---
title: "01-Pevermanni-lncRNA-discovery"
author: "Zach Bengtsson"
date: "2023-08-16"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# lncRNA Discovery for Porites evermanni

## HISAT

HISAT2 build to create an index for P evermanni...
```{bash hisat-build}
#FASTA file path input
ref_genome="/home/shared/8TB_HDD_02/zbengt/github/coral-lncRNA/data/Porites_evermanni_v1.fa"
index="/home/shared/8TB_HDD_02/zbengt/github/coral-lncRNA/ouput/Peve/Peve.index"

/home/shared/hisat2-2.2.1/hisat2-build \
-f "$ref_genome" \
"$index"
```

HISAT2 alignment...
```{bash}
find /home/shared/8TB_HDD_01/peve/trimmed/*gz \
| xargs basename -s -S1-TP2_R1_001.fastp-trim.20230519.fastq.gz | xargs -I{} \
/home/shared/hisat2-2.2.1/hisat2 \
-x /home/shared/8TB_HDD_02/zbengt/github/coral-lncRNA/ouput/Peve/Peve.index \
-p 20 \
-1 /home/shared/8TB_HDD_01/peve/trimmed/{}-S1-TP2_R1_001.fastp-trim.20230519.fastq.gz \
-2 /home/shared/8TB_HDD_01/peve/trimmed/{}-S1-TP2_R2_001.fastp-trim.20230519.fastq.gz \
-S /home/shared/8TB_HDD_01/peve/peve-sam/{}.sam
```

## Samtools

Samtools to convert the SAM files output from the previous HISAT2 command into sorted BAM files...
```{bash samtools}
for file in /home/shared/8TB_HDD_01/peve/peve-sam/*.sam; do
    base=$(basename "$file" .sam)
    /home/shared/samtools-1.12/samtools view -bS "$file" | \
    /home/shared/samtools-1.12/samtools sort \
    -o /home/shared/8TB_HDD_01/peve/bam/"$base".bam
done
```

## StringTie

StringTie to assemble transcripts from the sorted BAM files generated by the previous Samtools commands...
```{bash stringtie}
find /home/shared/8TB_HDD_01/peve/bam/*bam \
| xargs basename -s .bam | xargs -I{} \
/home/shared/stringtie-2.2.1.Linux_x86_64/stringtie \
-p 8 \
-G ../data/Porites_evermanni_v1.annot.gff \
-o /home/shared/8TB_HDD_01/peve/stringtie-assembly/{}.gtf \
/home/shared/8TB_HDD_01/peve/bam/{}.bam \
```

Use StringTie merge to merge all the individual GTF files into a single merged GTF file...

```{bash stringtie}
/home/shared/stringtie-2.2.1.Linux_x86_64/stringtie \
--merge \
-G ../data/Porites_evermanni_v1.annot.gff \
-o /home/shared/8TB_HDD_01/peve/merged-gtf/Peve_stringtie_merged.gtf \
/home/shared/8TB_HDD_01/peve/stringtie-assembly/*.gtf

```

```{bash}
head /home/shared/8TB_HDD_01/apul/merged-gtf/stringtie_merged.gtf
tail /home/shared/8TB_HDD_01/apul/merged-gtf/stringtie_merged.gtf
```

## GFFcompare

Use gffcompare to compare annotation file generated by StringTie to a reference annotation file and to produce a set of output files summarizing the results of the comparison, including classification of each transcript...
```{bash gffcompare}
/home/shared/gffcompare-0.12.6.Linux_x86_64/gffcompare \
-r ../data/Porites_evermanni_v1.annot.gff \
-G ../data/Porites_evermanni_v1.annot.gff \
-o /home/shared/8TB_HDD_01/peve/gffcompare/Peve_gffcompare_merged \
/home/shared/8TB_HDD_01/peve/merged-gtf/Peve_stringtie_merged.gtf \
```

```{bash}
head /home/shared/8TB_HDD_01/apul/gffcompare/gffcompare_merged.combined.gtf

```

## Filter

Filter to get a subset of the transcripts from the original GTF file that are putative lncRNA candidates based on their length and lack of overlap with known reference transcripts…
```{bash filter}
awk '$3 == "transcript" && $1 !~ /^#/ {print}' /home/shared/8TB_HDD_01/peve/gffcompare/Peve_gffcompare_merged.combined.gtf | grep 'class_code "u"' | awk '$5 - $4 > 199 {print}' > /home/shared/8TB_HDD_01/peve/filter-one/Peve_merged_lncRNA_candidates.gtf
```

```{bash}
head /home/shared/8TB_HDD_01/peve/filter-one/Peve_merged_lncRNA_candidates.gtf
```

## Bedtools

Get fasta of lncRNA candidate regions with Bedtools…
```{bash bedtools}
/home/shared/bedtools2/bin/bedtools \
getfasta -fi /home/shared/8TB_HDD_02/zbengt/github/coral-lncRNA/data/Porites_evermanni_v1.fa -bed /home/shared/8TB_HDD_01/peve/filter-one/Peve_merged_lncRNA_candidates.gtf -fo /home/shared/8TB_HDD_01/peve/get-fasta/Peve_merged_lncRNA_candidates.fasta -name -split
```

```{bash}
head /home/shared/8TB_HDD_01/peve/get-fasta/Peve_merged_lncRNA_candidates.fasta
```

## CPC2

Evaluate coding potential of transcripts with Coding Potential Calculator 2…
```{bash}
eval "$(/opt/anaconda/anaconda3/bin/conda shell.bash hook)"
python /home/shared/CPC2_standalone-1.0.1/bin/CPC2.py -i /home/shared/8TB_HDD_01/peve/get-fasta/Peve_merged_lncRNA_candidates.fasta -o ~/github/coral-lncRNA/ouput/peve_merged_cpc2_results
```

## Final Filter

Filter for those transcripts with label “noncoding”…
```{bash}
# Filter out transcripts with coding potential
awk '$8 == "noncoding" {print $1}' /home/shared/8TB_HDD_02/zbengt/github/coral-lncRNA/ouput/peve_merged_cpc2_results.txt > /home/shared/8TB_HDD_02/zbengt/github/coral-lncRNA/ouput/peve_noncoding_transcripts_ids.txt
grep -Fwf /home/shared/8TB_HDD_02/zbengt/github/coral-lncRNA/ouput/peve_noncoding_transcripts_ids.txt /home/shared/8TB_HDD_01/peve/get-fasta/Peve_merged_lncRNA_candidates.fasta > /home/shared/8TB_HDD_02/zbengt/github/coral-lncRNA/ouput/peve_merged_final_lncRNAs.gtf
```

```{bash}
head ~/github/coral-lncRNA/ouput/peve_merged_final_lncRNAs.gtf

tail ~/github/coral-lncRNA/ouput/peve_merged_final_lncRNAs.gtf
```

## Remove duplicate transcript IDs...

Get duplicate lines out of GTF file and save as new GTF...
```{bash}
awk '!seen[$0]++' ~/github/coral-lncRNA/ouput/peve_merged_final_lncRNAs.gtf > ~/github/coral-lncRNA/ouput/peve_deduplicated_final_lncRNAs.gtf

```

## Format GTF to BED

Format GTF to BED file...
```{bash}
awk -F":|-" '{print $2 "\t" $3 "\t" $4}' /home/shared/8TB_HDD_02/zbengt/github/coral-lncRNA/ouput/peve_deduplicated_final_lncRNAs.gtf > /home/shared/8TB_HDD_02/zbengt/github/coral-lncRNA/ouput/peve_deduplicated_final_lncRNAs.bed

```

## BEDTools

Use bedtools to get lncRNA fasta...
```{bash}
/home/shared/bedtools2/bin/bedtools \
getfasta -fi /home/shared/8TB_HDD_02/zbengt/github/coral-lncRNA/data/Porites_evermanni_v1.fa -bed /home/shared/8TB_HDD_02/zbengt/github/coral-lncRNA/ouput/peve_deduplicated_final_lncRNAs.bed -fo /home/shared/8TB_HDD_02/zbengt/github/coral-lncRNA/ouput/peve_bedtools_lncRNAs.fasta -name

```
