---
title: "02 Peve lncRNA align"
author: Steven Roberts
date: "`r format(Sys.time(), '%d %B, %Y')`" 
output: 
  github_document:
    toc: true
    toc_depth: 3
    number_sections: true
    html_preview: true
  html_document:
    theme: readable
    highlight: zenburn
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
    code_download: true

---

```{r setup, include=FALSE}
library(knitr)
library(tidyverse)
library(kableExtra)
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(data.table)
library(DT)
library(formattable)
library(Biostrings)
library(spaa)
library(tm)
knitr::opts_chunk$set(
  echo = TRUE,         # Display code chunks
  eval = FALSE,         # Evaluate code chunks
  warning = FALSE,     # Hide warnings
  message = FALSE,     # Hide messages
  fig.width = 6,       # Set plot width in inches
  fig.height = 4,      # Set plot height in inches
  fig.align = "center" # Align plots to the center
)

```


# Blast to sRNA, genome, proteome....


## lncRNA fasta 

```{r, engine='bash', eval=TRUE}
head ../../DEF-cross-species/data/peve_bedtools_lncRNAs.fasta
```


# make some blastdb


## sRNA

```{r, engine='bash', eval=TRUE}
ls ../../DEF-cross-species/data/blast/peve_sRNA*
```

## Genome
https://gannet.fish.washington.edu/seashell/snaps/Porites_evermanni_v1.fa

```{bash}
cd ../data/
curl -O https://gannet.fish.washington.edu/seashell/snaps/Porites_evermanni_v1.fa
```


```{r, engine='bash', eval=TRUE}
head ../data/Porites_evermanni_v1.fa

```

```{r, engine='bash'}
/home/shared/ncbi-blast-2.11.0+/bin/makeblastdb \
-in ../data/Porites_evermanni_v1.fa \
-dbtype nucl \
-out ../data/blast/pmea_genome
```

```{r, engine='bash', eval=TRUE}
ls ../data/blast/pmea_genome*
```




## Proteins

https://gannet.fish.washington.edu/seashell/snaps/Porites_evermanni_v1.annot.pep.fa

```{bash}
cd ../data/

curl -O https://gannet.fish.washington.edu/seashell/snaps/Porites_evermanni_v1.annot.pep.fa
```



```{r, engine='bash', eval=TRUE}
head ../data/Porites_evermanni_v1.annot.pep.fa
```




```{r, engine='bash'}
/home/shared/ncbi-blast-2.11.0+/bin/makeblastdb \
-in ../data/Porites_evermanni_v1.annot.pep.fa \
-dbtype prot \
-out ../data/blast/pmea_proteome
```



```{r, engine='bash', eval=TRUE}
ls ../data/blast/pmea_proteome*
```



## Genes
???



Blast comparison


# Result - database: sRNA

```{r, engine='bash'}
/home/shared/ncbi-blast-2.11.0+/bin/blastn \
-task blastn \
-query ../../DEF-cross-species/data/peve_bedtools_lncRNAs.fasta \
-db ../../DEF-cross-species/data/blast/peve_sRNA \
-out ../output/02-Peve-lncRNA-align/lncRNA_sRNA_blastn.tab \
-evalue 1E-40 \
-num_threads 40 \
-max_target_seqs 5 \
-max_hsps 1 \
-outfmt 6


```


```{r, engine='bash', eval=TRUE}
head ../output/02-Peve-lncRNA-align/lncRNA_sRNA_blastn.tab
```

(nothing)


```{r, engine='bash'}
/home/shared/ncbi-blast-2.11.0+/bin/blastn \
-task blastn \
-query ../../DEF-cross-species/data/peve_bedtools_lncRNAs.fasta \
-db ../../DEF-cross-species/data/blast/peve_sRNA \
-out ../output/02-Peve-lncRNA-align/lncRNA_sRNA_blastn02.tab \
-num_threads 20 \
-outfmt 6


```


```{r, engine='bash', eval=TRUE}
head ../output/02-Peve-lncRNA-align/lncRNA_sRNA_blastn02.tab
```


# Result -database: genome

```{r, engine='bash'}
/home/shared/ncbi-blast-2.11.0+/bin/blastn \
-task blastn \
-query ../../DEF-cross-species/data/peve_bedtools_lncRNAs.fasta \
-db ../data/blast/pmea_genome \
-out ../output/02-Peve-lncRNA-align/lncRNA_genome_blastn.tab \
-num_threads 20 \
-outfmt 6


```



```{r, engine='bash', eval=TRUE}
head ../output/02-Peve-lncRNA-align/lncRNA_genome_blastn.tab

wc -l ../output/02-Peve-lncRNA-align/lncRNA_genome_blastn.tab
```




```{r, engine='bash'}
/home/shared/ncbi-blast-2.11.0+/bin/blastn \
-task blastn \
-query ../../DEF-cross-species/data/peve_bedtools_lncRNAs.fasta \
-db ../data/blast/pmea_genome \
-out ../output/02-Peve-lncRNA-align/lncRNA_genome_blastn02.tab \
-evalue 1E-40 \
-num_threads 20 \
-max_target_seqs 5 \
-max_hsps 1 \
-outfmt 6


```



```{r, engine='bash', eval=TRUE}
head ../output/02-Peve-lncRNA-align/lncRNA_genome_blastn02.tab

wc -l ../output/02-Peve-lncRNA-align/lncRNA_genome_blastn02.tab
```


# Result - database: proteome


```{r, engine='bash'}
/home/shared/ncbi-blast-2.11.0+/bin/blastx \
-query ../../DEF-cross-species/data/peve_bedtools_lncRNAs.fasta \
-db ../data/blast/pmea_proteome \
-out ../output/02-Peve-lncRNA-align/lncRNA_proteome_blastx.tab \
-num_threads 20 \
-outfmt 6


```



```{r, engine='bash', eval=TRUE}
head ../output/02-Peve-lncRNA-align/lncRNA_proteome_blastx.tab
```


```{r, engine='bash'}
/home/shared/ncbi-blast-2.11.0+/bin/blastx \
-query ../../DEF-cross-species/data/peve_bedtools_lncRNAs.fasta \
-db ../data/blast/pmea_proteome \
-out ../output/02-Peve-lncRNA-align/lncRNA_proteome_blastx02.tab \
-evalue 1E-40 \
-num_threads 20 \
-max_target_seqs 5 \
-max_hsps 1 \


```



```{r, engine='bash', eval=TRUE}
head ../output/02-Peve-lncRNA-align/lncRNA_proteome_blastx02.tab

wc -l ../output/02-Peve-lncRNA-align/lncRNA_proteome_blastx02.tab
```

