---
title: "11-Peve-sRNAseq-miRdeep2"
author: "Sam White"
date: "2023-12-01"
output: 
  bookdown::html_document2:
    theme: cosmo
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
    code_download: true
  github_document:
    toc: true
    number_sections: true
  html_document:
    theme: cosmo
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
    code_download: true
bibliography: references.bib
link-citations: true
---


Use [miRDeep2](https://github.com/rajewsky-lab/mirdeep2) [@friedl√§nder2011] to identify potential miRNAs using _P.evermanni_ sRNAseq reads.

------------------------------------------------------------------------

Inputs:

-   Requires collapsed reads (i.e. concatenated, unique reads) in FastA format. See [10-Peve-sRNAseq-BLASTn.Rmd](https://github.com/urol-e5/deep-dive/blob/main/E-Peve/code/10-Peve-sRNAseq-BLASTn.Rmd) for code.

-   Genome FastA. See [07-Peve-sRNAseq-MirMachine.Rmd](https://github.com/urol-e5/deep-dive/blob/main/E-Peve/code/07-Peve-sRNAseq-MirMachine.Rmd) for download info if needed.

- [miRBase](https://mirbase.org/download/) mature miRNAs FastA.

Outputs:

-   Primary outputs are a result table in BED, CSV (tab-delimited), and HTML formats.

  - Due to the nature of mirDeep2's naming, trying to use variable names is challenging. As such, the chunks processing those files will require manual intervention to identify and provide the output filename(s); they are _not_ handled at in the `.bashvars` file at the top of this script.
  
- Other output files are too large for GitHub (and some (all?) are not needed for this analysis). Please find a full backup here:

https://gannet.fish.washington.edu/Atumefaciens/gitrepos/deep-dive/D-Apul/output/11-Apul-sRNAseq-miRdeep2/


------------------------------------------------------------------------