12-Peve-RNAseq-kallisto
================
Kathleen Durkin
2024-01-25

Description

Inputs:

trimmed RNAseq reads CDS gff

Outputs:

------------------------------------------------------------------------

# Create a Bash variables file

This allows usage of Bash variables (e.g. paths to common directories)
across R Markdown chunks.

``` bash
{
echo "#### Assign Variables ####"
echo ""

echo "# Data directories"
echo 'export Peve_dir=/home/shared/8TB_HDD_02/shedurkin/deep-dive/E-Peve'
echo 'export output_dir_top=${Peve_dir}/output/12-Peve-RNAseq-kallisto'
echo 'export trimmed_reads_dir=${output_dir_top}/trimmed-reads'
echo 'export kallisto_output_dir=${output_dir_top}/kallisto'
echo ""

echo "# Input/Output files"
echo 'export transcriptome_dir=${Peve_dir}/data'
echo 'export transcriptome_gff_name="Porites_evermanni_v1.annot.gff"'
echo 'export transcriptome_gff=${transcriptome_dir}/${transcriptome_gff_name}'
echo 'export transcriptome_gff_filtered_name="Porites_evermanni_v1_mRNA.annot.gff"'
echo 'export transcriptome_gff_filtered=${transcriptome_dir}/${transcriptome_gff_filtered_name}'
echo 'export genome_fasta_name="Porites_evermanni_v1.fa"'
echo 'export genome_fasta=${transcriptome_dir}/${genome_fasta_name}'
echo 'export transcriptome_fasta_name="Porites_evermanni_mRNA.fasta"'
echo 'export transcriptome_fasta=${transcriptome_dir}/${transcriptome_fasta_name}'
echo 'export transcriptome_fasta_collapsed_name="Porites_evermanni_mRNA_collapsed.fasta"'
echo 'export transcriptome_fasta_collapsed=${transcriptome_dir}/${transcriptome_fasta_collapsed_name}'
echo 'export kallisto_index_name="Peve_kallisto_index.idx"'

echo "# External data URLs"
echo 'export trimmed_reads_url="https://gannet.fish.washington.edu/Atumefaciens/20230519-E5_coral-fastqc-fastp-multiqc-RNAseq/P_evermanni/trimmed/"'
echo 'export transcriptome_url="https://www.genoscope.cns.fr/corals/data/Porites_evermanni_v1.annot.gff"'
echo 'export genome_url="https://www.genoscope.cns.fr/corals/data/Porites_evermanni_v1.fa"'
echo ""

echo "# Set filename patterns"
echo "export fastq_pattern='*.fastq.gz'"
echo "export R1_fastq_pattern='*_R1.fastq.gz'"
echo "export R2_fastq_pattern='*_R2.fastq.gz'"
echo ""

echo "# Paths to programs"
echo 'export kallisto=/home/shared/kallisto_linux-v0.50.1/kallisto'
echo 'export trinity_abund_to_matrix=/home/shared/trinityrnaseq-v2.12.0/util/abundance_estimates_to_matrix.pl'
echo 'export bedtools=/home/shared/bedtools2/bin/bedtools'
echo 'export fastx_toolkit="/home/shared/fastx_toolkit_0.0.13_binaries_Linux_2.6_amd64/bin"'
echo ""

echo "# Set number of CPUs to use"
echo 'export threads=20'
echo ""

echo "# Programs associative array"
echo "declare -A programs_array"
echo "programs_array=("
echo '[kallisto]="${kallisto}" \'
echo '[trinity_abund_to_matrix]="${trinity_abund_to_matrix}" \'
echo '[bedtools]="${bedtools}" \'
echo '[fastx_toolkit]="${fastx_toolkit}" \'
echo ")"

} > .bashvars

cat .bashvars
```

    #### Assign Variables ####

    # Data directories
    export Peve_dir=/home/shared/8TB_HDD_02/shedurkin/deep-dive/E-Peve
    export output_dir_top=${Peve_dir}/output/12-Peve-RNAseq-kallisto
    export trimmed_reads_dir=${output_dir_top}/trimmed-reads
    export kallisto_output_dir=${output_dir_top}/kallisto

    # Input/Output files
    export transcriptome_dir=${Peve_dir}/data
    export transcriptome_gff_name="Porites_evermanni_v1.annot.gff"
    export transcriptome_gff=${transcriptome_dir}/${transcriptome_gff_name}
    export transcriptome_gff_filtered_name="Porites_evermanni_v1_mRNA.annot.gff"
    export transcriptome_gff_filtered=${transcriptome_dir}/${transcriptome_gff_filtered_name}
    export genome_fasta_name="Porites_evermanni_v1.fa"
    export genome_fasta=${transcriptome_dir}/${genome_fasta_name}
    export transcriptome_fasta_name="Porites_evermanni_mRNA.fasta"
    export transcriptome_fasta=${transcriptome_dir}/${transcriptome_fasta_name}
    export transcriptome_fasta_collapsed_name="Porites_evermanni_mRNA_collapsed.fasta"
    export transcriptome_fasta_collapsed=${transcriptome_dir}/${transcriptome_fasta_collapsed_name}
    export kallisto_index_name="Peve_kallisto_index.idx"
    # External data URLs
    export trimmed_reads_url="https://gannet.fish.washington.edu/Atumefaciens/20230519-E5_coral-fastqc-fastp-multiqc-RNAseq/P_evermanni/trimmed/"
    export transcriptome_url="https://www.genoscope.cns.fr/corals/data/Porites_evermanni_v1.annot.gff"
    export genome_url="https://www.genoscope.cns.fr/corals/data/Porites_evermanni_v1.fa"

    # Set filename patterns
    export fastq_pattern='*.fastq.gz'
    export R1_fastq_pattern='*_R1.fastq.gz'
    export R2_fastq_pattern='*_R2.fastq.gz'

    # Paths to programs
    export kallisto=/home/shared/kallisto_linux-v0.50.1/kallisto
    export trinity_abund_to_matrix=/home/shared/trinityrnaseq-v2.12.0/util/abundance_estimates_to_matrix.pl
    export bedtools=/home/shared/bedtools2/bin/bedtools
    export fastx_toolkit="/home/shared/fastx_toolkit_0.0.13_binaries_Linux_2.6_amd64/bin"

    # Set number of CPUs to use
    export threads=20

    # Programs associative array
    declare -A programs_array
    programs_array=(
    [kallisto]="${kallisto}" \
    [trinity_abund_to_matrix]="${trinity_abund_to_matrix}" \
    [bedtools]="${bedtools}" \
    [fastx_toolkit]="${fastx_toolkit}" \
    )

# Download trimmed RNAseq reads

Reads are downloaded from:
<https://gannet.fish.washington.edu/Atumefaciens/20230519-E5_coral-fastqc-fastp-multiqc-RNAseq/P_evermanni/trimmed/>

The `--cut-dirs 4` command cuts the preceding directory structure
(i.e. `Atumefaciens/20230519-E5_coral-fastqc-fastp-multiqc-RNAseq/P_evermanni/trimmed/`)
so that we just end up with the reads.

``` bash
# Load bash variables into memory
source .bashvars

wget \
--directory-prefix ${trimmed_reads_dir} \
--recursive \
--no-check-certificate \
--continue \
--cut-dirs 4 \
--no-host-directories \
--no-parent \
--quiet \
--accept "RNA-*.fastq.gz" ${trimmed_reads_url}

ls -lh "${trimmed_reads_dir}"
```

# FastQC/MultiQC on trimmed reads

Already performed, can view multiqc report at
<https://gannet.fish.washington.edu/Atumefaciens/20230519-E5_coral-fastqc-fastp-multiqc-RNAseq/P_evermanni/trimmed/multiqc_report.html>

# Retrieve the reference transcriptome and genome

Provided by <https://www.genoscope.cns.fr/corals/genomes.html>

Download the gene CDS (coding sequence) gff file

``` bash
# Load bash variables into memory
source .bashvars

wget \
--directory-prefix ${transcriptome_dir} \
--no-check-certificate \
--continue \
--no-host-directories \
--no-directories \
--no-parent \
--quiet \
--execute robots=off \
--accept "${transcriptome_gff_name}" ${transcriptome_url}

ls -lh "${transcriptome_dir}"
```

Note that this is a CDS (coding sequence) gff file, not a FASTA, so
can’t input directly into kallisto. We’ll need the reference genome as
well to convert gff to FASTA

Download the genome scaffolds FASTA file

``` bash
# Load bash variables into memory
source .bashvars

wget \
--directory-prefix ${transcriptome_dir} \
--no-check-certificate \
--continue \
--no-host-directories \
--no-directories \
--no-parent \
--quiet \
--execute robots=off \
--accept "${genome_fasta_name}" ${genome_url}

ls -lh "${transcriptome_dir}"
```

## Verify transcriptome/genome FastA MD5 checksum

No checksum file(s) provided with download, so skipping this step

# Convert gff to FASTA format

``` bash
# Load bash variables into memory
source .bashvars

# The CDS gff we downloaded includes many different sequence types (CDS, UTR, mRNA, etc.). 
# We only want to the mRNA, so we first need to extract only the mRNA feature-type lines.
grep -w 'mRNA' ${transcriptome_gff} > ${transcriptome_gff_filtered}

head -n 3 ${transcriptome_gff_filtered}

# Now we can use bedtools to extract the fasta sequences for each feature listed in the gff
${programs_array[bedtools]} getfasta -fi ${genome_fasta} -bed ${transcriptome_gff_filtered} -fo ${transcriptome_fasta}

head -n 6 ${transcriptome_fasta}
```

    Porites_evermani_scaffold_1 Gmove   mRNA    3107    4488    543 -   .   ID=Peve_00000001;Name=Peve_00000001;start=0;stop=1;cds_size=543
    Porites_evermani_scaffold_1 Gmove   mRNA    424479  429034  2439.63 -   .   ID=Peve_00000002;Name=Peve_00000002;start=1;stop=1;cds_size=2019
    Porites_evermani_scaffold_1 Gmove   mRNA    429394  438909  1570.66 +   .   ID=Peve_00000003;Name=Peve_00000003;start=1;stop=1;cds_size=1458
    >Porites_evermani_scaffold_1:3106-4488
    TTACTGCTTCAGTATGTGAATTTCGATGGTGGCTTGACCGGAGTTAGACATGGCCCCCCTTGCTCGGAGTCCCGATCCCAAATCTCTCTCGTGCCAGTAGTTATCTCCTTTGAAGGGATTATCGTAATACATCCGGTACCAAAGATTGTAGTTGGCGCGTCTTTTACCACTGTAAAGCCAAACATCAAACCAGTTGCTGTACCAGTTGTAGTCAAATGGAACGGAATACATAACAGCAAGTGTTTTATCGATGTGTGGGATGTAGTATGTTAATACTCCTACTGCGCCTCTCGCAACGGGCCCCGCAGTTTTTCGTGCGCCGTAAAGCAAAGCTGTGCCTAGAAATAAGCAAACAAATTAGCAATTAAACAAAAATCATAAACATGATCATTCGTCACAATGATACATGGCAAACNTTGTGTCAGGTTAGTGATTGCTTATCTTTTCCCCAAATATTACACGCAAAGGGATTCTCATTCTGAACATGAATACTAAGTTTAGTTGGTTTAAGAAGCTCTTTCAACCCTGAAGTCATTTTGCCGTTCTGACAAAGCAAAACATAAACCTGTTGTCTGCAAAATTTCAAGAATGGGCAAATCCTCTAATACAGTTTCTTTACTCTTTAGGGTAAGTGGTCGCTGAGTTCGTCCTTGGTTTCATACGGTAATAGGCCCTTTGCACGCGATGAACTCATTTTACTACTACTACTACTACCAGAATCACTCGGGGTTTTGCTTTCTTGGACAAACGAGGGCTTTATTTTTCAAACCTCGCATAACCTGTGATCAGGCCCCCAAAACAAAAAGGGAAGAAGGACCGCCTGATTGCAGGTTAAATTCTTCGCAGAGATTAGCTAGAAAAATTAAATATGACAAGAAAACCAAAATGGATTCTATAGCAATAGATCAGACCTTTTTAGCTTGTATGGTTTGTTTACCCATTTCAGACCAAGTGATGGTACCCCAGAGAACTGTTTCTTTCAATGTCGTCTTTTGCACGTGCACCCACATGTTATGTATGCATAATTATATAAAAGACAAATTCCCCCAAGAACATCACGTGGTCTGAATTGGGAAAACAAAACAAACAAGCTAATGACACCATCCTGCATACTGGCCTTAGTGTTATTACTAAAAGAAAGGATTTTAAACTTTTATGTTATTAATATTAATTTTACCTGAAGAAACATCATGAGGCAAAACACGATTTGACGTCCCTGAATAAAAATATATGTTGACTGCTCTCCATTTATATCCACTTTCGTTATCAACACCAATGGCGACCTTGCGGCTTATGCTACCAAGGGTGTTTAAAATTGTTGTGAGAATGCCCAAGCCAAGTTGAGCACCGCTGATGACAGCACCAGCGTCAGCTAAAATTTT
    >Porites_evermani_scaffold_1:424478-429034
    TTACTGTTCAGCTAATTTGTAAAATAGCAAGTATGGCGAAGTGAATGATGACTCAGCTGAAGACAGCATTGCTGTTAAATCTGATTCCTCTAATATCATGACTTCAGCATCATCAAAGTGAATCCATTGAGATTTCTCAAGCTGTAAACTCTGTTCAACGGAATTTGATGATTCCTTAATAGAAACAGTGTTCTGTAATTTATTATCTTGTTTATTCTCTTGAAAGTTTAGCTGCCGAATTGCACTTCTGTTGGATTTTGCACCTCGATACCCAAAACTCTGGATTTTCTTTAAGCTGTGGGTTGGTGTAATATGATGTGCTTGTTTCTTGCCACCTACTGGTAAATCAACATTTTGATCCAGCTGTTCATTGGTTTCTTTTACAGGGTGCTTTCTTGATCTCTGAAAGTATCTTGAGATCCCAGACATACATGTTGTTGTACCTTTCCCTGCAGGACTCTCCAAAGCATGGATGTCATTTTCAGGATTGGTCATGTCCATGTTTTCCACTAATTTTCTGTCACTAGCCAAAGTGTCACAACCATCTGAAGTTCTATTGACTTCACTTGCAGGCATCTGACTCCAGTCAAACTCCCCAGAAGATACCATGTTTCTATGTACATGTCCATCTTTAACATGGTGCTTGTTGTCAGGATGGATGGGCTGCCCTTTGTTGTTATTTACATCGGCATTTGTGGGTGTGGTTACCTTAACGTAAGCCTGGTAATGTCCACTACATGAGGTCATACCACTGTGCATCACAATGCCAAATAAATGATACGTTGGATTGGCAATTGAACAAGTCTTTGAACACCATTCCTTGACTGACAGCTCCATTGGTGTCTCGAGATTTGAAGTGATCTTGGAGACCATTCCAGAAAATCTGCCAATTTAAAATATACACATCACTCATGTTTCCCCTTACTAATCTCTCAATGATTATACTGATTTGTCTCGAAGTGAATCATTGATCAGGGTAAAGTTATTGATGTATTGCACCATTGGTCAATGGAAGGACAAGAAACAGGTAACTACTCGTACTCATGTGAATCCATGAGAAAGACGTTAATCAATCAGAGGCATAAGTGGTTCAGAAGAAAAAATCTGAGTGTTTGCAATAGGAGTCGAACCTATGGCCTTTTGTTTACTAGTCCAAATGCTCTACCACTAAACTACCGGGGACTGGTGGGAGGTAAGAACACTAAACCACTGTAAATGCCTTTCAGGGCCCTGGGTTTAATTTGCAGCTGGTGCAGCTACAGTACAGGTGTGACTGATGGTGTTAGTATGAGGGTTCGAGTTTCATGACATTTAAGTTAGTCACACCTGTGCACCTGCAAATTAAACCCTCCAATTTTTGGGCCTGTTCCAGGTGTATGTATAGTAAGGATGGTGCAAAAAGGAGAGTAGAGCAAAAAAGGTGAGGGGTAACTTATTGCAACCTCTCTTCTCCGCCTGATGTGCAACTTCTATTAAATGCTGTTCTCTAAGTCCCAGAAGCGATCATTACTATCATTCTCTTTAATTTCTAATAAATGTGTAATTGATTAAAATCAGCCTCTGCAGGCAGGTTATGAGAATTGGGAACATGATCATAAGATCTATACATTTGCACAACTTATCTCTAAAAAGACTTGTGAGGGTACAAGAAAGGTCCTTCTTTTTTGACCTCAGGAAAGAAAAGGGTGCACAAAATCCTTACCCTGAAAGTGCGTTTGCAGTAAACCTCTTGAGGTGGATTGTGAGTATTCTTGGCAGCTTATCAAAGCTCGCTGAAATCTCAGCTTCAGTAAGAGTAAGGCAGTTCTCACAGAAATACTTGTTCTCTCCACTCAAATGTTCTACTGTTGCAAACTGCGACAGGGCCCAAGCTAAGCTCTGTGTGTCTCGGCCTTTCTTCGGTGTTGGTGAGAAAGCCTTCAGATCGCGCACTTTAGGCTCATCTTGAACTGGCACACTAATGTCTAGAAAGCTCTCAAAACGTTTCTTTGCATCTTCACAAGTCATACACTTGGTCTGGTGTAGAAGGCATCCTTGAAATAAATCTACAAAACTACTATGGCATTGGTTTGTGTTGCTGGTGAAGGAGTTCGCAGAAGGAGAATTTTTCACGTGCCTATCACCAATGTACATTTTTCGTGGATCTTCTCTCTTCTTTTTTATTTCTGTTTCTGTGTCTTGAAGATCTACTAGCAAACTGCACAGGAATTCTTGTGCATCATGTTGGAGATTGTCTTCAAACAATGGGTTTTTACCCCTATGTGTAAAAGAAAGGAAAATTTTTAAAAATTTTATGGCAATTTTTTGATGTGTTTGATGAGACAATAGCCATGTGACTGTGTGATTAACGGAGGTGTCAAACAAAAATGAGCTGATAGTCAACACGAAAATTATCGGCACATTTGTGGTGTGTCATTTATGTTGTAAGTATAATGGTGTTACTTTGTAACTGTATTAAGTAAACATGTATAATGAATTTCAAAAAACAACCAATTTTCACTACCAATAACCATGATACATGTATAAGCATCTACCTGAAGGTTTGCATGAAATCCAAAGGTTTTGCAGCAAGAACTAAAGTGTTTCTCCTTTGAACAAGATGTTCACAGTTCTCTTTATTGTCATTGTAGTCCTGTTCAAGCTCTTCCATCTGACTGAAAAGCTAGAAGGAAAAAGGATGCACTATTTAGTATATAAGTGATACATTCAAGTCAAAAATAGGCTAATACCCACCACTGAAATTGAAATTTGCTCCAAAAGCCACATGTAGATGGGTTACTAACTCAAACTCAGTTAAAAGGCATCAACAAAACCTGGATCTGTCCGGATCAGATACACTCCCTGTAATTCCTTAAACTTTTGTCACTACAAAAAAAACTCCAACTCGATTTTTATCTCAAAAACAAACTTCGCACAAAAATCCACAAGTAGCAAAGGCAATTCAAGGCAGTTTGTAAAATTTTAAAAAAATGTTACCGAAAACAGTGATACATGTAGTAAGACAAAAGATTAGGATTAAGGTTAGGGAAAAAAGGTTACAGATAGGCCCATTAAAAGAATTGATCTGACCTGGTTCAATCCAAGTTTTGCAAATGGCCCTCCCTTGAGTTTGAGTGGGAATCCATAAAATAAGGCTAATCCTGTGACTGCTTAGCCTTACTAGTTGTAAGTGTTATTTTTTCTTACCTCTCTTAGTTCACAAACAACATGTTTTTGATCACAACAATTTTCCAAATCCTCTGAACCCTGGCAAAAATAAAGACAAAATTACAGCATCATTGGTGTTAGATCCTAGTTTAAACCACTCAATATGGTATACATGTATTAAGCTATGTACATTAATGGAACGAAAAAAAGACATGCAAAACTATACAAAAAAAATTAGATAAGCTGCAGATATGCTAGCATTTCATACACAGTTAGACAAAATGTGGCAGCAGTTGGAGAAATACCAGTGATGACAATGATAATGTTAATAGTGTTAACAGCAATAACAATAAAATATTACCATTGGCACAATAGGCATTCACCTTGGTGTAAGGTTAATTCTGACAAAAAGCCAAATTGCCTAGACTTTCTAAATGTCCTTTTTTTCCTGGCTGGTTATGCCATTTTAAAGTTCTTTTTGTACCCATTTGGTGCAAAATTCACTGGTCAAAACTTCCACCAGTGACATTTTGACAATTAACCAATAGGAGAGTGAGTGATATTTGATGCCACTGCCGACACAGCATAAGCCAATAATTTTTCCCACATGAATTTATAGTGTGATTCAAATCCATTCAGGCAAAACAAGTTGACCTCGCGACTTGCTCAGTCACTGTGAGACTTCATTGGAGCTCACATTTAAGAACTAATGAGAGGTCAACTTATGGCAAGTCTACAAATTGCCCACCTCCTCACACAACCTTCTGCTAAGTGCCCCCCCTCCCCTACAATGAGACCACTTTACATCACCACCCTACCCTCCCATCCATAATTGTGGTGTATCTACCCTTTACGGAAACTTAGAGGGAAGGGGATGGAGATTTATCTTAAAGAGTAACACGTACGAAAAATGTATTGAATTACACCCACGTCCCCTTACGCACTCAAACTTCACTTTCCATCGCAAATAAGCCTAATCTTGTAATGCAACACTGAACCTAAGCTCCACTGCTTACGTTTTCTTTCCCGTCTGTTCTCGTGTCGAATGATTTCCCCTCTGTCGCACAATAGTTTTTTTCTTTGACAAGTGAATCGAGTTCATCAACAGACTCTAATATTCCCGGGCAATGCCGCAAAACTTGCACAACAGCGTTTGCGTAGCAAATATTTCCATGATTCCTTAATCCTGCGTAAGGTGGTGGTGGCTTTGGGGGCTCACATGTCTCATCGTCACTCGCATTTTGGGCTGTTGGAGGGCTGAAGATAGGATACAACTTCTCTCTCTTGTCAGGATCGCTTTCGCTCGAAGCTTTCTTTTTGCTTTTGAGCTTTAACGAGAATCTTGTTTTCTTACGTGGTGGGCTCATGGGACTCGACGGCGGTGAAGCTGACCTAGAAACTGATGGCAT
    >Porites_evermani_scaffold_1:429393-438909
    AAGTTATCCGGGAAAAAAGAAAATGGCGACGGCTAGTTCGCTGGTCAAATGTTATTGAGTCGTCTTTTAGCCTACACCTTGGGGTAACAGAATAATTTTAAAGAACATGTTCAGTGCAGAGGAGCTGGATGTAAAGACTACCCAGCTCAAAGCCCAAGGGCTGCTCTCTGAACCTTCAAATCACGAGGAGGAAGAAGTAGAGAGAATTGAAAGAGCGACTTGTTCTTATTCATATGAATCTGAACGGTTTTTATTGAAGGAGAGGAGCTTTTCAAGGCAGTATGCTCAGATTTATGCAGTACGTATTATGGCAATGAGGAAGAAGTTGGCTGCAGCCGCCAGGAGAAAATGGGGTTTGTGTAACATTCTCTTTGTATCATATATAAAATATTTAAGCCTTCATTGTTTTAATCATAAATGACCGGACTGTCACATCGATTGCGTTGAGAAGAGAAGGTTGTATTTAATTTAAGATTATCTTGTTTTTCACGGTTGCTGCTTACAAGCTGTAGTACTTAGCCGCAGCTGATGTACAAAAGGGTTAATTTTTCGCTTTCAGGTTGGTTGGGGGGGGGGGGGGGGGGGAGGGGGGGGGGGGGGGGGGGAACCAAAACTTGGGGCCGCTTTTTTTTTTCCGAAAGAACAAAAAGCAAAAATATTGGTTCCTTTTTACAAAATTTATATTTTAAAACAGAAAACAGATGAGTATCTATCAAGAATGTTGTTGTATTTGTAGGATTATTGTTTTTTGGTATGTTCTTCCGAATCAGATGAGGTCTTTCGCGGGGGGTGTAGGGAAAGGGGGGTTTGTTTTGGGTGGGGGGGGGGGTGGGGGGGGGGGGGGGGGGCGAGGGGAGGGGGGTAGTGGGATACCAATACTGGGACCCGATTTCTTTGTCCTGAAATACCAAACTGCAAAATTACTGGTACCTTTTACCAAAATTTATATTCTAACGCAGAATACAGTAGTGTATACCTGAAAATTTGGCAACAAAGTACAATTCAGTTGAACTTCCATGTGCGACCACCTCTCGTTAGCAGCCACCTTCCATAGGTGACCACCTATCCAAAACACTTAAGTTTTCCCAGTCAAAGCTTTACAGTTGAAACCTCTCAGAAACAACCACCTCCTGTAAGCGCCCACTACCACATATCATGGCTGACTTTCCATAATTTTCCTCTGTTTTTCACCTCTTGTAAGCAACCACTTGACAGATGGTCTGATCTCTATGTTCGCTGTTTGTACTATGCCACTTCGTATATAAGAAGAACTTTAAGCAACAACATGGAACTACTCATATTGCAACTTAGTAATTGCATGAAATAAAGATGTGTGTGGACCTTTTCTTGAAAGAGACACCCTGTGGTTCGATTTTTGCAAATGACCACTTCCTTTAAGTGCCCTTATGGGAGGTTCAACTGTATATACTCAATGGCCCTGTGTAGCTAAGGGTGAAAAGTTTGATTTCCATTTTATGCAATCCAATTTAGGTAACAGGTTTGAAATAAAGAAGAAACTGGCTGAGGTCAAGGGAAATGAGAAATGTGTCCTGATAGGAACTCTGTTCAAGAAAATGGAACTCAAGCCTTGTATATTAAAGGAGATAAGCGAAGAGGTGCCTAGTTTAAAGAAGTGCTATGTTTAAGAAGTAGATAATGACTGATTTCAAAAACGAGACAGTCTGAAAAATATTGGATGAAAGTGTTGGTATTGGCTCTGTGCCTTTGTTTGTCATATGCCACATACACTTTCCACATTGTTTCAAGATCAGGTTTCAGAATTTGACTACATTTACACTGGTGGACAACTGCTTTTCAACTTTTTTTTTTCGCTGCCTTACCAAAATTTTGCCTGTCCTGTTTTAAAAAGCAAATACACAATGAAGTGGAAATACAGTAAAATGTAATAAAAATGTACTATTTGTCCTTGCCGAATATCTTTCAATCCATTGAAGCAAATTTAATCCAAGGAAAATTATTTTCCCACTTTAGCAGAATTTTTTTTCTACTCCAGATGTGCTATCATACTTTATTAAAACACCAACTGAATGGTTTTGACAAACTTTCCGAGAAGACAGACAGTTATGATAAGGATAAAAACAACATTGTCACAAATGGATGGTTGCTCTTTCTTGGCAACATCCTTCCCAGGTTCCTGTCCTACTCGTCCTCCAGGGGAGGAGTTGGATAAGCCTCTAGGAATGAGATTGCTTTCTCGGGATTGTTTTGCAATAGTGGTGCTTTGCCAGAAGAGCAGAAAAATTGAGTTACACACAAGTCATGTTTAGGGACTGGGAGATCTTTTACATAGTCTCAATGGAAGGATATCACTGAACATGGGCTGCTTATCATACAGTTGAGAAAGGACAAGTTTTGTATTCAATGTTCTCTCTATTCTCTTTGTAGGGAGTTTCAATATATATTCATTGCTATTTAAAACAGGTTATCACTGATGTACACCCTAAGCACTAGAAAATTCCCTGTACAACTGGAAGATAGTTGCCTAGTAAAAAGGCAACTGGGCTGAAAATGTTCCTTTTTCTTGAAAAATATCACCTGCCCCAGGAAATCAGTTACTGTGGATTCACCCAGAAGTTTAAAATGACTTCTTATTTTCAATAATTTAAGTTTATTTCATGGTTCTTAACAGCATAACCTGATGCCTCAACCATCCAGAGTAAAATATGCAGATGACAGTGATGAACTTGTGATTGAGGATGAATCAGAGAGGGTGACTCTTACTGGTAACATTCCAGTCGGGACGTCAGTAACAGGTGACTCAACTTAAAATAATTCAATATCTAATCCTCCTGTTAATGATATCAATACTTGATTTGCTGAATTTATATTACATTCATTTGGTGTGTTTTGCTGTGGAAAGGAAAGAAACTGGGGTTTTGACCCCATGGACTAGTACCACCCTTTGGAACCTTACGAAGTAGGTTTAAATCATATAAAGGCTGTATGTTGAAATAGTGGTGTAATTATCATGAAATGATGGGATGATTGATTGGTAAAACTGGGCTTATCAATTGTGGTAATTTTCTCCCTGTTACATATGACAATGAGTTTGCAATCCCAGCAACAAAGGAAAAAAATTATTTGAACCAAGATTTAAAATGAATCACAGCATTATGGAATGATAAGCAATGGAAAAGAATAAAATTAAAACCTTTATGACTTTCTCTTCCAGTTTGTTCTTACAGTATTGAATGTTTTTTTTGCCATCTTTAGGAACTATAGTAGCTCTCTGTGGCCATACAACTGAAAGTGGCAAATTCCATGTTGATGAGTACTGTTTTAGTGGTCTTCCATACCAAACTTCACCCAACCTTGATCAGTGTTTAACCAAGGGTGAAGACAGGTAAAATGACATCAGTCACTGTGATATTGATTAATGTAATGTTTCTCATTCAGTCAGTCTTCTTCAAGTTGTTTTTGCATGAAAAAAAAAATTAGAAATGTAGACAAGACACCCAAGACACTTTTTCCTCAATAGATGAAAGATAGGCACTACTGACTATTACTGGTGGTGGTTTGCAGGCAATGGTCCCCCAGTGATCTGTTATGTCCCTAGTAGTGTGGTCCAGGGTGCTTTCACATCCCACAAGATTCAGATCTGTGAAGTGTTTGAGACTGGGCATGCAGTTTTTTATCCCTTACCGAGAAGTCTAAAAAATATAGAGATGTTTGCGCGATGGCGCAAGGGGGGGAGGGGGGGTGGGGGGTTGGAATCCAGGGATTTTCATTGGAGGGGTGCATAAATTGTGCTCTCCCTTAGAATTTAATTTTACAACTTGTTTACAGATTATCATTGCCACCTGTCACAAGCAGAGCTATCTAACAGGCAACCCTTCAATGCCAGTGCTTTCAGTATGTAAATTCTATAAAATTAATGTTTGTAGTCTGATGAGTGCATTTAGGGGTCCGGGATCCTCCCCTTGGAACTGTCACTCTGGCAATACAATCTCCTCAGTTACTCTAAGATCCTAAGTATTGGTCCAGCCAGAAATCAGGCCTGGACATACCACTCCAAAAGACAGGCACCTGAACTAACTGAGCTAACTAATTTTTCTCATCATTTGTCAGGTATGTGGCACTCATATCTGGTCTTGGCTTTGGCTCTGAAAGCCAGGATATGCTCAGCCTACAGATGTTCTCTGATCTGGTGACAGGAGAGCTTGGGAGTGTTCAGGTAAGATTCTTCTGTTTTACCTAATTTTGCTAACACTTGCCCAGTCCAGGCCCAAGTTGTTCAAAAGATGGATAGCGCAGTAATTGGTTTTGGTAACTTTTATCCACTGATTATAACTTTTATTAGTGATTAATAGCACCAGTTTATAGCAATAGTAATTCCTACAAATTGTTACTGAAAGTGTTACTCAATCAACTACAGAGCATCTGTGCCCTAGTTGTGCTGAATATCTTATTGACTGTGTAAAACAGGGAATATTGTACTAGTCTGTACCCATTCCCCTCATCAAATGAGTTTCATACTGTATTTTATCCTTTTTGAAAAATGAAGGCCCAACAACAAAAATTATATTTCTAATAGTCTAACACTAATATTACTGCAGCTTCCAGAAGGTGTTTAAAAGTAATGCCACTTAGCTTTAAATAGCCTCAGTTGGTCACATTTCAGTAACCGTCTCCAATGAAAACAGTAAGTCTGTGCAAAAAATTTTGTGTTTTATTAAAAGCTTAACAAGTCAGACCACCTTGCCTTTTTTGTGGACTTTTAAAGAAAGCCCACACTTTAATATTCTTGAGTAATTTAACAGGCTCTTTTGCACAGTAGCAGCTTCCTTTGTACTCTAGTTTGGCATGTTTTCTCACCATGATTTGTGTTTTAGGATCAAGAGTCATGTGCCAAAATAAGCCGTGTCATCATTGCTGGTAATTCTTTGAGCCAGTGTACTCAAACAAAAGAGAGTGAAACTAAGGTAACAAAAAAAAAATTCCTTTTCTTTTTGCTGGCAGTAGTTATCAGTATTTAGCACTTCTAGTGATTTGATGGATCAGTACCATCCCCCTTTCTCATTTGAGGCATCTGTGCATCTTCAAGTTTTCCTGCATCCTGTGTTCTTCTATTCCAACGCATTCTAAAAATCAGAATCCTGGTTCATTTGCACAAAATATTCACTAGGGAGATAACAGTCTGGATCTGACAGTCCCCAACCCTCCCCATCATTGACTGTTGGAATGGGTTGGTGAGCATCTTGCATGGGACCCATTCCATCCCCACTGCGTGAAAGGCTCCATGCAAAAACATCTCCTAGCCTACTCATCTCCTTTGTGTCTTTGCATGGAGCCAAAAACTACTGATGTATATCTAAGAGTTTGCTCCTACCCCAATTTGGAGAAAGCTGTTTGAAACAACTCCAAAATTACATGAGCTTGGTTTTCATAAACCACAATGCTTTGCATGATAATTGTGTGTAAACAGCAATAAATGATGATGATGATGATGATGATGATGATCAGCATTGTCTTCATCATCATTACATTATCATCATCATCATTATCATTAAGGACAAATTATCAAAAAGAGTGAATAATCTTACCATTAATATCTTGTATCACGTACGAGTTTGTTTTGACCACCTCCTGTCGGCTGATGTCTTATTTCATTTCATTAATGAGTTAGGTCAGTTACTGAACTTCTTTTGCCTGATCAATGGCTATAACAGCAAGGTATATGCATGTGTATCAGACTTTTTTTGTGTCATGACTGAAAAAAAAAACGATATATACAAATGTGTAGTGCTTGTTTCTTTTTGTAAGTGAAACTTGAACTTAAGTTGTTAGCAAGTGGGGAGTCAGCAAAACGAAGCATAGGGAAACAAGTGGTCACTGACAAAGTGACTTTCACGAGATAAGCTCCAAATGAGCCCTGTCATCCACAAACTATGGTCACTGATAACAGTATTCTCTTTTAAACTCCCTAATGATCTTTTTGTGACAGGCAAAATATTTAACAAGAAACACAGTGGCAGGAAGTGTAGAGGCTATTAAGAGTCTGGATGACTTTGTTCTACAATTAGCAGTAAGTGACTGATCATCTTTCCATAAAGTCCAAACAAAACATCTTACGTGAATGCACAATGTTAAAAAATTGAGCCTCTAGACACAAGGAAAATAATAATTACTGAGAACGCAAGACAATAATGTTGAAGGTCTTGGTGAAAAAAGGGCACTGTTAGTGATTACTTTTCTGGAAATAAAAAGTTGCTCAAAAATGTCAGAGCCAACATCAACATTTCAGTATATAACTGACAGTTACCAGTACTGGTCCTTCCAAAAAAAAAATGAAAAGTGATTGACCACCCCCGGAAGCTCTGTTTGCCCTTCTATAACCTGCAAGCAACCTTAAGAACATTTTTGTGTAAGAACAATATTTTATATTATGTTTCTTTATTTCAAATGCTATTTCTGTACCTTAGATGCAATCACAAGTCATACCCAGTCATTTTTCAAAGTCTTGTTGCAGAAAGTAGTATAAAATAATGTGATCAATGATGGACGAAGAAGGAAAAATCGTAAATAACTTCATGATGTGTTTTAGCCAATAGGGTGGGGGGTTAAGGAATGCCGATTTATTAAATAAATCTTAACCACTGAAAACTGTAATTTGTCCTTGAAAAATCCTTGAAAAGTCCTTGAAATTTTTTTGTTCAATTCCTGTCGGGAACTCAGATTTTTTCTTTGTCCCATGCTCGTGACATGTTGATTACACCATTTCTCATTTCTTTACCGAGCTCAAAATTTACCATCTTTCTTTATTTATCACAGCCCTGTTAATCTTTCTTGTTTTCCTTTACAGGCATGTGTTCCTGTAGAAATAATGCCAGGTGAATTTGATCCAGCAAATCACACGATGCCACAGCAGCCCCTCCATCGCTGTATGTTTCCTCAGGCAGTGGCTTATTCTACTTTCCAAAGTGTTACCAACCCATATGAGGCTGTAATTGGAGGTGTAAGGTAGCCAAAGGAAATTTACGTAATATCTTTTCATATAAGTTTATCTAACCAGCCTTTCATTAGCCGGGGACCAGGCTCCTAAGTGAGGGAAAAGGGCTAAAAAAAGGGCCAACAAAGCAAGACAAGTGGGAGTCTGGGGAGGGAGAGGGTGGCCGCTGCCCTTTCTCTCTCCCCAGTCGACCACTAGGCTCGCTTTGCTCACCAATTGTTTTTTATTTGTTTGTTTGTTTGTTTTTTTTGTTGTTTCACCCAAGTTTTTGCCTTTTTCCTTCACCACAGAGCCTGGTCTCAGGCTAGTTTTTTGTTTAAACCCTGAAAAAAATGGGGTCATTGCCATGGTAACATGAATAATTGAAAACAAGCCTAAAATTTTGAATTTATTAAGTTTCCAAAAAATTAAATCATTAGATTTTCTTACAAATTTGTACATAGCTTCCATAGCTTTCCTGGAAATTAATGTTGATTGTATCACAAAACCCCTTCGTTCAAAATTTGTTTTGAGGCTTTCATTGAATAGGATTTCTGCCAAATTTCGTCAAATTTTGAAATAGCATGTAGGTTTTAAGATATTGTCATTTATTTACAATTATTCTTATAAAAGTCAACAAAACAAGCACCCTCACATTGCTTCTTAATTACCTATGCTCAGCATTTGTGCGTCCGTTGAGGTAACATGATATGTCAAAAAGACAAAAACAATACTAAATAGCACGTTGTCACCCTTTTTTCCTTAAAAGCTTATTGGATGCATACCTCCCCATGTTTAAGATGTTAAGATGAACAGTAAAGGTATAAATCAGATCTGATCAAAACGTCTGGGGTGATAGGCTTCTGATGTCAAGAACCTAGTCGTAATTTGATAGGGACCTAACACTCCTCTGTTGAATGTGGACAGTTTTTGTTCGCCAAAACAATTTTTTAATTTTTTAAGCTGACAAAGTATTCTTTGCTACTGAAAGGTTTTGTACGTGTTCAAAAAGGATGTTAAAGAAAATTATATAATTTTCAGTCTCCATCAGTACTTTGATTGACTTTGTTTTGATTTTTGATGTCAGGGTGCTTGGTACTGCAGGACAGAATGTAAATGACATCTTCAAAGTGTCCACCTTTGAAGATCGCTTGGAGATTTTAGAGCACACTATGAAATGGAGTCACATTGCGCCCACAGCTCCTGACACCTTAGGTAAATTGTATTGTATTGGTCTAAGAACTCCAGACTTGTAAAAACGTTTTAAAATCAGATGGAAAAACTTCATCAAAATACTCTTTTTAGTGTCAGAGCATTGATTCTTGTCCTCATGCTGCTCAGAGACAGGGATTTGAATGAGAATTTTTGAAGAGGTGGTCACACATGGAGCTTTGAATTAAGCCTTTCAAACACCAAAGATAACCTTTGCACCTCATTGTGACTAAATACACCTTCCAAGAAATTCTCTTCCTCCCAGTCCAAACTTGTTCAATAACAAAGGAAGCGACTGCACGTTGTCCATTAGCTGGGACAAATATGCCAGTGTCAAGTTAACAAAGTTCATTCTTGGATCCGAGGCTGAGGGGAATTTGAAACAAAAGAAATGAATTATTTATCCATGAATCTCAAGGCAATTTCTTTTGTTTTATGCCCCTCAGCTTTGGGTCAAGTAGGAACTTTAATGATTCGAAACTGGCCAGTTGCTGGCCCTCATTAGCACTTCTGCAGTGTATCTGCCTTATACATGGCTATCTGGTTTTGCTAACAATTGCTATTTCAATTGTTCAGGTTGTTATCCCTATTATGAGAAGGATCCCTTCATTCTGGACAAATGCCCACATGTGTGCTTTGTTGGAAATCAGCCTAAGTACCAAAGTAAAACTATTCAAGGTAAAGTCCCCTTGTCAGTGTTTCATTTTACTATAATACAAGCAGCTTACGCATAATCACGGTTCTCATGTTGGTAGGTTTAAAGGGCAACAGTTTGCTGCTTTGTTGTAGCCACTCTCCGACTTTATTAGCCTCGTACGCAGAAGTTCTTAGGGCTTTGTCATGCGTGAATTTGTGACGAAGCCCTAAGCGCAGGGGAGGAAAGTGTGATAAAACCCTAAAAATTCTGAGGAGGCTACCCCTTTATAGGATGAAAAGGTAATGATGTTGATAAATTCCATTCTTTGTTTGTTGTGTAGGGGAAAATGGCCAGAAAATACTGCTGATAACAGTACCTACCTTTTCCAATACAAAGACATGTGTACTGTTGAATCTGAGGACACTCTCGTGCCATCCAGTGAGCTTCTCTGCCTCACTGTCAACTCAACAAGACAACAGTACCAGTGATGAAGAGATGGAATTGTGAATATGAGCCAAGATTAAGAGTGTACATTCTAGTACATGTATATTATGTTAGCATCAACACAAGAACACAATAGGTCATCATTCTGACATTGCTGCACTGGTATCTACCTATAACATATAAGCAGCAATTAGAGAAGGATTGTCCAGCTTGTACTTTCTTCCCAAATCTGTTCTGCATGAACTAATACTGTAGCATAAAAAAAAATTCATGA

# Align to reference transcriptome (Kallisto pseudoalignment)

## Building Index

``` bash
# Load bash variables into memory
source .bashvars

cd "${kallisto_output_dir}"

${programs_array[kallisto]} index \
--threads=${threads} \
--index="${kallisto_index_name}" \
"${transcriptome_fasta}"
```


    [build] loading fasta file /home/shared/8TB_HDD_02/shedurkin/deep-dive/E-Peve/data/Porites_evermanni_mRNA.fasta
    [build] k-mer length: 31
    [build] warning: clipped off poly-A tail (longer than 10)
            from 4 target sequences
    [build] warning: replaced 10722052 non-ACGUT characters in the input sequence
            with pseudorandom nucleotides
    KmerStream::KmerStream(): Start computing k-mer cardinality estimations (1/2)
    KmerStream::KmerStream(): Start computing k-mer cardinality estimations (1/2)
    KmerStream::KmerStream(): Finished
    CompactedDBG::build(): Estimated number of k-mers occurring at least once: 241599983
    CompactedDBG::build(): Estimated number of minimizer occurring at least once: 57847036
    CompactedDBG::filter(): Processed 281910300 k-mers in 40389 reads
    CompactedDBG::filter(): Found 241037665 unique k-mers
    CompactedDBG::filter(): Number of blocks in Bloom filter is 1651564
    CompactedDBG::construct(): Extract approximate unitigs (1/2)
    CompactedDBG::construct(): Extract approximate unitigs (2/2)
    CompactedDBG::construct(): Closed all input files

    CompactedDBG::construct(): Splitting unitigs (1/2)

    CompactedDBG::construct(): Splitting unitigs (2/2)
    CompactedDBG::construct(): Before split: 3234002 unitigs
    CompactedDBG::construct(): After split (1/1): 3234002 unitigs
    CompactedDBG::construct(): Unitigs split: 940
    CompactedDBG::construct(): Unitigs deleted: 0

    CompactedDBG::construct(): Joining unitigs
    CompactedDBG::construct(): After join: 3111634 unitigs
    CompactedDBG::construct(): Joined 122663 unitigs
    [build] building MPHF
    [build] creating equivalence classes ... 
    [build] target de Bruijn graph has k-mer length 31 and minimizer length 23
    [build] target de Bruijn graph has 3111634 contigs and contains 241249799 k-mers 

Note that, when building an index, kallisto warns us that it “replaced
10722052 non-ACGUT characters in the input sequence with pseudorandom
nucleotides.” This high number of identified “non-ACGUT” characters is
related to the type of reference sequences we used to build the index.
We obtained a coding sequence (CDS) gff file for P.evermanni and the
associated scaffold genome fasta, filtered the gff to retain only mRNAs,
and then used bedtools to extract the fasta sequences of every mRNA from
the scaffold fasta. Notably, scaffolds are basically fragments of known
DNA sequences “stitched” together by stretches of Ns to approximate the
full sequence structure without complete sequence data. This means some
of our mRNA sequences contain long, relatively-meaningless stretches of
Ns. This shouldn’t interfere with the kallisto pseudoalignment process,
so we will continue.

## Sample Quantification

Kallisto can run quantification on either single- or paired-end reads.
The default option is paired-end, which requires the input of an even
number of paired fastq files (e.g., pairA_R1.fastq, pairA_R2.fastq). To
use single-end mode, include the –single flag, as well as -l
(–fragment-length=DOUBLE, estimated avg. fragment length) and -s
(–sd=DOUBLE, estimates stand. dev. of fragment length), and a number of
fastq files. Again, gzipped files are acceptable.

Kallisto quant is rather finicky about how you input sets of paired
reads, and you can only input a single pair at a time. To circumvent,
I’ll create symlinks to each of the input files with simplified names,
create a quantification function, and apply it iteratively to each pair
using a loop.

``` bash
# Load bash variables into memory
source .bashvars

# Create sym links to each of the trimmed read files with simplified names

for file in "${trimmed_reads_dir}"/*.fastp-trim.20230519.fastq.gz; do
    # Extract sample ID and read number from the file name
    sample_id=$(echo "$file" | grep -oP 'RNA-POR-\K\d+')
    read_number=$(echo "$file" | grep -oP '_R\K\d+')

    # Create the shortened name
    shortened_name="sample${sample_id}_R${read_number}.fastq.gz"

    # Create symbolic link
    ln -s "$file" "${trimmed_reads_dir}/${shortened_name}"

done

ls -lh ${trimmed_reads_dir}
```

``` bash
# Load bash variables into memory
source .bashvars

# Function to run kallisto quant. Takes two (paired) reads as input, outputs to sample-associated directory
run_kallisto_quant() {
    source .bashvars  # Source .bashvars inside the function to make its variables accessible
    local R1_fastq=${1}
    local R2_fastq=${2}
    
    cd ${kallisto_output_dir}
    sample_num=$(basename "${R1_fastq}" "_R1.fastq.gz")
    mkdir kallisto_quant_${sample_num}

    ${programs_array[kallisto]} quant \
        --threads=${threads} \
        --index="${kallisto_output_dir}/${kallisto_index_name}" \
        --output-dir="${kallisto_output_dir}/kallisto_quant_${sample_num}" \
        --bootstrap-samples=100 \
        ${trimmed_reads_dir}/${R1_fastq} ${trimmed_reads_dir}/${R2_fastq}
}



# Iteratively apply run_kallisto_quant on each pair of input reads
for file_r1 in "${trimmed_reads_dir}"/*_R1.fastq.gz; do
    # Extract the sample name from the file name
    sample_name=$(basename "${file_r1}" "_R1.fastq.gz")

    # Form the file names (function takes input file names, not paths)
    file_r1_name="${sample_name}_R1.fastq.gz"
    file_r2_name="${sample_name}_R2.fastq.gz"

    # Check that the sample hasn't already been quantified
    if [ ! -d "${kallisto_output_dir}/kallisto_quant_${sample_name}" ]; then
    
        # Check if the corresponding R2 file exists
        if [ -e "${trimmed_reads_dir}/${file_r2}" ]; then
            # Run kallisto quant on the file pair
            run_kallisto_quant "${file_r1_name}" "${file_r2_name}" 

            echo "Processed sample: ${sample_name}"
        fi
    else
        echo "Sample already processed: ${sample_name}"
    fi
done
```

    Sample already processed: sample71
    Sample already processed: sample73
    Sample already processed: sample76
    Sample already processed: sample79
    Sample already processed: sample82

## Trinity Matrix with Kallisto Output

``` bash
# Load bash variables into memory
source .bashvars

cd ${kallisto_output_dir}

${programs_array[trinity_abund_to_matrix]} \
--est_method 'kallisto' \
--gene_trans_map 'none' \
--out_prefix 'kallisto' \
--name_sample_by_basedir ${kallisto_output_dir}/kallisto_quant_*/abundance.tsv
```

# Summary
