---
title: Identification of differentially expressed transcripts in P.verrucosa exposed
  to enriched nutrient supply.
author: "Sam White"
date: "2/24/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Use [Ballgown](https://github.com/alyssafrazee/ballgown) for identification of differentially expressed isoforms in _P.verrucosa_ exposed to elevated nutrient supplyu.

REQUIRES Linux-based system to run all chunks properly; some chunks will not work on Mac OS!

REQUIRES the following Bash programs:

- `wget`

- `tree`

- `md5sum`

REQUIRES the following R libraries:

- [`Ballgown`](https://github.com/alyssafrazee/ballgown) (Bioconductor)

- `tidyverse`

# Load `R` libraries
```{r load-libraries}
library("ballgown")
library("tidyverse")
library("ggplot2")
```


# Set user variables!
```{r set-variables}
# Set maximum pvalue for isoform expression cutoff
pvalue <- 0.05
# Set maximum qvalue (false discovery rate) for isoform expression cutoff
qvalue <- 0.05
```



# Download Ballgown input files.

Notebooks detailing their creation:

- [FastQ trimming](https://robertslab.github.io/sams-notebook/2023/02/15/FastQ-Trimming-and-QC-P.verrucosa-RNA-seq-Data-from-Danielle-Becker-in-Hollie-Putnam-Lab-Using-fastp-FastQC-and-MultiQC-on-Mox.html)

- [Genome indexing, and exon/splice sites with HISAT2](https://robertslab.github.io/sams-notebook/2023/01/31/Genome-Indexing-P.verrucosa-v1.0-Assembly-with-HiSat2-on-Mox.html)

- [Mapping and identificaion of isoforms with StingTie]()

```{bash download-ballgown-files}
# Download Ballgown input files and directory structure
wget \
--directory-prefix ../data/ballgown \
--recursive \
--no-check-certificate \
--continue \
--cut-dirs 2 \
--no-host-directories \
--no-parent \
--quiet \
--reject "input_fastqs_checksums.md5" \
--accept "*.ctab,*checksums.md5" https://gannet.fish.washington.edu/Atumefaciens/20230216-pver-stringtie-Pver_genome_assembly_v1.0-isoforms/
```

# Verify checksums

NOTE: Warnings are expected, as the checksums files have checksums for files that are not downloaded for this project.
```{bash checksum-verification}
cd ../data/ballgown

# Remove "top level" checksums file
# Prevents pattern matching downstream
rm checksums.md5

# Make a line
line="-----------------------------------------------------------------------------------------------"

# Set working directory
wd=$(pwd)

# Loop through directories and verify checksums
for directory in */
do
  cd "${directory}"
  # Get sample name; strips trailing slash from directory name
  sample="${directory%/}"
  
  echo ${line}
  echo "${sample}"
  echo ""
  
  # Confirm checksums; sorts for easier reading
  md5sum --check "${sample}"_checksums.md5 | sort -V
  echo ""
  echo "${line}"
  echo ""
  cd ${wd}
done
# Show downloaded directories/files
tree
```

# Find Ballgown installation location
```{r}
data_directory <-  system.file('extdata', package='ballgown') # automatically finds ballgown's installation directory
# examine data_directory:
data_directory
```

# Create Ballgown object
```{r}
# Uses regular expression in samplePattern to find all pertinent folders
# Load all measurement data
bg_all <- ballgown(dataDir="../data/ballgown/", samplePattern='[CE]*', meas='all')
bg_all
bg_controls <- ballgown(dataDir="../data/ballgown/", samplePattern='C.*', meas='all')
bg_controls
bg_enriched <- ballgown(dataDir="../data/ballgown/", samplePattern='E.*', meas='all')
bg_enriched
```


# Download and filter metadata file

Filtered metadata will be used to create a phenotype dataframe needed for Ballgown differential expression analysis.

```{r create-dataframes-for-ballgown-pData}
# Read in metadata file
sample_metadata_full <- read.csv("../data/metadata.RNAseq.csv")

# View full metadata
head(sample_metadata_full)

# Organize metadata in preparation of creating pData data frame far Ballgown
# Sort by "sample_id" to ensure matches directory structure (required for Ballgown)
sample_metadata_full <- sample_metadata_full %>% 
  select(fragment.ID, treatment, block, sample_id) %>% 
  arrange(sample_id)
# View organized metadata
head(sample_metadata_full)
```

# Load ALL phenotype dataframe into Ballgown object
```{r}
# Load phenotype info into Ballgown
pData(bg_all) <- sample_metadata_full
# Examine phenotype data as it exists in Ballgown
phenotype_table <-  pData(bg_all)
head(phenotype_table)
```


### Look at ALL exon info
```{r}
# Exon info
structure(bg_all)$exon
```

### Look at ALL intron info
```{r}
# Intron info
structure(bg_all)$intron
```

###  Look at ALL transcript (isoform) info
```{r}
# Transcript info
structure(bg_all)$trans
```

# Load ALL transcript expression data
```{r load-all-transcript-expression-data}
# Expression data
whole_tx_table <-  texpr(bg_all, 'all')

# Rename gene_names listed as a "."
whole_tx_table <- whole_tx_table %>% mutate(gene_name = ifelse(gene_name == ".", t_name, gene_name))
head(whole_tx_table)

# FPKM values for all transcripts
# Converts output to data frame
transcript_fpkm <- as.data.frame(texpr(bg_all, 'FPKM'))

# Put rownames into column for further manipulation
transcript_fpkm <- rownames_to_column(transcript_fpkm, "t_id")
head(transcript_fpkm)
```

## Write all transcript data to files
```{r write-transcript-data-to-file}
# Write all transcript FPKM data to file
write.csv(transcript_fpkm,
          file = "../output/02-ballgown-differential_gene_expression/fpkm-all_transcripts.csv",
          quote = FALSE,
          row.names = FALSE)

# Write whole_tx_table to file
write.csv(whole_tx_table,
          file ="../output/02-ballgown-differential_gene_expression/whole_tx_table.csv",
          quote = FALSE,
          row.names = FALSE)
```
