---
title: "Nov17-update"
format: revealjs
editor: visual
---


```{r setup, include=FALSE}
library(knitr)
library(kableExtra)
library(dplyr)
knitr::opts_chunk$set(
  echo = TRUE,         # Display code chunks
  eval = FALSE,        # Evaluate code chunks
  warning = FALSE,     # Hide warnings
  message = FALSE,     # Hide messages
  comment = ""          # Prevents appending '##' to beginning of lines in code output
)

```

## Quarto

Quarto enables you to weave together content and executable code into a finished presentation. To learn more about Quarto presentations see <https://quarto.org/docs/presentations/>.

## Bullets

When you click the **Render** button a document will be generated that includes:

-   Content authored with markdown
-   Output from executable code

## Code

When you click the **Render** button a presentation will be generated that includes both content and the output of embedded code. You can embed code like this:

```{r}
1 + 1
```

## lncRNA Discovery

- Aligned RNA-seq reads to genome -> SAM files
- SAM files converted to BAMS
- BAMS converted to GTF (Stringtie), merged to single GTF
- ID non-coding transcripts (GFFCompare, CPC2)
- Fasta and Bed file created for each species

## sRNAseq Analysis

- Trim reads to 25bp length
- miRTrace
  - Identify taxonomic origins (Clade-level) of sRNAseq data
- MirMachine
  - Identify potential miRNA homologs in genome (no sRNAseq analyses)
- ShortStack alignment of sRNAseq data and annotation of sRNA-producing genes.
  - Uses sRNAseq, genome, and miRNA database (miRBase).
- BLASTn
  - Align sRNAseq to miRNA databases (miRBase, MirGene)
- miRDeep2
  - Use sRNAseq, genome, and miRNA database (miRBase)


### MiRTrace

#### _A.pulchra_

miRNA Families: 5

Clades: 3
    - lophotrochozoa
    - rodents
    - primates

Number of samples with matches: 3


  