---
title: "Nov17-update"
format: revealjs
editor: visual
toc: true
toc_float: true
code_folding: show
---


## Non-coding RNA 

```{r setup, include=FALSE}
library(knitr)
library(kableExtra)
library(dplyr)
knitr::opts_chunk$set(
  echo = TRUE,         # Display code chunks
  eval = FALSE,        # Evaluate code chunks
  warning = FALSE,     # Hide warnings
  message = FALSE,     # Hide messages
  comment = ""          # Prevents appending '##' to beginning of lines in code output
)

```



Where things stand

## lncRNA Discovery


- Aligned RNA-seq reads to genome -> SAM files
- SAM files converted to BAMS
- BAMS converted to GTF (Stringtie), merged to single GTF
- ID non-coding transcripts (GFFCompare, CPC2)
- Fasta and Bed file created for each species


## lncRNA by the numbers

- Apul total transcripts / lncRNA = 106174 / 16206
- Peve total transcripts / lncRNA =     / 7378
- Pmea total transcripts / lncRNA = 31840 / 

## sRNAseq Analysis {.scrollable}

-   Trim reads to 25bp length
-   miRTrace
    -   Identify taxonomic origins (Clade-level) of sRNAseq data
-   MirMachine
    -   Identify potential miRNA homologs in genome (no sRNAseq analyses)
-   ShortStack alignment of sRNAseq data and annotation of sRNA-producing genes.
    -   Uses sRNAseq, genome, and miRNA database (miRBase).
-   BLASTn
    -   Align sRNAseq to miRNA databases (miRBase, MirGene)
-   miRDeep2
    -   Use sRNAseq, genome, and miRNA database (miRBase)

## MiRTrace

#### *A.pulchra*

```{r read-apul-mirtrace-table, eval=TRUE, echo=FALSE}
mirtrace.detailed.df <- read.csv("../../D-Apul/output/09-Apul-sRNAseq-miRTrace/mirtrace-stats-contamination_detailed.tsv", sep = "\t", header = TRUE)
```

```{r mirtrace-output-table, eval=TRUE, echo=FALSE}

mirtrace.detailed.df %>%
  mutate(
    across(
      starts_with("sRNA"),
      ~cell_spec(
        .,
        background = ifelse(
          . > 0,
          "lightgreen",
          "white"
          )
        )
      )
    ) %>%
  kable(escape = F, caption = "Clades identified as having sRNAseq matches.") %>%
  kable_styling("striped") %>% 
  scroll_box(width = "100%", height = "500px")
```

## MirMachine

### _A.millepora_

```{r count-predicted-hc-mirnas}
#| eval: true
#| echo: false

# Store Bash command to use with system()
cmd_predicted_loci <- paste(c('grep -c "^[^#]" "../../D-Apul/output/12-Apul-sRNAseq-MirMachine/results/predictions/filtered_gff/Amil-MirMachine.PRE.gff"'))


# Store Bash command to use with system()
cmd_unique_fams <- paste(c('grep "^[^#]" "../../D-Apul/output/12-Apul-sRNAseq-MirMachine/results/predictions/filtered_gff/Amil-MirMachine.PRE.gff"',
'awk -F"[\t=;]" \'{print $10}\'',
'sort -u',
'wc -l'),
collapse=" | ")

predicted_loci <- system(cmd_predicted_loci, intern = TRUE)
unique_fams <- system(cmd_unique_fams, intern = TRUE)
```

Predicted loci: `r predicted_loci`

Unique familes: `r unique_fams`

