# deep-dive

This repository is focused on takeing molecular data from three experiments where integrated epigenetic analysis has already been performed, and doing a _deeper dive_ into datasets to extract and analyze remaining points of excitement. 


In brief some extra special data sets that will be generated

### Epigenetic Data Types
- CpG methylation
- long non-coding RNA
- Gene Methylation Modules

### Gene Activity
- Transcript expression
- Gene Coeff Var
- Transcript Coeff Variation
- Predominat isoform
- Gene Expression Modules

### Other



Short-hand and other repos for the three experiments are

A) https://github.com/hputnam/Becker_E5 (Pocillopora verrucosa genome)

- https://gannet.fish.washington.edu/Atumefaciens/hputnam-Becker_E5/
- [OSF](https://osf.io/uayvk/)
- Nutrient exposure 


B) https://github.com/hputnam/HI_Bleaching_Timeseries/tree/main/Dec-July-2019-analysis (Montipora capitata genome)

C) https://github.com/emmastrand/Acclim_Dynamics_molecular (Pocillopora acuta genome)

- This effort is somewhat comprimised as ploidy (unintended), overwhelm signals

D) Acropora



---

## Genomes

- *Pocillopora verrucosa* genome v1.0 : http://pver.reefgenomics.org : https://www.ncbi.nlm.nih.gov/data-hub/genome/GCA_014529365.1/
- *Montipora capitata* v3 - http://cyanophora.rutgers.edu/montipora/ : https://www.ncbi.nlm.nih.gov/data-hub/genome/GCA_006542545.1/
- *Pocillopora acuta* v2 - http://cyanophora.rutgers.edu/Pocillopora_acuta/ : 
