---
title: "09-Apul-sRNAseq-miRTrace"
author: "Sam White"
date: "2023-11-01"
output: 
  bookdown::html_document2:
    theme: cosmo
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
    code_download: true
bibliography: references.bib
---

```{r setup, include=FALSE}
library(knitr)
library(kableExtra)
library(dplyr)
knitr::opts_chunk$set(
  echo = TRUE,         # Display code chunks
  eval = FALSE,        # Evaluate code chunks
  warning = FALSE,     # Hide warnings
  message = FALSE,     # Hide messages
  comment = ""         # Prevents appending '##' to beginning of lines in code output
)

```

Use [miRTrace](https://github.com/friedlanderlab/mirtrace) [@kang2018] to identify taxonomic origins of miRNA sequencing data.

NOTE: This requires you to have previously run [`08-Apul-sRNAseq-trimming.Rmd`](https://github.com/urol-e5/deep-dive/blob/main/D-Apul/code/08-Apul-sRNAseq-trimming.Rmd), as the code relies on the trimmed reads output from that code.

---


Inputs:

- Trimmed sRNAseq FastQs generated by [`08-Apul-sRNAseq-trimming.Rmd`](https://github.com/urol-e5/deep-dive/blob/main/D-Apul/code/08-Apul-sRNAseq-trimming.Rmd)

  - Filenames formatted: `*flexbar_trim.25bp*.gz`
  
Outputs:

- `mirtrace.config`: A [miRTrace](https://github.com/friedlanderlab/mirtrace) config file. A comma-separated file with this layout (one FastQ per line): `/path/to/fastq,custom_sample_name`

- "Collapsed" (i.e. unique sequences only) FastA for each corresponding input FastQ.

- `mirtrace-report.html`: HTML-formatted report generated by [miRTrace](https://github.com/friedlanderlab/mirtrace).

- `mirtrace-stats-contamination_basic.tsv`: Tab-delimited report with counts of sequences from each collapsed FastAs having matches to known miRNAs within each of the [miRTrace](https://github.com/friedlanderlab/mirtrace) Clades.

- `mirtrace-stats-contamination_detailed.csv`: Tab-delimited report of _only_ Clades with which sequences were matched, along with the corresponding miRNA families in each clade, and the sequence counts.


# Create a Bash variables file

This allows usage of Bash variables across R Markdown chunks.

```{r save-bash-variables-to-rvars-file, engine='bash', eval=TRUE}
{
echo "#### Assign Variables ####"
echo ""

echo "# Data directories"
echo 'export deep_dive_dir=/home/shared/8TB_HDD_01/sam/gitrepos/deep-dive'
echo 'export output_dir_top=${deep_dive_dir}/D-Apul/output/09-Apul-sRNAseq-miRTrace'
echo 'export trimmed_reads_dir=${deep_dive_dir}/D-Apul/output/08-Apul-sRNAseq-trimming/trimmed-reads'
echo ""

echo "# Paths to programs"
echo 'export mirtrace=/home/sam/programs/mambaforge/envs/miRTrace_env/bin/mirtrace'
echo ""

echo "# Set number of CPUs to use"
echo 'export threads=40'
echo ""

echo "export fastq_pattern='*flexbar_trim.25bp*.gz'"

echo "# Programs associative array"
echo "declare -A programs_array"
echo "programs_array=("
echo '[mirtrace]="${mirtrace}"'
echo ")"
} > .bashvars

cat .bashvars
```

# Create [miRTrace](https://github.com/friedlanderlab/mirtrace) config file
```{r create-config-file, engine='bash', eval=TRUE}
# Load bash variables into memory
source .bashvars

# Declare array
fastq_array=()

# Populate array
fastq_array=(${trimmed_reads_dir}/${fastq_pattern})

# Loop through read pairs
# Increment by 2 to process next pair of FastQ files
if [ -f "${output_dir_top}/mirtrace.config" ]; then
  echo "mirtrace.config already exists. Nothing to do."
  
else

  for (( i=0; i<${#fastq_array[@]} ; i+=2 ))
  do
    # Use first three parts of filename to create short sample name
    R1_name=$(echo "${fastq_array[i]##*/}" | awk -F "-" '{print $1"-"$2"-"$3}')
    R2_name=$(echo "${fastq_array[i+1]##*/}" | awk -F "-" '{print $1"-"$2"-"$3}')
    echo "${fastq_array[i]},${R1_name}_1"
    echo "${fastq_array[i+1]},${R2_name}_2"
  done >> "${output_dir_top}/mirtrace.config"

fi

cat "${output_dir_top}/mirtrace.config"
```

# Run [miRTrace](https://github.com/friedlanderlab/mirtrace)
```{r run-mirtrace, engine='bash', eval=TRUE}
# Load bash variables into memory
source .bashvars

time \
${programs_array[mirtrace]} trace \
--config ${output_dir_top}/mirtrace.config \
--write-fasta \
--num-threads ${threads} \
--output-dir ${output_dir_top} \
--force

tree -h ${output_dir_top}
```


# Results

## Read in table as data frame
```{r read-table, eval=TRUE}
mirtrace.detailed.df <- read.csv("../output/09-Apul-sRNAseq-miRTrace/mirtrace-stats-contamination_detailed.tsv", sep = "\t", header = TRUE)

str(mirtrace.detailed.df)
```


## Number of samples with matches
```{r counts-samples-with-matches, eval=TRUE}
# Select columns corresponding to sample names
sample_columns <- mirtrace.detailed.df %>%
  select(starts_with("sRNA.ACR."))

# Calculate the sum for each column
sample_sums <- colSums(sample_columns)

# Count the number of columns with a sum greater than 0
samples_with_sum_gt_0 <- sum(sample_sums > 0)

paste("Number of samples with matches: ", samples_with_sum_gt_0)
```
## Percentage of samples with matches
```{r percentage-samples-with-matches, eval=TRUE}
# Total number of samples (columns)
total_samples <- ncol(sample_columns)

# Percentage of samples with sums greater than 0
percentage_samples_gt_0 <- (samples_with_sum_gt_0 / total_samples) * 100

paste("Percentage of samples with matches: ", percentage_samples_gt_0)
```


## Number of clades with matches
```{r distinct-clades, eval=TRUE, class.source = 'fold-hide'}

unique_clade_count <- mirtrace.detailed.df %>%
  distinct(CLADE) %>%    # Get unique entries in CLADE column
  count()               # Count the number of unique entries



paste("Number of clades with matches:", unique_clade_count)


```


## [miRTrace](https://github.com/friedlanderlab/mirtrace) table

To make them easier to see, counts > 0 are highlighted in green.
```{r mirtrace-output-table, eval=TRUE, class.source = 'fold-hide'}

mirtrace.detailed.df %>%
  mutate(
    across(
      starts_with("sRNA"),
      ~cell_spec(
        .,
        background = ifelse(
          . > 0,
          "lightgreen",
          "white"
          )
        )
      )
    ) %>%
  kable(escape = F, caption = "Clades identified as having sRNAseq matches.") %>%
  kable_styling("striped") %>% 
  scroll_box(width = "100%", height = "500px")
```

# Citations


