---
title: "13-Apul-sRNAseq-ShortStack"
author: "Sam White"
date: "2023-11-03"
output: 
  bookdown::html_document2:
    theme: cosmo
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
    code_download: true
  github_document:
    theme: cosmo
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
    code_download: true
bibliography: references.bib
link-citations: true
---

```{r setup, include=FALSE}
library(knitr)
library(kableExtra)
library(dplyr)
library(reticulate)
knitr::opts_chunk$set(
  echo = TRUE,         # Display code chunks
  eval = FALSE,        # Evaluate code chunks
  warning = FALSE,     # Hide warnings
  message = FALSE,     # Hide messages
  comment = ""         # Prevents appending '##' to beginning of lines in code output
)
```

This notebook performs a simple [NCBI BLASTn](https://www.ncbi.nlm.nih.gov/books/NBK279690/) [@altschul1990] against two miRNA databases to attempt to identify miRNA in _A.pulchra_ sRNAseq:

- [miRBase](https://mirbase.org/download/)

- [MirGeneDB](https://www.mirgenedb.org/download)

Relies on the following software:

- [miRDeep-P2_pipeline](https://github.com/TF-Chan-Lab/miRDeep-P2_pipeline)

  - `fasta_U2T.pl`: Converts `U` to `T` in miRNA databases.
  
- [fastx_toolkit](http://hannonlab.cshl.edu/fastx_toolkit/)

  - `fastx_collapser`: Collapses duplicate sequences in FastA/Q into single sequence.

# Create a Bash variables file

This allows usage of Bash variables across R Markdown chunks.

```{r save-bash-variables-to-rvars-file, engine='bash', eval=TRUE}
{
echo "#### Assign Variables ####"
echo ""

echo "# Trimmed FastQ naming pattern"
echo "export trimmed_fastqs_pattern='*flexbar_trim.25bp*.fastq.gz'"

echo "# Data directories"
echo 'export deep_dive_dir=/home/shared/8TB_HDD_01/sam/gitrepos/deep-dive'
echo 'export deep_dive_data_dir="${deep_dive_dir}/data"'
echo 'export output_dir_top=${deep_dive_dir}/D-Apul/output/10-Apul-sRNAseq-BLASTn'
echo 'export trimmed_fastqs_dir="${deep_dive_dir}/D-Apul/output/08-Apul-sRNAseq-trimming/trimmed-reads"'
echo ""

echo "# Input/Output files"
echo 'export mirbase_mature_fasta_name="mirbase-mature-v22.1.fa"'
echo 'export mirgene_mature_fasta_name="mirgene-mature-all-v2.1.fa"'
echo ""

echo "# External data URLs"
echo 'export mirgenedb_fasta_url="https://www.mirgenedb.org/fasta/ALL?mat=1"'
echo ""

echo "# Set number of CPUs to use"
echo 'export threads=46'
echo ""

echo "# Initialize arrays"
echo 'export trimmed_fastqs_array=()'


} > .bashvars

cat .bashvars
```


# Download MirGeneDB Fasta
```{r download-mirgene-db-FastA, engine='bash', eval=TRUE}
# Load bash variables into memory
source .bashvars

wget \
--no-check-certificate \
--continue \
--no-host-directories \
--no-directories \
--no-parent \
--quiet \
--execute robots=off \
--output-document ${deep_dive_data_dir}/${mirgene_mature_fasta_name} \
 ${mirgenedb_fasta_url}
 
 ls -lh ${deep_dive_data_dir}
```
## Inspect miRNA FastAs
```{r inspect-miRNA-FastAs, engine='bash', eval=TRUE}
# Load bash variables into memory
source .bashvars

head "${deep_dive_data_dir}"/mir*.fa
```





------------------------------------------------------------------------

# Citations
