---
title: "11-Apul-sRNAseq-miRdeep2"
author: "Sam White"
date: "2023-11-08"
output: 
  bookdown::html_document2:
    theme: cosmo
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
    code_download: true
  github_document:
    toc: true
    number_sections: true
  html_document:
    theme: cosmo
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
    code_download: true
bibliography: references.bib
link-citations: true
---

```{r setup, include=FALSE}
library(knitr)
library(kableExtra)
library(dplyr)
library(reticulate)
knitr::opts_chunk$set(
  echo = TRUE,         # Display code chunks
  eval = FALSE,        # Evaluate code chunks
  warning = FALSE,     # Hide warnings
  message = FALSE,     # Hide messages
  comment = "",        # Prevents appending '##' to beginning of lines in code output
  width = 1000         # adds scroll bar
)
```




# Create a Bash variables file

This allows usage of Bash variables across R Markdown chunks.

```{r save-bash-variables-to-rvars-file, engine='bash', eval=TRUE}
{
echo "#### Assign Variables ####"
echo ""

echo "# Trimmed FastQ naming pattern"
echo "export trimmed_fastqs_pattern='*flexbar_trim.25bp*.fastq.gz'"
echo ""

echo "# miRTrace FastA naming pattern"
echo "export mirtrace_fasta_pattern='*flexbar_trim.25bp*.fasta'"

echo "# Data directories"
echo 'export deep_dive_dir=/home/shared/8TB_HDD_01/sam/gitrepos/deep-dive'
echo 'export deep_dive_data_dir="${deep_dive_dir}/data"'
echo 'export output_dir_top=${deep_dive_dir}/D-Apul/output/11-Apul-sRNAseq-miRdeep2'
echo 'export genome_fasta_dir=${deep_dive_dir}/D-Apul/data/Amil/ncbi_dataset/data/GCF_013753865.1'
echo 'export trimmed_fastqs_dir="${deep_dive_dir}/D-Apul/output/08-Apul-sRNAseq-trimming/trimmed-reads"'
echo 'export collapsed_reads_dir="${deep_dive_dir}/D-Apul/output/10-Apul-sRNAseq-BLASTn"'
echo 'export mirtrace_reads_dir="${deep_dive_dir}/D-Apul/output/09-Apul-sRNAseq-miRTrace/qc_passed_reads.all.collapsed"'
echo ""

echo "# Input/Output files"
echo 'export collapsed_reads_fasta="collapsed-reads-all.fasta"'
echo 'export collapsed_reads_mirdeep2="collapsed-reads-all-mirdeep2.fasta"'
echo 'export concatenated_trimmed_reads_fastq="concatenated-trimmed-reads-all.fastq.gz"'
echo 'export genome_fasta_name="GCF_013753865.1_Amil_v2.1_genomic.fna"'
echo 'export genome_fasta_no_spaces="GCF_013753865.1_Amil_v2.1_genomic-no_spaces.fna"'
echo 'export mirdeep2_mapping_file="Apul-mirdeep2-mapping.arf"'
echo 'export mirbase_mature_fasta_name="mirbase-mature-v22.1.fa"'
echo 'export mirbase_mature_fasta_no_spaces="mirbase-mature-v22.1-no_spaces.fa"'
echo 'export concatenated_mirtrace_reads_fasta="concatenated-mirtrace-reads.fasta"'
echo ""


echo "# Paths to programs"
echo 'export mirdeep2_mapper="mapper.pl"'
echo 'export mirdeep2="miRDeep2.pl"'
echo 'export bowtie_build="/home/shared/bowtie-1.3.1-linux-x86_64/bowtie-build"'
echo ""

echo "# Set number of CPUs to use"
echo 'export threads=46'
echo ""

echo "# Initialize arrays"
echo 'export trimmed_fastqs_array=()'


} > .bashvars

cat .bashvars
```

# Prepare reads for miRDeep2

Per miRDeep2 documentation: 

> The readID must end with _xNumber and is not allowed to contain whitespaces.
> has to have the format
> name_uniqueNumber_xnumber

```{r reformat-read-IDs, eval=TRUE, engine='bash'}
# Load bash variables into memory
source .bashvars


sed '/^>/ s/-/_x/g' "${collapsed_reads_dir}/${collapsed_reads_fasta}" \
| sed '/^>/ s/>/>seq_/' \
> "${output_dir_top}/${collapsed_reads_mirdeep2}"

grep "^>" ${collapsed_reads_dir}/${collapsed_reads_fasta} \
| head

echo ""
echo "--------------------------------------------------"
echo ""

grep "^>" ${output_dir_top}/${collapsed_reads_mirdeep2} \
| head
```

# Reformat genome FastA description lines

miRDeep2 can't process genome FastAs with spaces in
the description lines.

So, I'm replacing spaces with underscores.

And, for aesthetics, I'm also removing commas.

```{r remove-spaces-genome-FastA, eval=TRUE, engine='bash'}
# Load bash variables into memory
source .bashvars


sed '/^>/ s/ /_/g' "${genome_fasta_dir}/${genome_fasta_name}" \
| sed '/^>/ s/,//g' \
> "${genome_fasta_dir}/${genome_fasta_no_spaces}"

grep "^>" ${genome_fasta_dir}/${genome_fasta_name} \
| head

echo ""
echo "--------------------------------------------------"
echo ""

grep "^>" ${genome_fasta_dir}/${genome_fasta_no_spaces} \
| head
```

# Reformat miRBase FastA description lines
```{r remove-spaces-genome-FastA, eval=TRUE, engine='bash'}
# Load bash variables into memory
source .bashvars


sed '/^>/ s/ /_/g' "${deep_dive_data_dir}/${mirbase_mature_fasta_name}" \
| sed '/^>/ s/,//g' \
> "${deep_dive_data_dir}/${mirbase_mature_fasta_no_spaces}"

grep "^>" ${deep_dive_data_dir}/${mirbase_mature_fasta_name} \
| head

echo ""
echo "--------------------------------------------------"
echo ""

grep "^>" ${deep_dive_data_dir}/${mirbase_mature_fasta_no_spaces} \
| head
```



# Bowtie v1 genome index

miRDeep2 requires a Bowtie v1 genome index - cannot use Bowtie2 index
```{r bowtie1-index, eval=TRUE, engine='bash'}
# Load bash variables into memory
source .bashvars

# Check for existence of genome index first
if [ ! -f "${genome_fasta_no_spaces%.*}.ebwt" ]; then
  ${bowtie_build} \
  ${genome_fasta_dir}/${genome_fasta_no_spaces} \
  ${genome_fasta_dir}/${genome_fasta_no_spaces%.*} \
  --threads ${threads} \
  --quiet
fi
```

# Map reads to genome

Requires genome to be previously indexed with Bowtie.
```{r map-read-genome, eval=TRUE, engine='bash'}
# Load bash variables into memory
source .bashvars

# Append miRDeep2 to system PATH and set PERL5LIB
export PATH=$PATH:/home/shared/mirdeep2/bin
export PERL5LIB=$PERL5LIB:/home/shared/mirdeep2/lib/perl5

# Run miRDeep2 mapping
time \
${mirdeep2_mapper} \
${output_dir_top}/${collapsed_reads_mirdeep2} \
-c \
-p ${genome_fasta_dir}/${genome_fasta_no_spaces%.*} \
-t ${output_dir_top}/${mirdeep2_mapping_file} \
-o ${threads}
```

# Run miRDeep2

```{r mirdeep2, eval=FALSE, engine='bash'}
# Load bash variables into memory
source .bashvars

# Append miRDeep2 to system PATH and set PERL5LIB
export PATH=$PATH:/home/shared/mirdeep2/bin
export PERL5LIB=$PERL5LIB:/home/shared/mirdeep2/lib/perl5

time \
${mirdeep2} \
${output_dir_top}/${collapsed_reads_mirdeep2} \
${genome_fasta_dir}/${genome_fasta_no_spaces} \
${output_dir_top}/${mirdeep2_mapping_file} \
none \
${deep_dive_data_dir}/${mirbase_mature_fasta_no_spaces} \
none \
-t S.purpuratus \
-P \
-v \
-g -1 \
2>${output_dir_top}/miRDeep2-S.purpuratus-report.log
```

## Check runtime
```{r runtime-mirdeep2, eval=TRUE, engine='bash'}
# Load bash variables into memory
source .bashvars

tail -n 6 ${output_dir_top}/miRDeep2-S.purpuratus-report.log
```


# Move output files to output directory
MiRDeep2 outputs all files to the current working directly with no way to redirect
so want to move to intended output directory.

## Move output files

Output files will be in the format of `result_*` and `error_*`
```{r identify-mirdeep2-output-files, eval=FALSE, engine='bash'}
# Load bash variables into memory
source .bashvars

for file in result_* error_*
do
  mv "${file}" "${output_dir_top}/"
done
```



## Identify directories

```{r identify-mirdeep2-output-dirs, eval=TRUE, engine='bash'}
# Load bash variables into memory
source .bashvars

ls -l | grep "^d"
```

## Rsync directories

```{r move-mirdeep2-output-files, eval=TRUE, engine='bash'}
# Load bash variables into memory
source .bashvars

rsync -aP . --include='dir***' --exclude='*' --quiet "${output_dir_top}/"

rsync -aP . --include='map***' --exclude='*' --quiet "${output_dir_top}/"

rsync -aP . --exclude='mirgene*' --include='mir***' --exclude='*' --quiet "${output_dir_top}/"

rsync -aP . --include='pdfs***' --exclude='*' --quiet "${output_dir_top}/"

echo ""
echo "Check new location:"
echo ""

ls -l "${output_dir_top}/" | grep "^d"


```

## Confirm deletion patterns work _before_ deletion!
FYI - `eval=FALSE` is set because the following command will only work once...
```{r test-deletion-patterns, eval=FALSE, engine='bash'}
ls --directory dir_* mapper_logs mirdeep_runs mirna_results* pdfs_*
```


## Remove directores from code directory

```{r remove-mirdeep2-output-dirs, eval=FALSE, engine='bash'}
# Load bash variables into memory
source .bashvars

rm -rf dir_* mapper_logs mirdeep_runs mirna_results* pdfs_*

```

## Check to make sure they're gone
```{r check-deletion-success, eval=TRUE, engine='bash'}
ls --directory dir_* mapper_logs mirdeep_runs mirna_results* pdfs_*
```

# Results

## Peep output file format

The formatting of this CSV is terrible. It has a 3-column table on top of a
17-column table. This makes parsing a bit of a pain in its raw format.
```{r check-results-file, eval=TRUE, engine='bash'}
# Load bash variables into memory
source .bashvars


head -n 30 "${output_dir_top}/result_08_11_2023_t_14_47_36.csv"
```

## Create more easily parasable results file
This will match the line beginning with `provisional id` and print to the
end of the file (represented by the `$p`. `$` = end, `p` = print)
```{r parsable-results-file, eval=TRUE, engine='bash'}
# Load bash variables into memory
source .bashvars


sed --quiet '/provisional id/,$p' "${output_dir_top}/result_08_11_2023_t_14_47_36.csv" \
> "${output_dir_top}/parsable-result_08_11_2023_t_14_47_36.csv"

head "${output_dir_top}/parsable-result_08_11_2023_t_14_47_36.csv"
```



### Read in output CSV
```{r read-in-result-csv, eval=TRUE}

mirdeep_result.df <- read.csv("../output/11-Apul-sRNAseq-miRdeep2/parsable-result_08_11_2023_t_14_47_36.csv",
                              header = TRUE,
                              sep = "\t")

str(mirdeep_result.df)

```



---

# Citations