11.2-Apul-sRNAseq-miRdeep2-31bp-fastp-merged-cnidarian_only
================
Sam White
2024-05-16

- <a href="#1-create-a-bash-variables-file"
  id="toc-1-create-a-bash-variables-file">1 Create a Bash variables
  file</a>
- <a href="#2-prepare-reads-for-mirdeep2"
  id="toc-2-prepare-reads-for-mirdeep2">2 Prepare reads for miRDeep2</a>
  - <a href="#21-convert-fastq-to-fasta"
    id="toc-21-convert-fastq-to-fasta">2.1 Convert FastQ to FastA</a>
  - <a href="#22-remove-spaces-in-reads-fasta"
    id="toc-22-remove-spaces-in-reads-fasta">2.2 Remove Spaces in Reads
    FastA</a>
- <a href="#3-reformat-genome-fasta-description-lines"
  id="toc-3-reformat-genome-fasta-description-lines">3 Reformat genome
  FastA description lines</a>
- <a href="#4-bowtie-v1-genome-index" id="toc-4-bowtie-v1-genome-index">4
  Bowtie v1 genome index</a>
- <a href="#5-map-reads-to-genome" id="toc-5-map-reads-to-genome">5 Map
  reads to genome</a>
- <a href="#6-run-mirdeep2" id="toc-6-run-mirdeep2">6 Run miRDeep2</a>
  - <a href="#61-check-runtime" id="toc-61-check-runtime">6.1 Check
    runtime</a>
- <a href="#7-results" id="toc-7-results">7 Results</a>
  - <a href="#71-peep-output-file-format"
    id="toc-71-peep-output-file-format">7.1 Peep output file format</a>
  - <a href="#72-create-more-easily-parasable-results-file"
    id="toc-72-create-more-easily-parasable-results-file">7.2 Create more
    easily parasable results file</a>
  - <a href="#73-read-in-output-csv" id="toc-73-read-in-output-csv">7.3 Read
    in output CSV</a>
  - <a href="#74-mirnas-count-data" id="toc-74-mirnas-count-data">7.4 miRNAs
    count data</a>
- <a href="#8-citations" id="toc-8-citations">8 Citations</a>

Use [miRDeep2](https://github.com/rajewsky-lab/mirdeep2) ([Friedländer
et al. 2011](#ref-friedländer2011)) to identify potential miRNAs using
*A.pulchra* sRNAseq reads. The *A.millepora* genome will be used as the
reference genome for *A.pulchra*, as *A.pulchra* does not currently have
a sequenced genome and *A.millepora* had highest alignment rates for
standard RNAseq data compared to other published genomes tested.

------------------------------------------------------------------------

Inputs:

- Trimmed and merged reads. See
  [08.2-Apul-sRNAseq-trimming-31bp-fastp-merged.Rmd](./08.2-Apul-sRNAseq-trimming-31bp-fastp-merged.Rmd)
  for code.

- *A.millepora* genome FastA. See
  [12-Apul-sRNAseq-MirMachine.Rmd](https://github.com/urol-e5/deep-dive/blob/main/D-Apul/code/12-Apul-sRNAseq-MirMachine.Rmd)
  for download info if needed.

- [Cnidarian-only miRNA FastA](../../data/cnidarian_miRNAs.fasta)

  - Cnidarian miRNA culled from literature by Jill Ashey.

Outputs:

- Primary outputs are a result table in BED, CSV (tab-delimited), and
  HTML formats.

- Due to the nature of mirDeep2’s naming, trying to use variable names
  is challenging. As such, the chunks processing those files will
  require manual intervention to identify and provide the output
  filename(s); they are *not* handled at in the `.bashvars` file at the
  top of this script.

- Other output files are too large for GitHub (and some (all?) are not
  needed for this analysis). Please find a full backup here:

<https://gannet.fish.washington.edu/Atumefaciens/gitrepos/deep-dive/D-Apul/output/11.2-Apul-sRNAseq-miRdeep2-31bp-fastp-merged-cnidarian_only/>

------------------------------------------------------------------------

# 1 Create a Bash variables file

This allows usage of Bash variables across R Markdown chunks.

``` bash
{
echo "#### Assign Variables ####"
echo ""

echo "# Trimmed FastQ naming pattern"
echo "export trimmed_fastqs_pattern='*fastp-adapters-polyG-31bp-merged.fq.gz'"
echo ""

echo "# Data directories"
echo 'export deep_dive_dir=/home/shared/8TB_HDD_01/sam/gitrepos/deep-dive'
echo 'export deep_dive_data_dir="${deep_dive_dir}/data"'
echo 'export output_dir_top=${deep_dive_dir}/D-Apul/output/11.2-Apul-sRNAseq-miRdeep2-31bp-fastp-merged-cnidarian_only'
echo 'export genome_fasta_dir=${deep_dive_dir}/D-Apul/data/Amil/ncbi_dataset/data/GCF_013753865.1'
echo 'export trimmed_fastqs_dir="${deep_dive_dir}/D-Apul/output/08.2-Apul-sRNAseq-trimming-31bp-fastp-merged/trimmed-reads"'
echo ""

echo "# Input/Output files"
echo 'export collapsed_reads_gt17bp_mirdeep2="collapsed_reads_gt17bp_mirdeep2.fasta"'
echo 'export concatenated_trimmed_reads_fasta="concatenated-trimmed-reads-all.fasta"'
echo 'export concatenated_trimmed_reads_fasta_no_spaces="concatenated-trimmed-reads-all-no_spaces.fasta"'
echo 'export genome_fasta_name="GCF_013753865.1_Amil_v2.1_genomic.fna"'
echo 'export genome_fasta_no_spaces="GCF_013753865.1_Amil_v2.1_genomic-no_spaces.fna"'
echo 'export mirdeep2_mapping_file="Apul-mirdeep2-mapping.arf"'
echo 'export cnidarian_mirna_fasta_name="cnidarian_miRNAs.fasta"'
echo ""


echo "# Paths to programs"
echo 'export mirdeep2_mapper="mapper.pl"'
echo 'export mirdeep2="miRDeep2.pl"'
echo 'export mirdeep2_fastaparse="fastaparse.pl"'
echo 'export bowtie_build="/home/shared/bowtie-1.3.1-linux-x86_64/bowtie-build"'
echo 'export seqtk="/home/shared/seqtk-1.4/seqtk"'
echo ""

echo "# Set number of CPUs to use"
echo 'export threads=40'
echo ""



} > .bashvars

cat .bashvars
```

    #### Assign Variables ####

    # Trimmed FastQ naming pattern
    export trimmed_fastqs_pattern='*fastp-adapters-polyG-31bp-merged.fq.gz'

    # Data directories
    export deep_dive_dir=/home/shared/8TB_HDD_01/sam/gitrepos/deep-dive
    export deep_dive_data_dir="${deep_dive_dir}/data"
    export output_dir_top=${deep_dive_dir}/D-Apul/output/11.2-Apul-sRNAseq-miRdeep2-31bp-fastp-merged-cnidarian_only
    export genome_fasta_dir=${deep_dive_dir}/D-Apul/data/Amil/ncbi_dataset/data/GCF_013753865.1
    export trimmed_fastqs_dir="${deep_dive_dir}/D-Apul/output/08.2-Apul-sRNAseq-trimming-31bp-fastp-merged/trimmed-reads"

    # Input/Output files
    export collapsed_reads_gt17bp_mirdeep2="collapsed_reads_gt17bp_mirdeep2.fasta"
    export concatenated_trimmed_reads_fasta="concatenated-trimmed-reads-all.fasta"
    export concatenated_trimmed_reads_fasta_no_spaces="concatenated-trimmed-reads-all-no_spaces.fasta"
    export genome_fasta_name="GCF_013753865.1_Amil_v2.1_genomic.fna"
    export genome_fasta_no_spaces="GCF_013753865.1_Amil_v2.1_genomic-no_spaces.fna"
    export mirdeep2_mapping_file="Apul-mirdeep2-mapping.arf"
    export cnidarian_mirna_fasta_name="cnidarian_miRNAs.fasta"

    # Paths to programs
    export mirdeep2_mapper="mapper.pl"
    export mirdeep2="miRDeep2.pl"
    export mirdeep2_fastaparse="fastaparse.pl"
    export bowtie_build="/home/shared/bowtie-1.3.1-linux-x86_64/bowtie-build"
    export seqtk="/home/shared/seqtk-1.4/seqtk"

    # Set number of CPUs to use
    export threads=40

# 2 Prepare reads for miRDeep2

## 2.1 Convert FastQ to FastA

``` bash
# Load bash variables into memory
source .bashvars

# Make output directory, if it doesn't exist
mkdir --parents "${output_dir_top}"

if [ ! -f "${output_dir_top}/${concatenated_trimmed_reads_fasta}" ]; then
  for fastq in ${trimmed_fastqs_dir}/${trimmed_fastqs_pattern}
  do
    ${seqtk} seq -A ${fastq} >> ${output_dir_top}/${concatenated_trimmed_reads_fasta}
  done
fi

head ${output_dir_top}/${concatenated_trimmed_reads_fasta}
```

    >GWNJ-1013:680:GW2305123155th:1:1101:5900:1000 1:N:0:CAACTAAT merged_29_0
    GGCTGTTTGGTCTAGTGGTATGATTCTCA
    >GWNJ-1013:680:GW2305123155th:1:1101:9914:1000 1:N:0:CAACTAAT merged_27_0
    GGCTGTTTGGTCTAGTGGTATGATTCT
    >GWNJ-1013:680:GW2305123155th:1:1101:13982:1000 1:N:0:CAACTAAT merged_28_0
    TTCTAGTTGGGTTTGATTTTTTAATGGT
    >GWNJ-1013:680:GW2305123155th:1:1101:22535:1000 1:N:0:CAACTAAT merged_31_0
    TAATCGATCTCGAACAGGACAGACAGGATCA
    >GWNJ-1013:680:GW2305123155th:1:1101:30942:1000 1:N:0:CAACTAAT merged_29_0
    TGTATCTTCAGGATTTTTAGCATCTTCAG

## 2.2 Remove Spaces in Reads FastA

``` bash
# Load bash variables into memory
source .bashvars

if [ ! -f "${output_dir_top}/${concatenated_trimmed_reads_fasta_no_spaces}" ]; then
  sed '/^>/ s/ /_/g' "${output_dir_top}/${concatenated_trimmed_reads_fasta}" \
  > "${output_dir_top}/${concatenated_trimmed_reads_fasta_no_spaces}"
fi

grep "^>" ${output_dir_top}/${concatenated_trimmed_reads_fasta} \
| head

echo ""
echo "--------------------------------------------------"
echo ""

grep "^>" ${output_dir_top}/${concatenated_trimmed_reads_fasta_no_spaces} \
| head
```

    >GWNJ-1013:680:GW2305123155th:1:1101:5900:1000 1:N:0:CAACTAAT merged_29_0
    >GWNJ-1013:680:GW2305123155th:1:1101:9914:1000 1:N:0:CAACTAAT merged_27_0
    >GWNJ-1013:680:GW2305123155th:1:1101:13982:1000 1:N:0:CAACTAAT merged_28_0
    >GWNJ-1013:680:GW2305123155th:1:1101:22535:1000 1:N:0:CAACTAAT merged_31_0
    >GWNJ-1013:680:GW2305123155th:1:1101:30942:1000 1:N:0:CAACTAAT merged_29_0
    >GWNJ-1013:680:GW2305123155th:1:1101:4481:1016 1:N:0:CAACTAAT merged_28_0
    >GWNJ-1013:680:GW2305123155th:1:1101:12020:1016 1:N:0:CAACTAAT merged_28_0
    >GWNJ-1013:680:GW2305123155th:1:1101:20121:1016 1:N:0:CAACTAAT merged_22_0
    >GWNJ-1013:680:GW2305123155th:1:1101:20971:1016 1:N:0:CAACTAAT merged_30_0
    >GWNJ-1013:680:GW2305123155th:1:1101:28330:1016 1:N:0:CAACTAAT merged_28_0

    --------------------------------------------------

    >GWNJ-1013:680:GW2305123155th:1:1101:5900:1000_1:N:0:CAACTAAT_merged_29_0
    >GWNJ-1013:680:GW2305123155th:1:1101:9914:1000_1:N:0:CAACTAAT_merged_27_0
    >GWNJ-1013:680:GW2305123155th:1:1101:13982:1000_1:N:0:CAACTAAT_merged_28_0
    >GWNJ-1013:680:GW2305123155th:1:1101:22535:1000_1:N:0:CAACTAAT_merged_31_0
    >GWNJ-1013:680:GW2305123155th:1:1101:30942:1000_1:N:0:CAACTAAT_merged_29_0
    >GWNJ-1013:680:GW2305123155th:1:1101:4481:1016_1:N:0:CAACTAAT_merged_28_0
    >GWNJ-1013:680:GW2305123155th:1:1101:12020:1016_1:N:0:CAACTAAT_merged_28_0
    >GWNJ-1013:680:GW2305123155th:1:1101:20121:1016_1:N:0:CAACTAAT_merged_22_0
    >GWNJ-1013:680:GW2305123155th:1:1101:20971:1016_1:N:0:CAACTAAT_merged_30_0
    >GWNJ-1013:680:GW2305123155th:1:1101:28330:1016_1:N:0:CAACTAAT_merged_28_0

# 3 Reformat genome FastA description lines

miRDeep2 can’t process genome FastAs with spaces in the description
lines.

So, I’m replacing spaces with underscores.

And, for aesthetics, I’m also removing commas.

``` bash
# Load bash variables into memory
source .bashvars


sed '/^>/ s/ /_/g' "${genome_fasta_dir}/${genome_fasta_name}" \
| sed '/^>/ s/,//g' \
> "${genome_fasta_dir}/${genome_fasta_no_spaces}"

grep "^>" ${genome_fasta_dir}/${genome_fasta_name} \
| head

echo ""
echo "--------------------------------------------------"
echo ""

grep "^>" ${genome_fasta_dir}/${genome_fasta_no_spaces} \
| head
```

    >NC_058066.1 Acropora millepora isolate JS-1 chromosome 1, Amil_v2.1, whole genome shotgun sequence
    >NC_058067.1 Acropora millepora isolate JS-1 chromosome 2, Amil_v2.1, whole genome shotgun sequence
    >NC_058068.1 Acropora millepora isolate JS-1 chromosome 3, Amil_v2.1, whole genome shotgun sequence
    >NC_058069.1 Acropora millepora isolate JS-1 chromosome 4, Amil_v2.1, whole genome shotgun sequence
    >NC_058070.1 Acropora millepora isolate JS-1 chromosome 5, Amil_v2.1, whole genome shotgun sequence
    >NC_058071.1 Acropora millepora isolate JS-1 chromosome 6, Amil_v2.1, whole genome shotgun sequence
    >NC_058072.1 Acropora millepora isolate JS-1 chromosome 7, Amil_v2.1, whole genome shotgun sequence
    >NC_058073.1 Acropora millepora isolate JS-1 chromosome 8, Amil_v2.1, whole genome shotgun sequence
    >NC_058074.1 Acropora millepora isolate JS-1 chromosome 9, Amil_v2.1, whole genome shotgun sequence
    >NC_058075.1 Acropora millepora isolate JS-1 chromosome 10, Amil_v2.1, whole genome shotgun sequence

    --------------------------------------------------

    >NC_058066.1_Acropora_millepora_isolate_JS-1_chromosome_1_Amil_v2.1_whole_genome_shotgun_sequence
    >NC_058067.1_Acropora_millepora_isolate_JS-1_chromosome_2_Amil_v2.1_whole_genome_shotgun_sequence
    >NC_058068.1_Acropora_millepora_isolate_JS-1_chromosome_3_Amil_v2.1_whole_genome_shotgun_sequence
    >NC_058069.1_Acropora_millepora_isolate_JS-1_chromosome_4_Amil_v2.1_whole_genome_shotgun_sequence
    >NC_058070.1_Acropora_millepora_isolate_JS-1_chromosome_5_Amil_v2.1_whole_genome_shotgun_sequence
    >NC_058071.1_Acropora_millepora_isolate_JS-1_chromosome_6_Amil_v2.1_whole_genome_shotgun_sequence
    >NC_058072.1_Acropora_millepora_isolate_JS-1_chromosome_7_Amil_v2.1_whole_genome_shotgun_sequence
    >NC_058073.1_Acropora_millepora_isolate_JS-1_chromosome_8_Amil_v2.1_whole_genome_shotgun_sequence
    >NC_058074.1_Acropora_millepora_isolate_JS-1_chromosome_9_Amil_v2.1_whole_genome_shotgun_sequence
    >NC_058075.1_Acropora_millepora_isolate_JS-1_chromosome_10_Amil_v2.1_whole_genome_shotgun_sequence

# 4 Bowtie v1 genome index

miRDeep2 requires a Bowtie v1 genome index - cannot use Bowtie2 index

``` bash
# Load bash variables into memory
source .bashvars

# Check for existence of genome index first
if [ ! -f "${genome_fasta_no_spaces%.*}.1.ebwt" ]; then
  ${bowtie_build} \
  ${genome_fasta_dir}/${genome_fasta_no_spaces} \
  ${genome_fasta_dir}/${genome_fasta_no_spaces%.*} \
  --threads ${threads} \
  --quiet
fi
```

# 5 Map reads to genome

Requires genome to be previously indexed with Bowtie.

Additionally, requires user to enter path to their `mirdeep2` directory
as well as their `perl5` installation.

``` bash
# Load bash variables into memory
source .bashvars

# Move to output directory
# miRDeep2 puts output files in current working directory
cd ${output_dir_top}

# Append miRDeep2 to system PATH and set PERL5LIB
export PATH=$PATH:/home/shared/mirdeep2/bin
export PERL5LIB=$PERL5LIB:/home/shared/mirdeep2/lib/perl5

# Run miRDeep2 mapping
time \
${mirdeep2_mapper} \
${concatenated_trimmed_reads_fasta_no_spaces} \
-c \
-m \
-l 18 \
-p ${genome_fasta_dir}/${genome_fasta_no_spaces%.*} \
-s ${collapsed_reads_gt17bp_mirdeep2} \
-t ${mirdeep2_mapping_file} \
-o ${threads}
```

# 6 Run miRDeep2

Recommendation is to use the closest related species in the miRDeep2
options, even if the species isn’t very closely related. The
documentation indicates that miRDeep2 is always more accurate when at
least a species is provided.

The options provided to the command are as follows:

- `none`: Known miRNAs of the species being analyzed.
- `cnidarian_mirna_fasta_name`: Related species miRNAs.
- `none`: Known miRNA precursors in this species.
- `-P`: Specifies miRBase version \> 18.
- `-v`: Remove temporary files after completion.
- `-g -1`: Number of precursors to analyze. A setting of `-1` will
  analyze all. Default is 50,000 I set this to `-1` after multiple
  attempts to run using the default kept failing.

NOTE: This will take an *extremely* long time to run (days). Could
possibly be shortened by excluding `randfold` analysis.

``` bash
# Load bash variables into memory
source .bashvars

# Move to output directory
# miRDeep2 puts output files in current working directory
cd ${output_dir_top}

# Append miRDeep2 to system PATH and set PERL5LIB
export PATH=$PATH:/home/shared/mirdeep2/bin
export PERL5LIB=$PERL5LIB:/home/shared/mirdeep2/lib/perl5

time \
${mirdeep2} \
${collapsed_reads_gt17bp_mirdeep2} \
${genome_fasta_dir}/${genome_fasta_no_spaces} \
${mirdeep2_mapping_file} \
none \
${deep_dive_data_dir}/${cnidarian_mirna_fasta_name} \
none \
-P \
-v \
-g -1 \
2>miRDeep2-report.log
```

## 6.1 Check runtime

``` bash
# Load bash variables into memory
source .bashvars

tail -n 6 ${output_dir_top}/miRDeep2-report.log
```

    miRDeep runtime: 

    started: 11:15:00
    ended: 11:5:35
    total:23h:50m:35s

# 7 Results

REMEMBER - You’ll need to set the name of your results file before
running the chunks below.

## 7.1 Peep output file format

The formatting of this CSV is terrible. It has a 3-column table on top
of a 17-column table. This makes parsing a bit of a pain in its raw
format.

``` bash
# Load bash variables into memory
source .bashvars


head -n 30 "${output_dir_top}/result_16_05_2024_t_11_15_00.csv"
```

    miRDeep2 score  estimated signal-to-noise   excision gearing
    10  7.5 1
    9   7.5 1
    8   7.7 1
    7   7.5 1
    6   7.4 1
    5   7.3 1
    4   7.1 1
    3   7.3 1
    2   6.6 1
    1   5.7 1
    0   5.3 1
    -1  3.4 1
    -2  2.4 1
    -3  1.9 1
    -4  1.6 1
    -5  1.2 1
    -6  1   1
    -7  1   1
    -8  1   1
    -9  1   1
    -10 1   1



    novel miRNAs predicted by miRDeep2
    provisional id  miRDeep2 score  estimated probability that the miRNA candidate is a true positive   rfam alert  total read count    mature read count   loop read count star read count significant randfold p-value    miRBase miRNA   example miRBase miRNA with the same seed    UCSC browser    NCBI blastn consensus mature sequence   consensus star sequence consensus precursor sequence    precursor coordinate
    NC_058066.1_Acropora_millepora_isolate_JS-1_chromosome_1_Amil_v2.1_whole_genome_shotgun_sequence_49320  127017.5        -   249134  248655  0   479 no  -   apa-mir-12434_Exaiptasia_pallida_Baumgarten_et_al._2017_NA  -   -   uuaacgaguagauaaaugaagaga    cuucguuuauucacucguucaua cuucguuuauucacucguucauauuuauuauuaacgaguagauaaaugaagaga  NC_058066.1_Acropora_millepora_isolate_JS-1_chromosome_1_Amil_v2.1_whole_genome_shotgun_sequence:20346247..20346301:-
    NC_058070.1_Acropora_millepora_isolate_JS-1_chromosome_5_Amil_v2.1_whole_genome_shotgun_sequence_208678 40612       -   79660   79120   0   540 no  -   -   -   -   ucucagauuacaguaguuaagu  cuuaacuacugcaaucugaacau ucucagauuacaguaguuaaguaauuaaauacuuaacuacugcaaucugaacau  NC_058070.1_Acropora_millepora_isolate_JS-1_chromosome_5_Amil_v2.1_whole_genome_shotgun_sequence:11598966..11599020:+
    NC_058068.1_Acropora_millepora_isolate_JS-1_chromosome_3_Amil_v2.1_whole_genome_shotgun_sequence_123927 40334.8     -   79120   73933   0   5187    no  -   -   -   -   ucugcguuaucggugaaauugu  auuucacuagauaagcgcuaac  auuucacuagauaagcgcuaacuguuauuguucugcguuaucggugaaauugu   NC_058068.1_Acropora_millepora_isolate_JS-1_chromosome_3_Amil_v2.1_whole_genome_shotgun_sequence:597846..597899:+

## 7.2 Create more easily parasable results file

This will match the line beginning with `provisional id` and print to
the end of the file (represented by the `$p`. `$` = end, `p` = print)

``` bash
# Load bash variables into memory
source .bashvars


sed --quiet '/provisional id/,$p' "${output_dir_top}/result_16_05_2024_t_11_15_00.csv" \
> "${output_dir_top}/parsable-result_16_05_2024_t_11_15_00.csv"

head "${output_dir_top}/parsable-result_16_05_2024_t_11_15_00.csv"
```

    provisional id  miRDeep2 score  estimated probability that the miRNA candidate is a true positive   rfam alert  total read count    mature read count   loop read count star read count significant randfold p-value    miRBase miRNA   example miRBase miRNA with the same seed    UCSC browser    NCBI blastn consensus mature sequence   consensus star sequence consensus precursor sequence    precursor coordinate
    NC_058066.1_Acropora_millepora_isolate_JS-1_chromosome_1_Amil_v2.1_whole_genome_shotgun_sequence_49320  127017.5        -   249134  248655  0   479 no  -   apa-mir-12434_Exaiptasia_pallida_Baumgarten_et_al._2017_NA  -   -   uuaacgaguagauaaaugaagaga    cuucguuuauucacucguucaua cuucguuuauucacucguucauauuuauuauuaacgaguagauaaaugaagaga  NC_058066.1_Acropora_millepora_isolate_JS-1_chromosome_1_Amil_v2.1_whole_genome_shotgun_sequence:20346247..20346301:-
    NC_058070.1_Acropora_millepora_isolate_JS-1_chromosome_5_Amil_v2.1_whole_genome_shotgun_sequence_208678 40612       -   79660   79120   0   540 no  -   -   -   -   ucucagauuacaguaguuaagu  cuuaacuacugcaaucugaacau ucucagauuacaguaguuaaguaauuaaauacuuaacuacugcaaucugaacau  NC_058070.1_Acropora_millepora_isolate_JS-1_chromosome_5_Amil_v2.1_whole_genome_shotgun_sequence:11598966..11599020:+
    NC_058068.1_Acropora_millepora_isolate_JS-1_chromosome_3_Amil_v2.1_whole_genome_shotgun_sequence_123927 40334.8     -   79120   73933   0   5187    no  -   -   -   -   ucugcguuaucggugaaauugu  auuucacuagauaagcgcuaac  auuucacuagauaagcgcuaacuguuauuguucugcguuaucggugaaauugu   NC_058068.1_Acropora_millepora_isolate_JS-1_chromosome_3_Amil_v2.1_whole_genome_shotgun_sequence:597846..597899:+
    NC_058071.1_Acropora_millepora_isolate_JS-1_chromosome_6_Amil_v2.1_whole_genome_shotgun_sequence_256047 39736.9     -   77937   65834   0   12103   no  -   spi-mir-temp-4_Stylophora_pistillata_Liew_et_al._2014_Exact_match_of_nve-miR-2023.  -   -   aaagaaguacaagugguaggg   cugcuacuuggacuuuuuuca   cugcuacuuggacuuuuuucacguuuaucgaugaaagaaguacaagugguaggg  NC_058071.1_Acropora_millepora_isolate_JS-1_chromosome_6_Amil_v2.1_whole_genome_shotgun_sequence:8840769..8840823:+
    NC_058068.1_Acropora_millepora_isolate_JS-1_chromosome_3_Amil_v2.1_whole_genome_shotgun_sequence_123929 38705.7     -   75922   73933   0   1989    no  -   -   -   -   ucugcguuaucggugaaauugu  auuucacuagaugagcgcuaac  auuucacuagaugagcgcuaacuguuauuguucugcguuaucggugaaauugu   NC_058068.1_Acropora_millepora_isolate_JS-1_chromosome_3_Amil_v2.1_whole_genome_shotgun_sequence:598142..598195:+
    NC_058073.1_Acropora_millepora_isolate_JS-1_chromosome_8_Amil_v2.1_whole_genome_shotgun_sequence_341478 25652.3     -   50311   49076   0   1235    no  -   apa-mir-2050_Exaiptasia_pallida_Baumgarten_et_al._2017_miR-2050;_Nve;_Spis;_Adi -   -   auugauuguagacaagccucuga aggguauugucuaugaucaaaaa aggguauugucuaugaucaaaaauuuuauuauugauuguagacaagccucuga   NC_058073.1_Acropora_millepora_isolate_JS-1_chromosome_8_Amil_v2.1_whole_genome_shotgun_sequence:11426819..11426872:-
    NC_058075.1_Acropora_millepora_isolate_JS-1_chromosome_10_Amil_v2.1_whole_genome_shotgun_sequence_406883    20352.5     -   39917   39897   0   20  no  -   hma-mir-new54_Hydra_magnipapillata_Krishna_et_al._2013_NA   -   -   uaguuuacaggcucaacuuugcu ggaguugaucuuguaaac  uaguuuacaggcucaacuuugcuuuaaagggaguugaucuuguaaac NC_058075.1_Acropora_millepora_isolate_JS-1_chromosome_10_Amil_v2.1_whole_genome_shotgun_sequence:15858707..15858754:-
    NC_058067.1_Acropora_millepora_isolate_JS-1_chromosome_2_Amil_v2.1_whole_genome_shotgun_sequence_81558  14705.5     -   28847   27971   0   876 no  -   -   -   -   uaggcguauuuccgauugucc   gacagccggaacuacgucugu   uaggcguauuuccgauuguccuuuuacuacgacagccggaacuacgucugu NC_058067.1_Acropora_millepora_isolate_JS-1_chromosome_2_Amil_v2.1_whole_genome_shotgun_sequence:21549048..21549099:+
    NC_058077.1_Acropora_millepora_isolate_JS-1_chromosome_12_Amil_v2.1_whole_genome_shotgun_sequence_459486    12860       -   25226   25161   0   65  no  -   -   -   -   agugcacuuuucucaggaugaa  cuucuugagaaaauuucaccca  agugcacuuuucucaggaugaauugaauucuucuugagaaaauuucaccca NC_058077.1_Acropora_millepora_isolate_JS-1_chromosome_12_Amil_v2.1_whole_genome_shotgun_sequence:21629096..21629147:+

## 7.3 Read in output CSV

This chunk provides a more concise overview of the data and it’s
columns.

``` r
mirdeep_result.df <- read.csv("../output/11.2-Apul-sRNAseq-miRdeep2-31bp-fastp-merged-cnidarian_only/parsable-result_16_05_2024_t_11_15_00.csv",
                              header = TRUE,
                              sep = "\t")

str(mirdeep_result.df)
```

    'data.frame':   438 obs. of  17 variables:
     $ provisional.id                                                   : chr  "NC_058066.1_Acropora_millepora_isolate_JS-1_chromosome_1_Amil_v2.1_whole_genome_shotgun_sequence_49320" "NC_058070.1_Acropora_millepora_isolate_JS-1_chromosome_5_Amil_v2.1_whole_genome_shotgun_sequence_208678" "NC_058068.1_Acropora_millepora_isolate_JS-1_chromosome_3_Amil_v2.1_whole_genome_shotgun_sequence_123927" "NC_058071.1_Acropora_millepora_isolate_JS-1_chromosome_6_Amil_v2.1_whole_genome_shotgun_sequence_256047" ...
     $ miRDeep2.score                                                   : num  127018 40612 40335 39737 38706 ...
     $ estimated.probability.that.the.miRNA.candidate.is.a.true.positive: logi  NA NA NA NA NA NA ...
     $ rfam.alert                                                       : chr  "-" "-" "-" "-" ...
     $ total.read.count                                                 : int  249134 79660 79120 77937 75922 50311 39917 28847 25226 24754 ...
     $ mature.read.count                                                : int  248655 79120 73933 65834 73933 49076 39897 27971 25161 24435 ...
     $ loop.read.count                                                  : int  0 0 0 0 0 0 0 0 0 0 ...
     $ star.read.count                                                  : int  479 540 5187 12103 1989 1235 20 876 65 319 ...
     $ significant.randfold.p.value                                     : chr  "no" "no" "no" "no" ...
     $ miRBase.miRNA                                                    : chr  "-" "-" "-" "-" ...
     $ example.miRBase.miRNA.with.the.same.seed                         : chr  "apa-mir-12434_Exaiptasia_pallida_Baumgarten_et_al._2017_NA" "-" "-" "spi-mir-temp-4_Stylophora_pistillata_Liew_et_al._2014_Exact_match_of_nve-miR-2023." ...
     $ UCSC.browser                                                     : chr  "-" "-" "-" "-" ...
     $ NCBI.blastn                                                      : chr  "-" "-" "-" "-" ...
     $ consensus.mature.sequence                                        : chr  "uuaacgaguagauaaaugaagaga" "ucucagauuacaguaguuaagu" "ucugcguuaucggugaaauugu" "aaagaaguacaagugguaggg" ...
     $ consensus.star.sequence                                          : chr  "cuucguuuauucacucguucaua" "cuuaacuacugcaaucugaacau" "auuucacuagauaagcgcuaac" "cugcuacuuggacuuuuuuca" ...
     $ consensus.precursor.sequence                                     : chr  "cuucguuuauucacucguucauauuuauuauuaacgaguagauaaaugaagaga" "ucucagauuacaguaguuaaguaauuaaauacuuaacuacugcaaucugaacau" "auuucacuagauaagcgcuaacuguuauuguucugcguuaucggugaaauugu" "cugcuacuuggacuuuuuucacguuuaucgaugaaagaaguacaagugguaggg" ...
     $ precursor.coordinate                                             : chr  "NC_058066.1_Acropora_millepora_isolate_JS-1_chromosome_1_Amil_v2.1_whole_genome_shotgun_sequence:20346247..20346301:-" "NC_058070.1_Acropora_millepora_isolate_JS-1_chromosome_5_Amil_v2.1_whole_genome_shotgun_sequence:11598966..11599020:+" "NC_058068.1_Acropora_millepora_isolate_JS-1_chromosome_3_Amil_v2.1_whole_genome_shotgun_sequence:597846..597899:+" "NC_058071.1_Acropora_millepora_isolate_JS-1_chromosome_6_Amil_v2.1_whole_genome_shotgun_sequence:8840769..8840823:+" ...

## 7.4 miRNAs count data

This provides some rudimentary numbers for the miRDeep2 output.

Further analysis is possibly desired to evaluate score thresholds, miRNA
families, etc.

``` bash
# Load bash variables into memory
source .bashvars

# Total predicted miRNAS
total_miRNAs=$(awk 'NR > 1' ${output_dir_top}/parsable-result_16_05_2024_t_11_15_00.csv \
| wc -l
)

echo "Total of predicted miRNAs: ${total_miRNAs}"
echo ""

# Matches to known mature miRNAs
mature_miRNAs=$(awk -F'\t' '$11 != "-" && $11 != "" {print $11}' ${output_dir_top}/parsable-result_16_05_2024_t_11_15_00.csv \
| wc -l
)

echo "Number of seed matches to known miRNAS: ${mature_miRNAs}"
echo ""

# Novel miRNAs
novel_miRNAs=$(awk -F "\t" '$11 == "-" || $11 == "" {print $11}' ${output_dir_top}/parsable-result_16_05_2024_t_11_15_00.csv \
| awk 'NR > 1' \
| wc -l
)

echo "Number of novel miRNAs: ${novel_miRNAs}"
```

    Total of predicted miRNAs: 438

    Number of seed matches to known miRNAS: 80

    Number of novel miRNAs: 358

------------------------------------------------------------------------

# 8 Citations

<div id="refs" class="references csl-bib-body hanging-indent">

<div id="ref-friedländer2011" class="csl-entry">

Friedländer, Marc R., Sebastian D. Mackowiak, Na Li, Wei Chen, and
Nikolaus Rajewsky. 2011. “miRDeep2 Accurately Identifies Known and
Hundreds of Novel microRNA Genes in Seven Animal Clades.” *Nucleic Acids
Research* 40 (1): 37–52. <https://doi.org/10.1093/nar/gkr688>.

</div>

</div>
