---
title: "13-Apul-sRNAseq-ShortStack"
author: "Sam White"
date: "2023-11-03"
output: 
  bookdown::html_document2:
    theme: cosmo
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
    code_download: true
bibliography: references.bib
---

```{r setup, include=FALSE}
library(knitr)
library(kableExtra)
library(dplyr)
library(reticulate)
knitr::opts_chunk$set(
  echo = TRUE,         # Display code chunks
  eval = FALSE,        # Evaluate code chunks
  warning = FALSE,     # Hide warnings
  message = FALSE,     # Hide messages
  comment = ""         # Prevents appending '##' to beginning of lines in code output
)
```

Use [ShortStack](https://github.com/MikeAxtell/ShortStack) [@johnson2016][@axtell2013][@shahid2014] to perform alignment of sRNAseq data and annotation of sRNA-producing genes.

Uses trimmed reads from [08-Apul-sRNAseq-trimming.Rmd](https://github.com/urol-e5/deep-dive/blob/main/D-Apul/code/08-Apul-sRNAseq-trimming.Rmd).  Expect FastQ filename format like:

`*flexbar_trim.20230621_[12]*.fastq.gz`

---
Inputs:

- Genome FastA.

Outputs:

- Confidence-filtered GFF of predicted miRNAs.
- Unfiltered FastA of predicted miRNAs.
- Unfiltered GFF of predicted miRNAs.

Software requirements:

- Utilizes a [ShortStack](https://github.com/MikeAxtell/ShortStack#installation) Conda/Mamba environment, per the installation instructions.

---


# Set R variables
Replace with name of your ShortStack environment and the path to the corresponding
conda installation (find this _after_ you've activated the environment).

E.g. 

```{r conda-evn-example, eval=FALSE}
# Activate environment
conda activate ShortStack4_env

# Find conda path
which conda
```


```{r R-variables, eval=TRUE}
shortstack_conda_env_name <- c("ShortStack4_env")
shortstack_cond_path <- c("/home/sam/programs/mambaforge/condabin/conda")
```

# Create a Bash variables file

This allows usage of Bash variables across R Markdown chunks.

```{r save-bash-variables-to-rvars-file, engine='bash', eval=TRUE}
{
echo "#### Assign Variables ####"
echo ""

echo "# Trimmed FastQ naming pattern"
echo "export trimmed_fastq_pattern='*flexbar_trim.25bp.fastq.gz'"

echo "# Data directories"
echo 'export deep_dive_dir=/home/shared/8TB_HDD_01/sam/gitrepos/deep-dive'
echo 'export deep_dive_data_dir="${deep_dive_dir}/data"'
echo 'export output_dir_top=${deep_dive_dir}/D-Apul/output/13-Apul-sRNAseq-ShortStack'
echo 'export trimmed_fastqs_dir="${deep_dive_dir}/D-Apul/output/08-Apul-sRNAseq-trimming/trimmed-reads"'
echo ""

echo "# Input/Output files"
echo 'export genome_fasta_dir=${deep_dive_dir}/D-Apul/data/Amil/ncbi_dataset/data/GCF_013753865.1'
echo 'export genome_fasta_name="GCF_013753865.1_Amil_v2.1_genomic.fna"'
echo 'export genome_fasta="${genome_fasta_dir}/${genome_fasta_name}"'
echo 'export mirbase_mature_fasta=mature.fa'
echo 'export mirbase_mature_fasta_version=mirbase-mature-v22.1.fa'
echo ""

echo "# External data URLs"
echo 'export mirbase_fasta_url="https://mirbase.org/download_version_files/22.1/"'
echo ""

echo "# Set number of CPUs to use"
echo 'export threads=40'
echo ""

echo "# Initialize arrays"
echo 'export trimmed_fastqs_array=()'


} > .bashvars

cat .bashvars
```

# Load [ShortStack](https://github.com/MikeAxtell/ShortStack conda environment

If this is successful, the first line of output should show that the Python being used
is the one in your [ShortStack](https://github.com/MikeAxtell/ShortStack conda environment path.

E.g.

`python:         /home/sam/programs/mambaforge/envs/mirmachine_env/bin/python`

```{r load-shortstack-conda-env, eval=TRUE}
use_condaenv(condaenv = shortstack_conda_env_name, conda = shortstack_cond_path)

# Check successful env loading
py_config()
```


# Download [miRBase](https://mirbase.org/) mature miRNA FastA
```{r download-mirbase-mature-fasta, engine='bash', eval=TRUE}
# Load bash variables into memory
source .bashvars

wget \
--directory-prefix ${deep_dive_data_dir} \
--recursive \
--no-check-certificate \
--continue \
--no-host-directories \
--no-directories \
--no-parent \
--quiet \
--execute robots=off \
 ${mirbase_fasta_url}/${mirbase_mature_fasta}

# Rename to indicate miRBase FastA version
mv ${deep_dive_data_dir}/${mirbase_mature_fasta} ${deep_dive_data_dir}/${mirbase_mature_fasta_version}

ls -lh "${deep_dive_data_dir}"
```
# Run ShortStack

```{r shortstack, engine='bash', eval=TRUE}
# Load bash variables into memory
source .bashvars

# Create array of trimmed FastQs
trimmed_fastqs_array=(${fastq_pattern})

# Pass array contents to new variable as space-delimited list
trimmed_fastqs_list=$(echo "${trimmed_fastq_array[*]}")

###### Run ShortStack ######
time \
${programs_array[ShortStack]} \
--genomefile "${genome_fasta}" \
--readfile ${trimmed_fastqs_list} \
--known_miRNAs ${deep_dive_data_dir}/${mirbase_mature_fasta_version} \
--dn_mirna \
--threads ${threads} \
--outdir ${output_dir_top}

######## Create MD5 checksums for trimmed FastQs ########
# Serves a way to list FastQs used for ShortStack and record checksums
for fastq in ${fastq_pattern}
do
  md5sum "${fastq}"
done
```



------------------------------------------------------------------------

# Citations
