13.1-Apul-sRNAseq-ShortStack-R1-reads
================
Sam White
2023-11-03

- [1 Set R variables](#1-set-r-variables)
- [2 Create a Bash variables file](#2-create-a-bash-variables-file)
- [3 Load ShortStack conda
  environment](#3-load-shortstack-conda-environment)
- [4 Download miRBase mature miRNA
  FastA](#4-download-mirbase-mature-mirna-fasta)
- [5 Run ShortStack](#5-run-shortstack)
  - [5.1 Modify genome filename for ShortStack
    compatability](#51-modify-genome-filename-for-shortstack-compatability)
  - [5.2 Excecute ShortStack command](#52-excecute-shortstack-command)
  - [5.3 Check runtime](#53-check-runtime)
- [6 Results](#6-results)
  - [6.1 ShortStack synopsis](#61-shortstack-synopsis)
  - [6.2 Inspect `Results.txt`](#62-inspect-resultstxt)
    - [6.2.1 Directory tree of all ShortStack
      outputs](#621-directory-tree-of-all-shortstack-outputs)
- [7 Citations](#7-citations)

Use [ShortStack](https://github.com/MikeAxtell/ShortStack) ([Axtell
2013](#ref-axtell2013a); [Shahid and Axtell 2014](#ref-shahid2014);
[Johnson et al. 2016](#ref-johnson2016a))to perform alignment of sRNAseq
data and annotation of sRNA-producing genes.

The *A.millepora* genome will be used as the reference genome for
*A.pulchra*, as *A.pulchra* does not currently have a sequenced genome
and *A.millepora* had highest alignment rates for standard RNAseq data
compared to other published genomes tested.

------------------------------------------------------------------------

Inputs:

- Requires trimmed sRNAseq files generated by
  [08.1-Dapul-sRNAseq-trimming-R1-only.Rmd](https://github.com/urol-e5/deep-dive/blob/main/D-Apul/code/08.1-Dapul-sRNAseq-trimming-R1-only.Rmd)

  - Filenames formatted: `*flexbar_trim.25bp*.gz`

- *A.millepora* genome FastA. See
  [12-Apul-sRNAseq-MirMachine.Rmd](https://github.com/urol-e5/deep-dive/blob/main/D-Apul/code/12-Apul-sRNAseq-MirMachine.Rmd)
  for download info if needed.

Outputs:

- See [ShortStack outputs
  documentation](https://github.com/MikeAxtell/ShortStack#outputs) for
  full list and detailed descriptions.

Software requirements:

- Utilizes a
  [ShortStack](https://github.com/MikeAxtell/ShortStack#installation)
  Conda/Mamba environment, per the installation instructions.

Replace with name of your ShortStack environment and the path to the
corresponding conda installation (find this *after* you’ve activated the
environment).

E.g.

``` bash
# Activate environment
conda activate ShortStack4_env

# Find conda path
which conda
```

------------------------------------------------------------------------

# 1 Set R variables

``` r
shortstack_conda_env_name <- c("ShortStack-4.0.3_env")
shortstack_cond_path <- c("/home/sam/programs/mambaforge/condabin/conda")
```

# 2 Create a Bash variables file

This allows usage of Bash variables across R Markdown chunks.

``` bash
{
echo "#### Assign Variables ####"
echo ""

echo "# Trimmed FastQ naming pattern"
echo "export trimmed_fastqs_pattern='*fastp-R1-31bp-auto_adapters-polyG.fq.gz'"

echo "# Data directories"
echo 'export deep_dive_dir=/home/shared/8TB_HDD_01/sam/gitrepos/deep-dive'
echo 'export deep_dive_data_dir="${deep_dive_dir}/data"'
echo 'export output_dir_top=${deep_dive_dir}/D-Apul/output/13.1-Apul-sRNAseq-ShortStack-R1-reads'
echo 'export trimmed_fastqs_dir="${deep_dive_dir}/D-Apul/output/08.1-Dapul-sRNAseq-trimming-R1-only/trimmed-reads"'
echo ""

echo "# Input/Output files"
echo 'export genome_fasta_dir=${deep_dive_dir}/D-Apul/data/Amil/ncbi_dataset/data/GCF_013753865.1'
echo 'export genome_fasta_name="GCF_013753865.1_Amil_v2.1_genomic.fna"'
echo 'export shortstack_genome_fasta_name="GCF_013753865.1_Amil_v2.1_genomic.fa"'
echo 'export mirbase_mature_fasta=mature.fa'
echo 'export mirbase_mature_fasta_version=mirbase-mature-v22.1.fa'
echo 'export genome_fasta="${genome_fasta_dir}/${shortstack_genome_fasta_name}"'
echo ""

echo "# External data URLs"
echo 'export mirbase_fasta_url="https://mirbase.org/download_version_files/22.1/"'
echo ""

echo "# Set number of CPUs to use"
echo 'export threads=40'
echo ""

echo "# Initialize arrays"
echo 'export trimmed_fastqs_array=()'


} > .bashvars

cat .bashvars
```

    #### Assign Variables ####

    # Trimmed FastQ naming pattern
    export trimmed_fastqs_pattern='*fastp-R1-31bp-auto_adapters-polyG.fq.gz'
    # Data directories
    export deep_dive_dir=/home/shared/8TB_HDD_01/sam/gitrepos/deep-dive
    export deep_dive_data_dir="${deep_dive_dir}/data"
    export output_dir_top=${deep_dive_dir}/D-Apul/output/13.1-Apul-sRNAseq-ShortStack-R1-reads
    export trimmed_fastqs_dir="${deep_dive_dir}/D-Apul/output/08.1-Dapul-sRNAseq-trimming-R1-only/trimmed-reads"

    # Input/Output files
    export genome_fasta_dir=${deep_dive_dir}/D-Apul/data/Amil/ncbi_dataset/data/GCF_013753865.1
    export genome_fasta_name="GCF_013753865.1_Amil_v2.1_genomic.fna"
    export shortstack_genome_fasta_name="GCF_013753865.1_Amil_v2.1_genomic.fa"
    export mirbase_mature_fasta=mature.fa
    export mirbase_mature_fasta_version=mirbase-mature-v22.1.fa
    export genome_fasta="${genome_fasta_dir}/${shortstack_genome_fasta_name}"

    # External data URLs
    export mirbase_fasta_url="https://mirbase.org/download_version_files/22.1/"

    # Set number of CPUs to use
    export threads=40

    # Initialize arrays
    export trimmed_fastqs_array=()

# 3 Load [ShortStack](https://github.com/MikeAxtell/ShortStack) conda environment

If this is successful, the first line of output should show that the
Python being used is the one in your
$$ShortStack$$(<https://github.com/MikeAxtell/ShortStack> conda
environment path.

E.g.

`python:         /home/sam/programs/mambaforge/envs/mirmachine_env/bin/python`

``` r
use_condaenv(condaenv = shortstack_conda_env_name, conda = shortstack_cond_path)

# Check successful env loading
py_config()
```

    python:         /home/sam/programs/mambaforge/envs/ShortStack-4.0.3_env/bin/python
    libpython:      /home/sam/programs/mambaforge/envs/ShortStack-4.0.3_env/lib/libpython3.10.so
    pythonhome:     /home/sam/programs/mambaforge/envs/ShortStack-4.0.3_env:/home/sam/programs/mambaforge/envs/ShortStack-4.0.3_env
    version:        3.10.13 | packaged by conda-forge | (main, Dec 23 2023, 15:36:39) [GCC 12.3.0]
    numpy:          /home/sam/programs/mambaforge/envs/ShortStack-4.0.3_env/lib/python3.10/site-packages/numpy
    numpy_version:  1.26.4

    NOTE: Python version was forced by use_python() function

# 4 Download [miRBase](https://mirbase.org/) mature miRNA FastA

``` bash
# Load bash variables into memory
source .bashvars

wget \
--directory-prefix ${deep_dive_data_dir} \
--recursive \
--no-check-certificate \
--continue \
--no-host-directories \
--no-directories \
--no-parent \
--quiet \
--execute robots=off \
 ${mirbase_fasta_url}/${mirbase_mature_fasta}

# Rename to indicate miRBase FastA version
mv ${deep_dive_data_dir}/${mirbase_mature_fasta} ${deep_dive_data_dir}/${mirbase_mature_fasta_version}

ls -lh "${deep_dive_data_dir}"
```

    total 13M
    drwxr-xr-x 2 sam sam 4.0K Nov  7 14:36 blast_dbs
    -rw-r--r-- 1 sam sam 3.7M Feb 13 11:18 mirbase-mature-v22.1.fa
    -rw-r--r-- 1 sam sam 3.7M Dec  4 11:05 mirbase-mature-v22.1-no_spaces.fa
    -rw-r--r-- 1 sam sam 3.7M Nov 17 07:59 mirbase-mature-v22.1-no_U.fa
    -rw-r--r-- 1 sam sam 726K Nov  7 14:36 mirgene-mature-all-v2.1.fa
    -rw-r--r-- 1 sam sam 726K Nov 17 07:59 mirgene-mature-all-v2.1-no_U.fa

# 5 Run ShortStack

## 5.1 Modify genome filename for ShortStack compatability

``` bash
# Load bash variables into memory
source .bashvars

# Check for FastA file first
# Then create rename file if doesn't exist
if [ -f "${genome_fasta_dir}/${shortstack_genome_fasta_name}" ]; then
  echo "${genome_fasta_dir}/${shortstack_genome_fasta_name} already exists. Nothing to do."
  echo ""
else

  # Copy genome FastA to ShortStack-compatible filename (ending with .fa)
  cp ${genome_fasta_dir}/${genome_fasta_name} ${genome_fasta_dir}/${shortstack_genome_fasta_name}
fi

# Confirm
ls -lh ${genome_fasta_dir}/${shortstack_genome_fasta_name}
```

    /home/shared/8TB_HDD_01/sam/gitrepos/deep-dive/D-Apul/data/Amil/ncbi_dataset/data/GCF_013753865.1/GCF_013753865.1_Amil_v2.1_genomic.fa already exists. Nothing to do.

    -rw-r--r-- 1 sam sam 460M Nov  6 12:40 /home/shared/8TB_HDD_01/sam/gitrepos/deep-dive/D-Apul/data/Amil/ncbi_dataset/data/GCF_013753865.1/GCF_013753865.1_Amil_v2.1_genomic.fa

## 5.2 Excecute ShortStack command

Uses the `--dn_mirna` option to identify miRNAs in the genome, without
relying on the `--known_miRNAs`.

This part of the code redirects the output of `time` to the end of
`shortstack.log` file.

- `; } \ 2>> ${output_dir_top}/shortstack.log`

``` bash
# Load bash variables into memory
source .bashvars

# Make output directory, if it doesn't exist
mkdir "${output_dir_top}"

# Create array of trimmed FastQs
trimmed_fastqs_array=(${trimmed_fastqs_dir}/${trimmed_fastqs_pattern})


# Pass array contents to new variable as space-delimited list
trimmed_fastqs_list=$(echo "${trimmed_fastqs_array[*]}")


###### Run ShortStack ######
{ time \
ShortStack \
--genomefile "${genome_fasta}" \
--readfile ${trimmed_fastqs_list} \
--known_miRNAs ${deep_dive_data_dir}/${mirbase_mature_fasta_version} \
--dn_mirna \
--threads ${threads} \
--outdir ${output_dir_top}/ShortStack_out \
&> ${output_dir_top}/shortstack.log ; } \
2>> ${output_dir_top}/shortstack.log
```

## 5.3 Check runtime

``` bash
# Load bash variables into memory
source .bashvars

tail -n 3 ${output_dir_top}/shortstack.log \
| grep "real" \
| awk '{print "ShortStack runtime:" "\t" $2}'
```

    ShortStack runtime: 70m8.432s

# 6 Results

## 6.1 ShortStack synopsis

``` bash
# Load bash variables into memory
source .bashvars

tail -n 25 ${output_dir_top}/shortstack.log
```

    Writing final files

    Found a total of 25 MIRNA loci


    Non-MIRNA loci by DicerCall:
    N 18280
    22 31
    23 26
    21 9
    24 6

    Creating visualizations of microRNA loci with strucVis
    <<< WARNING >>>
    Do not rely on these results alone to annotate new MIRNA loci!
    The false positive rate for de novo MIRNA identification is low, but NOT ZERO
    Insepct each mirna locus, especially the strucVis output, and see
    https://doi.org/10.1105/tpc.17.00851 , https://doi.org/10.1093/nar/gky1141

    Tue 13 Feb 2024 10:36:14 -0800 PST
    Run Completed!

    real    70m8.432s
    user    1142m17.752s
    sys 354m13.933s

ShortStack identified 25 miRNAs.

## 6.2 Inspect `Results.txt`

``` bash
# Load bash variables into memory
source .bashvars

head ${output_dir_top}/ShortStack_out/Results.txt

echo ""
echo "----------------------------------------------------------"
echo ""

echo "Nummber of potential loci:"
awk '(NR>1)' ${output_dir_top}/ShortStack_out/Results.txt | wc -l
```

    Locus   Name    Chrom   Start   End Length  Reads   DistinctSequences   FracTop Strand  MajorRNA    MajorRNAReads   Short   Long    21  22  23  24  DicerCall   MIRNA   known_miRNAs
    NC_058066.1:22060-22476 Cluster_1   NC_058066.1 22060   22476   417 336 13  0.9851190476190477  +   UGGUACAAGAAUGUCUC   322 333 3   0   0   0   0   N   N   NA
    NC_058066.1:25992-26407 Cluster_2   NC_058066.1 25992   26407   416 3949    13  0.006583945302608255    -   UAGGGUACUGUCGGCU    2852    3948    1   0   0   0   0   N   N   NA
    NC_058066.1:97939-98598 Cluster_3   NC_058066.1 97939   98598   660 950 247 0.9673684210526315  +   UGUCUGUUGUAGCUUUGUAUUUUGCCAGC   271 25  905 1   3   6   10  N   N   NA
    NC_058066.1:152483-152901   Cluster_4   NC_058066.1 152483  152901  419 135 34  0.044444444444444446    -   UAAGUACUUUAUCAACUAACUCUAGGCA    70  3   123 0   2   0   7   N   N   NA
    NC_058066.1:161122-161538   Cluster_5   NC_058066.1 161122  161538  417 433 186 0.3071593533487298  .   UUUUAGCCUAGUGCGGGUUUCCAGACGU    42  29  380 8   2   1   13  N   N   NA
    NC_058066.1:204731-205142   Cluster_6   NC_058066.1 204731  205142  412 298 149 0.8221476510067114  +   UCCCAACACGUCUAGACUGUACAAUUUCU   34  0   294 1   0   1   2   N   N   NA
    NC_058066.1:205941-206966   Cluster_7   NC_058066.1 205941  206966  1026    1889    368 0.35150873478030703 .   CAAAAGAGCGGACAAAAUAGUCGACAGAUU  727 11  1834    9   2   13  20  N   N   NA
    NC_058066.1:210855-211343   Cluster_8   NC_058066.1 210855  211343  489 1227    337 0.7612061939690301  .   UAAUACUUGUAGUGAAGGUUCAAUCUCGA   95  12  1117    7   10  21  60  N   N   NA
    NC_058066.1:349656-351297   Cluster_9   NC_058066.1 349656  351297  1642    3430    1231    0.810204081632653   +   UCAGCUUGGAAAUGACAGCUUUUGACGU    303 53  3274    10  19  20  54  N   N   NA

    ----------------------------------------------------------

    Nummber of potential loci:
    18377

Column 20 of the `Results.txt` file identifies if a cluster is a miRNA
or not (`Y` or `N`).

``` bash
# Load bash variables into memory
source .bashvars

echo "Number of loci characterized as miRNA:"
awk '$20=="Y" {print $0}' ${output_dir_top}/ShortStack_out/Results.txt \
| wc -l
echo ""

echo "----------------------------------------------------------"

echo ""
echo "Number of loci _not_ characterized as miRNA:"
awk '$20=="N" {print $0}' ${output_dir_top}/ShortStack_out/Results.txt \
| wc -l
```

    Number of loci characterized as miRNA:
    25

    ----------------------------------------------------------

    Number of loci _not_ characterized as miRNA:
    18352

Column 21 of the `Results.txt` file identifies if a cluster aligned to a
known miRNA (miRBase) or not (`Y` or `NA`).

Since there are no miRNAs, the following code will *not* print any
output.

The `echo` command after the `awk` command is simply there to prove that
the chunk executed.

``` bash
# Load bash variables into memory
source .bashvars

echo "Number of loci matching miRBase miRNAs:"
awk '$21!="NA" {print $0}' ${output_dir_top}/ShortStack_out/Results.txt \
| wc -l
echo ""

echo "----------------------------------------------------------"

echo ""
echo "Number of loci _not_ matching miRBase miRNAs:"
awk '$21=="NA" {print $0}' ${output_dir_top}/ShortStack_out/Results.txt \
| wc -l
```

    Number of loci matching miRBase miRNAs:
    54

    ----------------------------------------------------------

    Number of loci _not_ matching miRBase miRNAs:
    18324

Although there are loci with matches to miRBase miRNAs, ShortStack did
*not* annotate these clusters as miRNAs likely [because they do not
*also* match secondary structure
criteria](https://github.com/MikeAxtell/ShortStack#mirna-annotation).

### 6.2.1 Directory tree of all ShortStack outputs

Many of these are large (by GitHub standards) BAM files, so will not be
added to the repo.

Additionally, it’s unlikely we’ll utilize most of the other files
(bigwig) generated by ShortStack.

``` bash
# Load bash variables into memory
source .bashvars

tree -h ${output_dir_top}/
```

    /home/shared/8TB_HDD_01/sam/gitrepos/deep-dive/D-Apul/output/13.1-Apul-sRNAseq-ShortStack-R1-reads/
    ├── [ 27K]  shortstack.log
    └── [ 36K]  ShortStack_out
        ├── [ 45K]  alignment_details.tsv
        ├── [1.2M]  Counts.txt
        ├── [ 87K]  known_miRNAs.gff3
        ├── [1.8M]  known_miRNAs_unaligned.fasta
        ├── [6.0M]  merged_alignments_21_m.bw
        ├── [6.5M]  merged_alignments_21_p.bw
        ├── [5.6M]  merged_alignments_22_m.bw
        ├── [6.0M]  merged_alignments_22_p.bw
        ├── [ 11M]  merged_alignments_23-24_m.bw
        ├── [ 12M]  merged_alignments_23-24_p.bw
        ├── [2.2G]  merged_alignments.bam
        ├── [216K]  merged_alignments.bam.csi
        ├── [ 94M]  merged_alignments_other_m.bw
        ├── [ 97M]  merged_alignments_other_p.bw
        ├── [ 48M]  merged_alignments_sRNA-ACR-140-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bw
        ├── [ 54M]  merged_alignments_sRNA-ACR-145-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bw
        ├── [ 51M]  merged_alignments_sRNA-ACR-150-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bw
        ├── [ 44M]  merged_alignments_sRNA-ACR-173-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bw
        ├── [ 45M]  merged_alignments_sRNA-ACR-178-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bw
        ├── [ 19M]  merged_alignments_sRNA-POR-73-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bw
        ├── [ 22M]  merged_alignments_sRNA-POR-79-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bw
        ├── [ 26M]  merged_alignments_sRNA-POR-82-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bw
        ├── [7.3K]  mir.fasta
        ├── [1.8M]  Results.gff3
        ├── [2.8M]  Results.txt
        ├── [261M]  sRNA-ACR-140-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bam
        ├── [227K]  sRNA-ACR-140-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bam.csi
        ├── [299M]  sRNA-ACR-145-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bam
        ├── [227K]  sRNA-ACR-145-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bam.csi
        ├── [311M]  sRNA-ACR-150-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bam
        ├── [231K]  sRNA-ACR-150-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bam.csi
        ├── [275M]  sRNA-ACR-173-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bam
        ├── [228K]  sRNA-ACR-173-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bam.csi
        ├── [248M]  sRNA-ACR-178-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bam
        ├── [229K]  sRNA-ACR-178-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bam.csi
        ├── [218M]  sRNA-POR-73-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bam
        ├── [228K]  sRNA-POR-73-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bam.csi
        ├── [223M]  sRNA-POR-79-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bam
        ├── [230K]  sRNA-POR-79-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bam.csi
        ├── [255M]  sRNA-POR-82-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bam
        ├── [230K]  sRNA-POR-82-S1-TP2_R1_001.fastp-R1-31bp-auto_adapters-polyG.bam.csi
        └── [4.0K]  strucVis
            ├── [ 12K]  Cluster_10044.ps
            ├── [ 34K]  Cluster_10044.txt
            ├── [ 11K]  Cluster_10197.ps
            ├── [5.8K]  Cluster_10197.txt
            ├── [ 12K]  Cluster_10499.ps
            ├── [ 11K]  Cluster_10499.txt
            ├── [ 11K]  Cluster_10724.ps
            ├── [7.5K]  Cluster_10724.txt
            ├── [ 12K]  Cluster_1877.ps
            ├── [9.9K]  Cluster_1877.txt
            ├── [ 12K]  Cluster_2008.ps
            ├── [ 36K]  Cluster_2008.txt
            ├── [ 12K]  Cluster_2101.ps
            ├── [ 45K]  Cluster_2101.txt
            ├── [ 12K]  Cluster_2543.ps
            ├── [ 44K]  Cluster_2543.txt
            ├── [ 11K]  Cluster_2544.ps
            ├── [ 32K]  Cluster_2544.txt
            ├── [ 11K]  Cluster_3043.ps
            ├── [5.1K]  Cluster_3043.txt
            ├── [ 12K]  Cluster_3121.ps
            ├── [7.4K]  Cluster_3121.txt
            ├── [ 11K]  Cluster_346.ps
            ├── [ 20K]  Cluster_346.txt
            ├── [ 12K]  Cluster_3701.ps
            ├── [ 18K]  Cluster_3701.txt
            ├── [ 12K]  Cluster_4295.ps
            ├── [ 31K]  Cluster_4295.txt
            ├── [ 11K]  Cluster_4890.ps
            ├── [ 16K]  Cluster_4890.txt
            ├── [ 12K]  Cluster_5293.ps
            ├── [ 70K]  Cluster_5293.txt
            ├── [ 18K]  Cluster_5486.ps
            ├── [ 16K]  Cluster_5486.txt
            ├── [ 12K]  Cluster_6406.ps
            ├── [ 22K]  Cluster_6406.txt
            ├── [ 11K]  Cluster_6481.ps
            ├── [ 46K]  Cluster_6481.txt
            ├── [ 12K]  Cluster_6968.ps
            ├── [ 47K]  Cluster_6968.txt
            ├── [ 12K]  Cluster_6973.ps
            ├── [ 13K]  Cluster_6973.txt
            ├── [ 12K]  Cluster_6988.ps
            ├── [ 92K]  Cluster_6988.txt
            ├── [ 11K]  Cluster_7124.ps
            ├── [ 27K]  Cluster_7124.txt
            ├── [ 11K]  Cluster_8463.ps
            ├── [5.6K]  Cluster_8463.txt
            ├── [ 12K]  Cluster_9446.ps
            └── [ 31K]  Cluster_9446.txt

    2 directories, 92 files

------------------------------------------------------------------------

# 7 Citations

<div id="refs" class="references csl-bib-body hanging-indent">

<div id="ref-axtell2013a" class="csl-entry">

Axtell, Michael J. 2013. “ShortStack: Comprehensive Annotation and
Quantification of Small RNA Genes.” *RNA* 19 (6): 740–51.
<https://doi.org/10.1261/rna.035279.112>.

</div>

<div id="ref-johnson2016a" class="csl-entry">

Johnson, Nathan R, Jonathan M Yeoh, Ceyda Coruh, and Michael J Axtell.
2016. “Improved Placement of Multi-Mapping Small RNAs.” *G3
Genes\|Genomes\|Genetics* 6 (7): 2103–11.
<https://doi.org/10.1534/g3.116.030452>.

</div>

<div id="ref-shahid2014" class="csl-entry">

Shahid, Saima, and Michael J. Axtell. 2014. “Identification and
Annotation of Small RNA Genes Using ShortStack.” *Methods* 67 (1):
20–27. <https://doi.org/10.1016/j.ymeth.2013.10.004>.

</div>

</div>
