---
title: "methods section"
editor: visual
format:
  html:
    code-fold: true
    code-summary: "Show the code"
---

::: callout-important
This is draft method for Subproject "A"

Nutrient exposure experiment - P verrocosa genome
:::

### DNA Methylation Analysis

Deduplicated Alignment files (BAMs) were downloaded [from Becker](https://gannet.fish.washington.edu/Atumefaciens/hputnam-Becker_E5/WGBS_Becker_E5/Becker_WGBS/bismark_deduplicated/)

Methylation level was extracted

    ${bismark_dir}/bismark_methylation_extractor \
    --bedGraph --counts --scaffolds \
    --multicore 28 \
    --counts \
    --comprehensive \
    --merge_non_CpG \
    --buffer_size 75% \
    *deduplicated.bam

This was followed by coverage file generation

    find *deduplicated.bismark.cov.gz \
    | xargs basename -s _R1_001_val_1_bismark_bt2_pe.deduplicated.bismark.cov.gz \
    | xargs -I{} ${bismark_dir}/coverage2cytosine \
    --genome_folder ${genome_folder} \
    -o {} \
    --merge_CpG \
    --zero_based \
    {}_R1_001_val_1_bismark_bt2_pe.deduplicated.bismark.cov.gz

and files generated fofre
