09-Pmea-sRNAseq-miRTrace
================
Sam White (modified by K Durkin for P. meandrina analysis)
2023-11-14

- <a href="#1-create-a-bash-variables-file"
  id="toc-1-create-a-bash-variables-file">1 Create a Bash variables
  file</a>
- <a href="#2-create-mirtrace-config-file"
  id="toc-2-create-mirtrace-config-file">2 Create miRTrace config file</a>
- <a href="#3-run-mirtrace" id="toc-3-run-mirtrace">3 Run miRTrace</a>
- <a href="#4-results" id="toc-4-results">4 Results</a>
  - <a href="#41-read-in-table-as-data-frame"
    id="toc-41-read-in-table-as-data-frame">4.1 Read in table as data
    frame</a>
  - <a href="#42-number-of-samples-with-matches"
    id="toc-42-number-of-samples-with-matches">4.2 Number of samples with
    matches</a>
  - <a href="#43-percentage-of-samples-with-matches"
    id="toc-43-percentage-of-samples-with-matches">4.3 Percentage of samples
    with matches</a>
  - <a href="#44-number-of-clades-with-matches"
    id="toc-44-number-of-clades-with-matches">4.4 Number of clades with
    matches</a>
  - <a href="#45-mirtrace-table" id="toc-45-mirtrace-table">4.5 miRTrace
    table</a>
- <a href="#5-citations" id="toc-5-citations">5 Citations</a>

Use [miRTrace](https://github.com/friedlanderlab/mirtrace) \[@kang2018\]
to identify taxonomic origins of miRNA sequencing data.

NOTE: This requires you to have previously run
[`08-Pmea-sRNAseq-trimming.Rmd`](https://github.com/urol-e5/deep-dive/blob/main/F-Pmea/code/08-Pmea-sRNAseq-trimming.Rmd),
as the code relies on the trimmed reads output from that code.

------------------------------------------------------------------------

Inputs:

- Trimmed sRNAseq FastQs generated by
  [`08-Pmea-sRNAseq-trimming.Rmd`](https://github.com/urol-e5/deep-dive/blob/main/F-Pmea/code/08-Pmea-sRNAseq-trimming.Rmd)

  - Filenames formatted: `*flexbar_trim.25bp*.gz`

Outputs:

- `mirtrace.config`: A
  [miRTrace](https://github.com/friedlanderlab/mirtrace) config file. A
  comma-separated file with this layout (one FastQ per line):
  `/path/to/fastq,custom_sample_name`

- “Collapsed” (i.e. unique sequences only) FastA for each corresponding
  input FastQ.

- `mirtrace-report.html`: HTML-formatted report generated by
  [miRTrace](https://github.com/friedlanderlab/mirtrace).

- `mirtrace-stats-contamination_basic.tsv`: Tab-delimited report with
  counts of sequences from each collapsed FastAs having matches to known
  miRNAs within each of the
  [miRTrace](https://github.com/friedlanderlab/mirtrace) Clades.

- `mirtrace-stats-contamination_detailed.csv`: Tab-delimited report of
  *only* Clades with which sequences were matched, along with the
  corresponding miRNA families in each clade, and the sequence counts.

# 1 Create a Bash variables file

This allows usage of Bash variables across R Markdown chunks.

``` bash
{
echo "#### Assign Variables ####"
echo ""

echo "# Data directories"
echo 'export deep_dive_dir=/home/shared/8TB_HDD_02/shedurkin/deep-dive'
echo 'export output_dir_top=${deep_dive_dir}/F-Pmea/output/09-Pmea-sRNAseq-miRTrace'
echo 'export trimmed_reads_dir=${deep_dive_dir}/F-Pmea/output/08-Pmea-sRNAseq-trimming/trimmed-reads'
echo ""

echo "# Paths to programs"
echo 'export mirtrace=/home/sam/programs/mambaforge/envs/miRTrace_env/bin/mirtrace'
echo ""

echo "# Set number of CPUs to use"
echo 'export threads=40'
echo ""

echo "export fastq_pattern='*flexbar_trim.25bp*.gz'"

echo "# Programs associative array"
echo "declare -A programs_array"
echo "programs_array=("
echo '[mirtrace]="${mirtrace}"'
echo ")"
} > .bashvars

cat .bashvars
```

    #### Assign Variables ####

    # Data directories
    export deep_dive_dir=/home/shared/8TB_HDD_02/shedurkin/deep-dive
    export output_dir_top=${deep_dive_dir}/F-Pmea/output/09-Pmea-sRNAseq-miRTrace
    export trimmed_reads_dir=${deep_dive_dir}/F-Pmea/output/08-Pmea-sRNAseq-trimming/trimmed-reads

    # Paths to programs
    export mirtrace=/home/sam/programs/mambaforge/envs/miRTrace_env/bin/mirtrace

    # Set number of CPUs to use
    export threads=40

    export fastq_pattern='*flexbar_trim.25bp*.gz'
    # Programs associative array
    declare -A programs_array
    programs_array=(
    [mirtrace]="${mirtrace}"
    )

# 2 Create [miRTrace](https://github.com/friedlanderlab/mirtrace) config file

``` bash
# Load bash variables into memory
source .bashvars

# Declare array
fastq_array=()

# Populate array
fastq_array=(${trimmed_reads_dir}/${fastq_pattern})

# Loop through read pairs
# Increment by 2 to process next pair of FastQ files
if [ -f "${output_dir_top}/mirtrace.config" ]; then
  echo "mirtrace.config already exists. Nothing to do."
  
else

  for (( i=0; i<${#fastq_array[@]} ; i+=2 ))
  do
    # Use first three parts of filename to create short sample name
    R1_name=$(echo "${fastq_array[i]##*/}" | awk -F "-" '{print $1"-"$2"-"$3}')
    R2_name=$(echo "${fastq_array[i+1]##*/}" | awk -F "-" '{print $1"-"$2"-"$3}')
    echo "${fastq_array[i]},${R1_name}_1"
    echo "${fastq_array[i+1]},${R2_name}_2"
  done >> "${output_dir_top}/mirtrace.config"

fi

cat "${output_dir_top}/mirtrace.config"
```

    mirtrace.config already exists. Nothing to do.
    /home/shared/8TB_HDD_02/shedurkin/deep-dive/F-Pmea/output/08-Pmea-sRNAseq-trimming/trimmed-reads/sRNA-POC-47-S1-TP2.flexbar_trim.25bp_1.fastq.gz,sRNA-POC-47_1
    /home/shared/8TB_HDD_02/shedurkin/deep-dive/F-Pmea/output/08-Pmea-sRNAseq-trimming/trimmed-reads/sRNA-POC-47-S1-TP2.flexbar_trim.25bp_2.fastq.gz,sRNA-POC-47_2
    /home/shared/8TB_HDD_02/shedurkin/deep-dive/F-Pmea/output/08-Pmea-sRNAseq-trimming/trimmed-reads/sRNA-POC-48-S1-TP2.flexbar_trim.25bp_1.fastq.gz,sRNA-POC-48_1
    /home/shared/8TB_HDD_02/shedurkin/deep-dive/F-Pmea/output/08-Pmea-sRNAseq-trimming/trimmed-reads/sRNA-POC-48-S1-TP2.flexbar_trim.25bp_2.fastq.gz,sRNA-POC-48_2
    /home/shared/8TB_HDD_02/shedurkin/deep-dive/F-Pmea/output/08-Pmea-sRNAseq-trimming/trimmed-reads/sRNA-POC-50-S1-TP2.flexbar_trim.25bp_1.fastq.gz,sRNA-POC-50_1
    /home/shared/8TB_HDD_02/shedurkin/deep-dive/F-Pmea/output/08-Pmea-sRNAseq-trimming/trimmed-reads/sRNA-POC-50-S1-TP2.flexbar_trim.25bp_2.fastq.gz,sRNA-POC-50_2
    /home/shared/8TB_HDD_02/shedurkin/deep-dive/F-Pmea/output/08-Pmea-sRNAseq-trimming/trimmed-reads/sRNA-POC-53-S1-TP2.flexbar_trim.25bp_1.fastq.gz,sRNA-POC-53_1
    /home/shared/8TB_HDD_02/shedurkin/deep-dive/F-Pmea/output/08-Pmea-sRNAseq-trimming/trimmed-reads/sRNA-POC-53-S1-TP2.flexbar_trim.25bp_2.fastq.gz,sRNA-POC-53_2
    /home/shared/8TB_HDD_02/shedurkin/deep-dive/F-Pmea/output/08-Pmea-sRNAseq-trimming/trimmed-reads/sRNA-POC-57-S1-TP2.flexbar_trim.25bp_1.fastq.gz,sRNA-POC-57_1
    /home/shared/8TB_HDD_02/shedurkin/deep-dive/F-Pmea/output/08-Pmea-sRNAseq-trimming/trimmed-reads/sRNA-POC-57-S1-TP2.flexbar_trim.25bp_2.fastq.gz,sRNA-POC-57_2

# 3 Run [miRTrace](https://github.com/friedlanderlab/mirtrace)

``` bash
# Load bash variables into memory
source .bashvars

time \
${programs_array[mirtrace]} trace \
--config ${output_dir_top}/mirtrace.config \
--write-fasta \
--num-threads ${threads} \
--output-dir ${output_dir_top} \
--force

tree -h ${output_dir_top}
```

    miRTrace version 1.0.1 starting. Processing 10 sample(s).
    NOTE: reusing existing output directory, outdated files may be present.

    Run complete. Processed 10 sample(s) in 31 s.

    Reports written to /home/shared/8TB_HDD_02/shedurkin/deep-dive/F-Pmea/output/09-Pmea-sRNAseq-miRTrace/
    For information about citing our paper, run miRTrace in mode "cite".

    real    0m32.248s
    user    3m55.713s
    sys 0m6.702s
    /home/shared/8TB_HDD_02/shedurkin/deep-dive/F-Pmea/output/09-Pmea-sRNAseq-miRTrace
    ├── [1.6K]  mirtrace.config
    ├── [302K]  mirtrace-report.html
    ├── [ 563]  mirtrace-stats-contamination_basic.tsv
    ├── [1.1K]  mirtrace-stats-contamination_detailed.tsv
    └── [4.0K]  qc_passed_reads.all.collapsed
        ├── [ 71M]  sRNA-POC-47-S1-TP2.flexbar_trim.25bp_1.fasta
        ├── [ 89M]  sRNA-POC-47-S1-TP2.flexbar_trim.25bp_2.fasta
        ├── [ 75M]  sRNA-POC-48-S1-TP2.flexbar_trim.25bp_1.fasta
        ├── [ 84M]  sRNA-POC-48-S1-TP2.flexbar_trim.25bp_2.fasta
        ├── [ 68M]  sRNA-POC-50-S1-TP2.flexbar_trim.25bp_1.fasta
        ├── [ 83M]  sRNA-POC-50-S1-TP2.flexbar_trim.25bp_2.fasta
        ├── [ 99M]  sRNA-POC-53-S1-TP2.flexbar_trim.25bp_1.fasta
        ├── [116M]  sRNA-POC-53-S1-TP2.flexbar_trim.25bp_2.fasta
        ├── [120M]  sRNA-POC-57-S1-TP2.flexbar_trim.25bp_1.fasta
        └── [129M]  sRNA-POC-57-S1-TP2.flexbar_trim.25bp_2.fasta

    1 directory, 14 files

# 4 Results

## 4.1 Read in table as data frame

``` r
mirtrace.detailed.df <- read.csv("../output/09-Pmea-sRNAseq-miRTrace/mirtrace-stats-contamination_detailed.tsv", sep = "\t", header = TRUE)

str(mirtrace.detailed.df)
```

    'data.frame':   8 obs. of  14 variables:
     $ CLADE        : chr  "nematode" "insects" "insects" "lophotrochozoa" ...
     $ FAMILY_ID    : int  86 14 957 1996 2689 1994 1998 3122
     $ MIRBASE_IDS  : chr  "asu-miR-86,bma-miR-86,cbn-miR-86,cbr-miR-86,cel-miR-86,crm-miR-86,hco-miR-86,prd-miR-86,str-miR-86" "aae-miR-14,aga-miR-14,ame-miR-14,api-miR-14,bmo-miR-14,cqu-miR-14,dan-miR-14,der-miR-14,dgr-miR-14,dme-miR-14,d"| __truncated__ "aae-miR-957,aga-miR-957,cqu-miR-957,dme-miR-957,dps-miR-957,dsi-miR-957,dvi-miR-957" "cte-miR-1996a" ...
     $ SEQ          : chr  "TAAGTGAATGCTTTGCCACA" "TCAGTCTTTTTCTCTCTCCT" "TGAAACCGTCCAAAACTGAG" "CTCAAGTGAGGTCAGTGCTG" ...
     $ sRNA.POC.47_1: int  0 0 0 0 2 0 1 0
     $ sRNA.POC.47_2: int  0 0 0 0 0 0 0 0
     $ sRNA.POC.48_1: int  0 2 0 0 0 0 0 0
     $ sRNA.POC.48_2: int  0 0 0 0 0 0 0 0
     $ sRNA.POC.50_1: int  0 0 0 0 0 0 0 0
     $ sRNA.POC.50_2: int  0 0 0 0 0 0 0 0
     $ sRNA.POC.53_1: int  0 0 1 1 0 0 0 1
     $ sRNA.POC.53_2: int  0 0 0 0 0 0 0 0
     $ sRNA.POC.57_1: int  3 2 0 0 0 1 0 0
     $ sRNA.POC.57_2: int  0 0 0 0 0 0 0 0

## 4.2 Number of samples with matches

``` r
# Select columns corresponding to sample names
sample_columns <- mirtrace.detailed.df %>%
  select(starts_with("sRNA.POC."))

# Calculate the sum for each column
sample_sums <- colSums(sample_columns)

# Count the number of columns with a sum greater than 0
samples_with_sum_gt_0 <- sum(sample_sums > 0)

paste("Number of samples with matches: ", samples_with_sum_gt_0)
```

    [1] "Number of samples with matches:  4"

## 4.3 Percentage of samples with matches

``` r
# Total number of samples (columns)
total_samples <- ncol(sample_columns)

# Percentage of samples with sums greater than 0
percentage_samples_gt_0 <- (samples_with_sum_gt_0 / total_samples) * 100

paste("Percentage of samples with matches: ", percentage_samples_gt_0)
```

    [1] "Percentage of samples with matches:  40"

## 4.4 Number of clades with matches

``` r
unique_clade_count <- mirtrace.detailed.df %>%
  distinct(CLADE) %>%    # Get unique entries in CLADE column
  count()               # Count the number of unique entries



paste("Number of clades with matches:", unique_clade_count)
```

    [1] "Number of clades with matches: 4"

## 4.5 [miRTrace](https://github.com/friedlanderlab/mirtrace) table

To make them easier to see, counts \> 0 are highlighted in green.

``` r
mirtrace.detailed.df %>%
  mutate(
    across(
      starts_with("sRNA"),
      ~cell_spec(
        .,
        background = ifelse(
          . > 0,
          "lightgreen",
          "white"
          )
        )
      )
    ) %>%
  kable(escape = F, caption = "Clades identified as having sRNAseq matches.") %>%
  kable_styling("striped") %>% 
  scroll_box(width = "100%", height = "500px")
```

<div
style="border: 1px solid #ddd; padding: 0px; overflow-y: scroll; height:500px; overflow-x: scroll; width:100%; ">

<table class="table table-striped" style="margin-left: auto; margin-right: auto;">
<caption>

Clades identified as having sRNAseq matches.

</caption>
<thead>
<tr>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">

CLADE

</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">

FAMILY_ID

</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">

MIRBASE_IDS

</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">

SEQ

</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">

sRNA.POC.47_1

</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">

sRNA.POC.47_2

</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">

sRNA.POC.48_1

</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">

sRNA.POC.48_2

</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">

sRNA.POC.50_1

</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">

sRNA.POC.50_2

</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">

sRNA.POC.53_1

</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">

sRNA.POC.53_2

</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">

sRNA.POC.57_1

</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">

sRNA.POC.57_2

</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">

nematode

</td>
<td style="text-align:right;">

86

</td>
<td style="text-align:left;">

asu-miR-86,bma-miR-86,cbn-miR-86,cbr-miR-86,cel-miR-86,crm-miR-86,hco-miR-86,prd-miR-86,str-miR-86

</td>
<td style="text-align:left;">

TAAGTGAATGCTTTGCCACA

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[3\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: lightgreen !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
</tr>
<tr>
<td style="text-align:left;">

insects

</td>
<td style="text-align:right;">

14

</td>
<td style="text-align:left;">

aae-miR-14,aga-miR-14,ame-miR-14,api-miR-14,bmo-miR-14,cqu-miR-14,dan-miR-14,der-miR-14,dgr-miR-14,dme-miR-14,dmo-miR-14,dpe-miR-14,dps-miR-14,dse-miR-14,dsi-miR-14,dvi-miR-14,dwi-miR-14,dya-miR-14,hme-miR-14,mse-miR-14,ngi-miR-14,nvi-miR-14,tca-miR-14

</td>
<td style="text-align:left;">

TCAGTCTTTTTCTCTCTCCT

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[2\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: lightgreen !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[2\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: lightgreen !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
</tr>
<tr>
<td style="text-align:left;">

insects

</td>
<td style="text-align:right;">

957

</td>
<td style="text-align:left;">

aae-miR-957,aga-miR-957,cqu-miR-957,dme-miR-957,dps-miR-957,dsi-miR-957,dvi-miR-957

</td>
<td style="text-align:left;">

TGAAACCGTCCAAAACTGAG

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[1\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: lightgreen !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
</tr>
<tr>
<td style="text-align:left;">

lophotrochozoa

</td>
<td style="text-align:right;">

1996

</td>
<td style="text-align:left;">

cte-miR-1996a

</td>
<td style="text-align:left;">

CTCAAGTGAGGTCAGTGCTG

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[1\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: lightgreen !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
</tr>
<tr>
<td style="text-align:left;">

lophotrochozoa

</td>
<td style="text-align:right;">

2689

</td>
<td style="text-align:left;">

cte-miR-2689

</td>
<td style="text-align:left;">

TATCCTGGCCTGCAAGTGCA

</td>
<td style="text-align:left;">

\[2\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: lightgreen !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
</tr>
<tr>
<td style="text-align:left;">

lophotrochozoa

</td>
<td style="text-align:right;">

1994

</td>
<td style="text-align:left;">

cla-miR-1994,cte-miR-1994,lgi-miR-1994a,lgi-miR-1994b

</td>
<td style="text-align:left;">

TGAGACAGTGTGTCCTCCCT

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[1\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: lightgreen !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
</tr>
<tr>
<td style="text-align:left;">

lophotrochozoa

</td>
<td style="text-align:right;">

1998

</td>
<td style="text-align:left;">

cte-miR-1998

</td>
<td style="text-align:left;">

TTGAACGCAGAGATGTACAT

</td>
<td style="text-align:left;">

\[1\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: lightgreen !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
</tr>
<tr>
<td style="text-align:left;">

primates

</td>
<td style="text-align:right;">

3122

</td>
<td style="text-align:left;">

hsa-miR-3122

</td>
<td style="text-align:left;">

GTTGGGACAAGAGGACGGTC

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[1\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: lightgreen !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
<td style="text-align:left;">

\[0\]{style=” border-radius: 4px; padding-right: 4px; padding-left: 4px;
background-color: white !important;“}

</td>
</tr>
</tbody>
</table>

</div>

# 5 Citations
