10.1-Pmea-sRNAseq-BLASTn-31bp-fastp-merged-cnidarian_miRBase
================
Sam White (modified by K Durkin for P. meandrina analysis)
2024-04-19

- <a href="#1-create-a-bash-variables-file"
  id="toc-1-create-a-bash-variables-file">1 Create a Bash variables
  file</a>
  - <a href="#11-inspect-mirna-fastas" id="toc-11-inspect-mirna-fastas">1.1
    Inspect miRNA FastAs</a>
- <a href="#2-convert-u-to-t-in-mirna-fastas"
  id="toc-2-convert-u-to-t-in-mirna-fastas">2 Convert <code>U</code> to
  <code>T</code> in miRNA FastAs</a>
- <a href="#3-create-blast-databases" id="toc-3-create-blast-databases">3
  Create BLAST Databases</a>
- <a href="#4-prepare-reads-for-blasting"
  id="toc-4-prepare-reads-for-blasting">4 Prepare reads for BLASTing</a>
  - <a href="#41-concatenate-all-trimmed-reads"
    id="toc-41-concatenate-all-trimmed-reads">4.1 Concatenate all trimmed
    reads</a>
  - <a href="#42-collapse-reads-to-fasta"
    id="toc-42-collapse-reads-to-fasta">4.2 Collapse reads to FastA</a>
- <a href="#5-run-blastn-default-e-value"
  id="toc-5-run-blastn-default-e-value">5 Run BLASTn Default E-value</a>
  - <a href="#51-cnidarian-mirbase-blastn-default-e-value"
    id="toc-51-cnidarian-mirbase-blastn-default-e-value">5.1 cnidarian
    miRBase BLASTn Default e-value</a>
- <a href="#6-blastn-e-value--10" id="toc-6-blastn-e-value--10">6 BLASTn
  E-value = 10</a>
  - <a href="#61-cnidarian-mirbase-blastn-e-value--10"
    id="toc-61-cnidarian-mirbase-blastn-e-value--10">6.1 cnidarian miRBase
    BLASTn e-value = 10</a>
- <a href="#7-results" id="toc-7-results">7 Results</a>
  - <a href="#71-check-blastn-default-e-value-results"
    id="toc-71-check-blastn-default-e-value-results">7.1 Check BLASTn
    Default e-value results</a>
  - <a href="#72-check-blastn-e-value--10-results"
    id="toc-72-check-blastn-e-value--10-results">7.2 Check BLASTn e-value =
    10 results</a>
- <a href="#8-citations" id="toc-8-citations">8 Citations</a>

This notebook performs a simple [NCBI
BLASTn](https://www.ncbi.nlm.nih.gov/books/NBK279690/) ([Altschul et al.
1990](#ref-altschul1990)) against an miRNA database to attempt to
identify miRNA in *P.meandrina* sRNAseq:

- [miRBase](https://mirbase.org/download/)
  - Utilizes a modified version, which includes cnidarian miRNA culled
    from literature by Jill Ahsley.

  - [`cnidarian-mirbase-mature-v22.1.fasta`](../../data/cnidarian-mirbase-mature-v22.1.fasta)

Relies on the following software:

- [fastx_toolkit](http://hannonlab.cshl.edu/fastx_toolkit/)

  - `fastx_collapser`: Collapses duplicate sequences in FastA/Q into
    single sequence.

# 1 Create a Bash variables file

This allows usage of Bash variables across R Markdown chunks.

``` bash
{
echo "#### Assign Variables ####"
echo ""

echo "# Trimmed FastQ naming pattern"
echo "export trimmed_fastqs_pattern='*31bp-merged.fq'"

echo "# Data directories"
echo 'export deep_dive_dir=/home/shared/8TB_HDD_02/shedurkin/deep-dive'
echo 'export deep_dive_data_dir="${deep_dive_dir}/data"'
echo 'export output_dir_top=${deep_dive_dir}/F-Pmea/output/10.1-Pmea-sRNAseq-BLASTn-31bp-fastp-merged-cnidarian_miRBase'
echo 'export trimmed_fastqs_dir="${deep_dive_dir}/F-Pmea/output/08.2-Pmea-sRNAseq-trimming-31bp-fastp-merged/trimmed-reads"'
echo 'export blast_dbs_dir="${deep_dive_dir}/data/blast_dbs"'
echo ""

echo "# Input/Output files"
echo 'export collapsed_reads_fasta="collapsed-reads-all.fasta"'
echo 'export concatenated_trimmed_reads_fastq="concatenated-trimmed-reads-all.fastq.gz"'
echo 'export mirbase_mature_fasta_name="cnidarian-mirbase-mature-v22.1.fasta"'
echo 'export mirbase_mature_fasta_no_U="cnidarian-mirbase-mature-v22.1-no_U.fa"'
echo ""

echo "# External data URLs"
echo 'export mirgenedb_fasta_url="https://www.mirgenedb.org/fasta/ALL?mat=1"'
echo ""

echo "# Paths to programs"
echo 'export ncbi_blast_dir="/home/shared/ncbi-blast-2.15.0+/bin/"'
echo 'export ncbi_blastn="${ncbi_blast_dir}/blastn"'
echo 'export ncbi_makeblast_db="${ncbi_blast_dir}/makeblastdb"'
echo 'export fastx_collapser="/home/shared/fastx_toolkit_0.0.13_binaries_Linux_2.6_amd64/bin/fastx_collapser"'

echo "# Set number of CPUs to use"
echo 'export threads=46'
echo ""

echo "# Initialize arrays"
echo 'export trimmed_fastqs_array=()'


} > .bashvars

cat .bashvars
```

    #### Assign Variables ####

    # Trimmed FastQ naming pattern
    export trimmed_fastqs_pattern='*31bp-merged.fq'
    # Data directories
    export deep_dive_dir=/home/shared/8TB_HDD_02/shedurkin/deep-dive
    export deep_dive_data_dir="${deep_dive_dir}/data"
    export output_dir_top=${deep_dive_dir}/F-Pmea/output/10.1-Pmea-sRNAseq-BLASTn-31bp-fastp-merged-cnidarian_miRBase
    export trimmed_fastqs_dir="${deep_dive_dir}/F-Pmea/output/08.2-Pmea-sRNAseq-trimming-31bp-fastp-merged/trimmed-reads"
    export blast_dbs_dir="${deep_dive_dir}/data/blast_dbs"

    # Input/Output files
    export collapsed_reads_fasta="collapsed-reads-all.fasta"
    export concatenated_trimmed_reads_fastq="concatenated-trimmed-reads-all.fastq.gz"
    export mirbase_mature_fasta_name="cnidarian-mirbase-mature-v22.1.fasta"
    export mirbase_mature_fasta_no_U="cnidarian-mirbase-mature-v22.1-no_U.fa"

    # External data URLs
    export mirgenedb_fasta_url="https://www.mirgenedb.org/fasta/ALL?mat=1"

    # Paths to programs
    export ncbi_blast_dir="/home/shared/ncbi-blast-2.15.0+/bin/"
    export ncbi_blastn="${ncbi_blast_dir}/blastn"
    export ncbi_makeblast_db="${ncbi_blast_dir}/makeblastdb"
    export fastx_collapser="/home/shared/fastx_toolkit_0.0.13_binaries_Linux_2.6_amd64/bin/fastx_collapser"
    # Set number of CPUs to use
    export threads=46

    # Initialize arrays
    export trimmed_fastqs_array=()

## 1.1 Inspect miRNA FastAs

``` bash
# Load bash variables into memory
source .bashvars

head "${deep_dive_data_dir}"/cni*.fasta
```

    ==> /home/shared/8TB_HDD_02/shedurkin/deep-dive/data/cnidarian-mirbase-mature-v22.1.fasta <==
    >cel-let-7-5p MIMAT0000001 Caenorhabditis elegans let-7-5p
    UGAGGUAGUAGGUUGUAUAGUU
    >cel-let-7-3p MIMAT0015091 Caenorhabditis elegans let-7-3p
    CUAUGCAAUUUUCUACCUUACC
    >cel-lin-4-5p MIMAT0000002 Caenorhabditis elegans lin-4-5p
    UCCCUGAGACCUCAAGUGUGA
    >cel-lin-4-3p MIMAT0015092 Caenorhabditis elegans lin-4-3p
    ACACCUGGGCUCUCCGGGUACC
    >cel-miR-1-5p MIMAT0020301 Caenorhabditis elegans miR-1-5p
    CAUACUUCCUUACAUGCCCAUA

    ==> /home/shared/8TB_HDD_02/shedurkin/deep-dive/data/cnidarian_miRNAs.fasta <==
    >spi-mir-temp-1_Stylophora_pistillata_Liew_et_al._2014_Matches_miR-100_family.
    ACCCGUAGAUCCGAACUUGUGG
    >spi-mir-temp-2_Stylophora_pistillata_Liew_et_al._2014_NA
    UAUCGAAUCCGUCAAAAAGAGA
    >spi-mir-temp-3_Stylophora_pistillata_Liew_et_al._2014_NA
    UCAGGGAUUGUGGUGAGUUAGUU
    >spi-mir-temp-4_Stylophora_pistillata_Liew_et_al._2014_Exact_match_of_nve-miR-2023.
    AAAGAAGUACAAGUGGUAGGG
    >spi-mir-temp-5_Stylophora_pistillata_Liew_et_al._2014_NA
    GAGGUCCGGAUGGUUGA

# 2 Convert `U` to `T` in miRNA FastAs

This is needed because the sRNAseq sequences do *not* have uracils
(`U`) - they have thymines (`T`).

``` bash
# Load bash variables into memory
source .bashvars

# Convert miRBase FastA
sed '/^[^>]/s/U/T/g' "${deep_dive_data_dir}/${mirbase_mature_fasta_name}" \
> "${deep_dive_data_dir}/${mirbase_mature_fasta_no_U}"

# Convert MirGene FastA
sed '/^[^>]/s/U/T/g' "${deep_dive_data_dir}/${mirgene_mature_fasta_name}" \
> "${deep_dive_data_dir}/${mirgene_mature_fasta_no_U}"

head ${deep_dive_data_dir}/*.fa
  
```

    bash: line 8: /home/shared/8TB_HDD_02/shedurkin/deep-dive/data/: Is a directory
    ==> /home/shared/8TB_HDD_02/shedurkin/deep-dive/data/cnidarian-mirbase-mature-v22.1-no_U.fa <==
    >cel-let-7-5p MIMAT0000001 Caenorhabditis elegans let-7-5p
    TGAGGTAGTAGGTTGTATAGTT
    >cel-let-7-3p MIMAT0015091 Caenorhabditis elegans let-7-3p
    CTATGCAATTTTCTACCTTACC
    >cel-lin-4-5p MIMAT0000002 Caenorhabditis elegans lin-4-5p
    TCCCTGAGACCTCAAGTGTGA
    >cel-lin-4-3p MIMAT0015092 Caenorhabditis elegans lin-4-3p
    ACACCTGGGCTCTCCGGGTACC
    >cel-miR-1-5p MIMAT0020301 Caenorhabditis elegans miR-1-5p
    CATACTTCCTTACATGCCCATA

    ==> /home/shared/8TB_HDD_02/shedurkin/deep-dive/data/mirbase-mature-v22.1.fa <==
    >cel-let-7-5p MIMAT0000001 Caenorhabditis elegans let-7-5p
    UGAGGUAGUAGGUUGUAUAGUU
    >cel-let-7-3p MIMAT0015091 Caenorhabditis elegans let-7-3p
    CUAUGCAAUUUUCUACCUUACC
    >cel-lin-4-5p MIMAT0000002 Caenorhabditis elegans lin-4-5p
    UCCCUGAGACCUCAAGUGUGA
    >cel-lin-4-3p MIMAT0015092 Caenorhabditis elegans lin-4-3p
    ACACCUGGGCUCUCCGGGUACC
    >cel-miR-1-5p MIMAT0020301 Caenorhabditis elegans miR-1-5p
    CAUACUUCCUUACAUGCCCAUA

    ==> /home/shared/8TB_HDD_02/shedurkin/deep-dive/data/mirbase-mature-v22.1-no_spaces.fa <==
    >cel-let-7-5p_MIMAT0000001_Caenorhabditis_elegans_let-7-5p
    UGAGGUAGUAGGUUGUAUAGUU
    >cel-let-7-3p_MIMAT0015091_Caenorhabditis_elegans_let-7-3p
    CUAUGCAAUUUUCUACCUUACC
    >cel-lin-4-5p_MIMAT0000002_Caenorhabditis_elegans_lin-4-5p
    UCCCUGAGACCUCAAGUGUGA
    >cel-lin-4-3p_MIMAT0015092_Caenorhabditis_elegans_lin-4-3p
    ACACCUGGGCUCUCCGGGUACC
    >cel-miR-1-5p_MIMAT0020301_Caenorhabditis_elegans_miR-1-5p
    CAUACUUCCUUACAUGCCCAUA

    ==> /home/shared/8TB_HDD_02/shedurkin/deep-dive/data/mirbase-mature-v22.1-no_U.fa <==
    >cel-let-7-5p MIMAT0000001 Caenorhabditis elegans let-7-5p
    TGAGGTAGTAGGTTGTATAGTT
    >cel-let-7-3p MIMAT0015091 Caenorhabditis elegans let-7-3p
    CTATGCAATTTTCTACCTTACC
    >cel-lin-4-5p MIMAT0000002 Caenorhabditis elegans lin-4-5p
    TCCCTGAGACCTCAAGTGTGA
    >cel-lin-4-3p MIMAT0015092 Caenorhabditis elegans lin-4-3p
    ACACCTGGGCTCTCCGGGTACC
    >cel-miR-1-5p MIMAT0020301 Caenorhabditis elegans miR-1-5p
    CATACTTCCTTACATGCCCATA

    ==> /home/shared/8TB_HDD_02/shedurkin/deep-dive/data/mirgene-mature-all-v2.1.fa <==
    >Aae-Bantam_3p
    UGAGAUCAUUUUGAAAGCUGAUU
    >Bge-Bantam_3p
    UGAGAUCAUUGUGAAAGCUGAUU
    >Bpl-Bantam_3p
    UGAGAUCAUUGUGAAAACUGAU
    >Cgi-Bantam_3p
    UGAGAUCAUUGUGAAAACUGAUU
    >Cte-Bantam_3p
    UGAGAUCAUUGUGAAAACUAAUC

    ==> /home/shared/8TB_HDD_02/shedurkin/deep-dive/data/mirgene-mature-all-v2.1-no_U.fa <==
    >Aae-Bantam_3p
    TGAGATCATTTTGAAAGCTGATT
    >Bge-Bantam_3p
    TGAGATCATTGTGAAAGCTGATT
    >Bpl-Bantam_3p
    TGAGATCATTGTGAAAACTGAT
    >Cgi-Bantam_3p
    TGAGATCATTGTGAAAACTGATT
    >Cte-Bantam_3p
    TGAGATCATTGTGAAAACTAATC

# 3 Create BLAST Databases

``` bash
# Load bash variables into memory
source .bashvars

# miRBase BLAST DB
## Make sure output directory exists
if [ ! -d "${blast_dbs_dir}" ]; then
  mkdir --parents "${blast_dbs_dir}"
fi

## Check for pre-exising database
if [ ! -f "${blast_dbs_dir}/${mirbase_mature_fasta_no_U%.*}.blastdb.log" ]; then
  ${ncbi_makeblast_db} \
  -in ${deep_dive_data_dir}/${mirbase_mature_fasta_no_U} \
  -title ${mirbase_mature_fasta_no_U%.*} \
  -dbtype nucl \
  -out ${blast_dbs_dir}/${mirbase_mature_fasta_no_U%.*} \
  -logfile ${blast_dbs_dir}/${mirbase_mature_fasta_no_U%.*}.blastdb.log
fi
```

# 4 Prepare reads for BLASTing

## 4.1 Concatenate all trimmed reads

``` bash
# Load bash variables into memory
source .bashvars

# Check for existence of concatenated FastA before running
if [ ! -f "${output_dir_top}/${concatenated_trimmed_reads_fastq}" ]; then
  cat ${trimmed_fastqs_dir}/${trimmed_fastqs_pattern}.gz \
  > "${output_dir_top}/${concatenated_trimmed_reads_fastq}"
fi

ls -lh "${output_dir_top}/${concatenated_trimmed_reads_fastq}"
```

    -rw-r--r-- 1 shedurkin labmembers 1.1G Apr 19 12:40 /home/shared/8TB_HDD_02/shedurkin/deep-dive/F-Pmea/output/10.1-Pmea-sRNAseq-BLASTn-31bp-fastp-merged-cnidarian_miRBase/concatenated-trimmed-reads-all.fastq.gz

## 4.2 Collapse reads to FastA

Uses `fastx_collapser` to collapse to unique reads.

Requires undocumented quality setting. Have selected `30` as cuttoff:
`-Q30`.

``` bash
# Load bash variables into memory
source .bashvars

# Check for existence of collapsed FastA before running
time \
if [ ! -f "${output_dir_top}/${collapsed_reads_fasta}" ]; then
  zcat ${output_dir_top}/${concatenated_trimmed_reads_fastq} \
  | ${fastx_collapser} \
  -Q30 \
  -o "${output_dir_top}/${collapsed_reads_fasta}"
fi

head "${output_dir_top}/${collapsed_reads_fasta}"
```

    real    0m0.000s
    user    0m0.000s
    sys 0m0.000s
    >1-10257663
    TGAAAATCTTTGCTCTGAAGTGGAA
    >2-2952210
    CTCGGGCTGAGACTTGAAGCG
    >3-1527059
    GCACTGGTGGTTCAGTGGTAGAATTCTC
    >4-1205244
    GCACTGGTGGTTCAGTGGTAGAATTCTCGCC
    >5-1189262
    CTCGGGCTGAGACTTGAAGCA

# 5 Run BLASTn Default E-value

- 1000 for blastn-short

Runs BLASTn using the `blastn-short` task for sequences \< 30bp.

Look for top match (`-max_hsps 1` & `-max_target_seqs 1`) for each
query.

- Suppress subsequent warning
  `Examining 5 or more matches is recommended` by redirecting stdout:
  `2> /dev/null`

## 5.1 cnidarian miRBase BLASTn Default e-value

``` bash
# Load bash variables into memory
source .bashvars

time \
${ncbi_blastn} \
-db ${blast_dbs_dir}/${mirbase_mature_fasta_no_U%.*} \
-query ${output_dir_top}/${collapsed_reads_fasta} \
-out ${output_dir_top}/miRBase-BLASTn-eval_1000.outfmt6 \
-task blastn-short \
-max_hsps 1 \
-max_target_seqs 1 \
-outfmt 6 \
-num_threads ${threads} \
2> /dev/null
```

# 6 BLASTn E-value = 10

Running this for simple comparison to the defaul `blastn-short` value of
1000.

## 6.1 cnidarian miRBase BLASTn e-value = 10

``` bash
# Load bash variables into memory
source .bashvars

time \
${ncbi_blastn} \
-db ${blast_dbs_dir}/${mirbase_mature_fasta_no_U%.*} \
-query ${output_dir_top}/${collapsed_reads_fasta} \
-out ${output_dir_top}/miRBase-BLASTn-eval_10.outfmt6 \
-task blastn-short \
-evalue 10 \
-max_hsps 1 \
-max_target_seqs 1 \
-outfmt 6 \
-num_threads ${threads} \
2> /dev/null
```

# 7 Results

## 7.1 Check BLASTn Default e-value results

``` bash
# Load bash variables into memory
source .bashvars

head ${output_dir_top}/*eval_1000.outfmt6

echo ""
echo "-----------------------------------------"
echo ""

wc -l ${output_dir_top}/*eval_1000.outfmt6
```

    1-10257663  ssc-miR-7143-5p 100.000 11  0   0   12  22  8   18  1.5 22.3
    2-2952210   cpo-miR-1379-5p 100.000 12  0   0   1   12  18  7   0.26    24.3
    3-1527059   ppc-miR-8214-5p 100.000 13  0   0   12  24  18  6   0.11    26.3
    4-1205244   ppc-miR-8214-5p 100.000 13  0   0   12  24  18  6   0.11    26.3
    5-1189262   cpo-miR-1379-5p 100.000 12  0   0   1   12  18  7   0.26    24.3
    6-630167    cpo-miR-509c-3p 100.000 12  0   0   13  24  8   19  0.37    24.3
    7-548099    ppc-miR-8214-5p 100.000 13  0   0   12  24  18  6   0.11    26.3
    8-512116    hpo-miR-10025-5p    100.000 10  0   0   11  20  17  8   4.9 20.3
    9-432671    cpo-miR-1379-5p 100.000 12  0   0   1   12  18  7   0.34    24.3
    10-407718   ppc-miR-8214-5p 100.000 13  0   0   12  24  18  6   0.11    26.3

    -----------------------------------------

    6393700 /home/shared/8TB_HDD_02/shedurkin/deep-dive/F-Pmea/output/10.1-Pmea-sRNAseq-BLASTn-31bp-fastp-merged-cnidarian_miRBase/miRBase-BLASTn-eval_1000.outfmt6

## 7.2 Check BLASTn e-value = 10 results

``` bash
# Load bash variables into memory
source .bashvars

head ${output_dir_top}/*eval_10.outfmt6

echo ""
echo "-----------------------------------------"
echo ""

wc -l ${output_dir_top}/*eval_10.outfmt6
```

    1-10257663  ssc-miR-7143-5p 100.000 11  0   0   12  22  8   18  1.5 22.3
    2-2952210   cpo-miR-1379-5p 100.000 12  0   0   1   12  18  7   0.26    24.3
    3-1527059   ppc-miR-8214-5p 100.000 13  0   0   12  24  18  6   0.11    26.3
    4-1205244   ppc-miR-8214-5p 100.000 13  0   0   12  24  18  6   0.11    26.3
    5-1189262   cpo-miR-1379-5p 100.000 12  0   0   1   12  18  7   0.26    24.3
    6-630167    cpo-miR-509c-3p 100.000 12  0   0   13  24  8   19  0.37    24.3
    7-548099    ppc-miR-8214-5p 100.000 13  0   0   12  24  18  6   0.11    26.3
    8-512116    hpo-miR-10025-5p    100.000 10  0   0   11  20  17  8   4.9 20.3
    9-432671    cpo-miR-1379-5p 100.000 12  0   0   1   12  18  7   0.34    24.3
    10-407718   ppc-miR-8214-5p 100.000 13  0   0   12  24  18  6   0.11    26.3

    -----------------------------------------

    6367488 /home/shared/8TB_HDD_02/shedurkin/deep-dive/F-Pmea/output/10.1-Pmea-sRNAseq-BLASTn-31bp-fastp-merged-cnidarian_miRBase/miRBase-BLASTn-eval_10.outfmt6

The default e-value results in alignments for all query sequences, which
is likely not what we’d expect.

Decreasing the e-value resulted in fewer queary alignments, as we’d
expect.

For further analysis, we should probably discuss a reasonable e-value to
use for filtering.

------------------------------------------------------------------------

# 8 Citations

<div id="refs" class="references csl-bib-body hanging-indent">

<div id="ref-altschul1990" class="csl-entry">

Altschul, Stephen F., Warren Gish, Webb Miller, Eugene W. Myers, and
David J. Lipman. 1990. “Basic Local Alignment Search Tool.” *Journal of
Molecular Biology* 215 (3): 403–10.
<https://doi.org/10.1016/s0022-2836(05)80360-2>.

</div>

</div>
