13-Pmea-sRNAseq-ShortStack
================
Sam White (modified by K Durkin for P. meandrina analysis)
2023-11-30

- <a href="#1-set-r-variables" id="toc-1-set-r-variables">1 Set R
  variables</a>
- <a href="#2-create-a-bash-variables-file"
  id="toc-2-create-a-bash-variables-file">2 Create a Bash variables
  file</a>
- <a href="#3-load-shortstack-conda-environment"
  id="toc-3-load-shortstack-conda-environment">3 Load ShortStack conda
  environment</a>
- <a href="#4-download-mirbase-mature-mirna-fasta"
  id="toc-4-download-mirbase-mature-mirna-fasta">4 Download miRBase mature
  miRNA FastA</a>
- <a href="#5-run-shortstack" id="toc-5-run-shortstack">5 Run
  ShortStack</a>
  - <a href="#51-modify-genome-filename-for-shortstack-compatability"
    id="toc-51-modify-genome-filename-for-shortstack-compatability">5.1
    Modify genome filename for ShortStack compatability</a>
  - <a href="#52-excecute-shortstack-command"
    id="toc-52-excecute-shortstack-command">5.2 Excecute ShortStack
    command</a>
  - <a href="#53-check-runtime" id="toc-53-check-runtime">5.3 Check
    runtime</a>
- <a href="#6-results" id="toc-6-results">6 Results</a>
  - <a href="#61-shortstack-synopsis" id="toc-61-shortstack-synopsis">6.1
    ShortStack synopsis</a>
  - <a href="#62-inspect-resultstxt" id="toc-62-inspect-resultstxt">6.2
    Inspect <code>Results.txt</code></a>
    - <a href="#621-directory-tree-of-all-shortstack-outputs"
      id="toc-621-directory-tree-of-all-shortstack-outputs">6.2.1 Directory
      tree of all ShortStack outputs</a>
- <a href="#7-citations" id="toc-7-citations">7 Citations</a>

Use [ShortStack](https://github.com/MikeAxtell/ShortStack) ([Axtell
2013](#ref-axtell2013a); [Shahid and Axtell 2014](#ref-shahid2014);
[Johnson et al. 2016](#ref-johnson2016a))to perform alignment of sRNAseq
data and annotation of sRNA-producing genes.

The *P.meandrina* genome will be used as the reference genome.

------------------------------------------------------------------------

Inputs:

- Requires trimmed sRNAseq files generated by
  [08-Pmea-sRNAseq-trimming.Rmd](https://github.com/urol-e5/deep-dive/blob/main/F-Pmea/code/08-Pmea-sRNAseq-trimming.Rmd)

  - Filenames formatted: `*flexbar_trim.25bp*.gz`

- *P.meandrina* genome FastA. See
  [12-Pmea-sRNAseq-MirMachine.Rmd](https://github.com/urol-e5/deep-dive/blob/main/F-Pmea/code/12-Pmea-sRNAseq-MirMachine.Rmd)
  for download info if needed.

Outputs:

- See [ShortStack outputs
  documentation](https://github.com/MikeAxtell/ShortStack#outputs) for
  full list and detailed descriptions.

Software requirements:

- Utilizes a
  [ShortStack](https://github.com/MikeAxtell/ShortStack#installation)
  Conda/Mamba environment, per the installation instructions.

Replace with name of your ShortStack environment and the path to the
corresponding conda installation (find this *after* you’ve activated the
environment).

E.g.

``` bash
# Activate environment
conda activate ShortStack4_env

# Find conda path
which conda
```

------------------------------------------------------------------------

# 1 Set R variables

``` r
shortstack_conda_env_name <- c("ShortStack4_env")
shortstack_cond_path <- c("/home/sam/programs/mambaforge/condabin/conda")
```

# 2 Create a Bash variables file

This allows usage of Bash variables across R Markdown chunks.

``` bash
{
echo "#### Assign Variables ####"
echo ""

echo "# Trimmed FastQ naming pattern"
echo "export trimmed_fastqs_pattern='*flexbar_trim.25bp*.fastq.gz'"

echo "# Data directories"
echo 'export deep_dive_dir=/home/shared/8TB_HDD_02/shedurkin/deep-dive'
echo 'export deep_dive_data_dir="${deep_dive_dir}/data"'
echo 'export output_dir_top=${deep_dive_dir}/F-Pmea/output/13-Pmea-sRNAseq-ShortStack'
echo 'export trimmed_fastqs_dir="${deep_dive_dir}/F-Pmea/output/08-Pmea-sRNAseq-trimming/trimmed-reads"'
echo ""

echo "# Input/Output files"
echo 'export genome_fasta_dir=${deep_dive_dir}/F-Pmea/data/Pmea'
echo 'export genome_fasta_name="Pocillopora_meandrina_HIv1.assembly.fasta"'
echo 'export shortstack_genome_fasta_name="Pocillopora_meandrina_HIv1.assembly.fa"'
echo 'export mirbase_mature_fasta=mature.fa'
echo 'export mirbase_mature_fasta_version=mirbase-mature-v22.1.fa'
echo 'export genome_fasta="${genome_fasta_dir}/${shortstack_genome_fasta_name}"'
echo ""

echo "# External data URLs"
echo 'export mirbase_fasta_url="https://mirbase.org/download_version_files/22.1/"'
echo ""

echo "# Set number of CPUs to use"
echo 'export threads=46'
echo ""

echo "# Initialize arrays"
echo 'export trimmed_fastqs_array=()'


} > .bashvars

cat .bashvars
```

    #### Assign Variables ####

    # Trimmed FastQ naming pattern
    export trimmed_fastqs_pattern='*flexbar_trim.25bp*.fastq.gz'
    # Data directories
    export deep_dive_dir=/home/shared/8TB_HDD_02/shedurkin/deep-dive
    export deep_dive_data_dir="${deep_dive_dir}/data"
    export output_dir_top=${deep_dive_dir}/F-Pmea/output/13-Pmea-sRNAseq-ShortStack
    export trimmed_fastqs_dir="${deep_dive_dir}/F-Pmea/output/08-Pmea-sRNAseq-trimming/trimmed-reads"

    # Input/Output files
    export genome_fasta_dir=${deep_dive_dir}/F-Pmea/data/Pmea
    export genome_fasta_name="Pocillopora_meandrina_HIv1.assembly.fasta"
    export shortstack_genome_fasta_name="Pocillopora_meandrina_HIv1.assembly.fa"
    export mirbase_mature_fasta=mature.fa
    export mirbase_mature_fasta_version=mirbase-mature-v22.1.fa
    export genome_fasta="${genome_fasta_dir}/${shortstack_genome_fasta_name}"

    # External data URLs
    export mirbase_fasta_url="https://mirbase.org/download_version_files/22.1/"

    # Set number of CPUs to use
    export threads=46

    # Initialize arrays
    export trimmed_fastqs_array=()

# 3 Load [ShortStack](https://github.com/MikeAxtell/ShortStack) conda environment

If this is successful, the first line of output should show that the
Python being used is the one in your
$$ShortStack$$(<https://github.com/MikeAxtell/ShortStack> conda
environment path.

E.g.

`python:         /home/sam/programs/mambaforge/envs/mirmachine_env/bin/python`

``` r
use_condaenv(condaenv = shortstack_conda_env_name, conda = shortstack_cond_path)

# Check successful env loading
py_config()
```

    python:         /home/sam/programs/mambaforge/envs/ShortStack4_env/bin/python
    libpython:      /home/sam/programs/mambaforge/envs/ShortStack4_env/lib/libpython3.10.so
    pythonhome:     /home/sam/programs/mambaforge/envs/ShortStack4_env:/home/sam/programs/mambaforge/envs/ShortStack4_env
    version:        3.10.13 | packaged by conda-forge | (main, Oct 26 2023, 18:07:37) [GCC 12.3.0]
    numpy:          /home/sam/programs/mambaforge/envs/ShortStack4_env/lib/python3.10/site-packages/numpy
    numpy_version:  1.26.0

    NOTE: Python version was forced by use_python() function

# 4 Download [miRBase](https://mirbase.org/) mature miRNA FastA

``` bash
# Load bash variables into memory
source .bashvars

wget \
--directory-prefix ${deep_dive_data_dir} \
--recursive \
--no-check-certificate \
--continue \
--no-host-directories \
--no-directories \
--no-parent \
--quiet \
--execute robots=off \
 ${mirbase_fasta_url}/${mirbase_mature_fasta}

# Rename to indicate miRBase FastA version
mv ${deep_dive_data_dir}/${mirbase_mature_fasta} ${deep_dive_data_dir}/${mirbase_mature_fasta_version}

ls -lh "${deep_dive_data_dir}"
```

    total 13M
    drwxr-xr-x 2 shedurkin labmembers 4.0K Nov 14 09:39 blast_dbs
    -rw-r--r-- 1 shedurkin labmembers 3.7M Dec  1 08:51 mirbase-mature-v22.1.fa
    -rw-r--r-- 1 shedurkin labmembers 3.7M Nov 30 08:45 mirbase-mature-v22.1-no_spaces.fa
    -rw-r--r-- 1 shedurkin labmembers 3.7M Nov 14 15:35 mirbase-mature-v22.1-no_U.fa
    -rw-r--r-- 1 shedurkin labmembers 726K Nov 14 09:39 mirgene-mature-all-v2.1.fa
    -rw-r--r-- 1 shedurkin labmembers 726K Nov 14 15:35 mirgene-mature-all-v2.1-no_U.fa

# 5 Run ShortStack

## 5.1 Modify genome filename for ShortStack compatability

``` bash
# Load bash variables into memory
source .bashvars

# Copy genome FastA to ShortStack-compatible filename (ending with .fa)
cp ${genome_fasta_dir}/${genome_fasta_name} ${genome_fasta_dir}/${shortstack_genome_fasta_name}

# Confirm
ls -lh ${genome_fasta_dir}/${shortstack_genome_fasta_name}
```

    -rw-r--r-- 1 shedurkin labmembers 360M Dec  1 08:51 /home/shared/8TB_HDD_02/shedurkin/deep-dive/F-Pmea/data/Pmea/Pocillopora_meandrina_HIv1.assembly.fa

## 5.2 Excecute ShortStack command

Uses the `--dn_mirna` option to identify miRNAs in the genome, without
relying on the `--known_miRNAs`.

This part of the code redirects the output of `time` to the end of
`shortstack.log` file.

- `; } \ 2>> ${output_dir_top}/shortstack.log`

``` bash
# Load bash variables into memory
source .bashvars

# Create array of trimmed FastQs
trimmed_fastqs_array=(${trimmed_fastqs_dir}/${trimmed_fastqs_pattern})


# Pass array contents to new variable as space-delimited list
trimmed_fastqs_list=$(echo "${trimmed_fastqs_array[*]}")


###### Run ShortStack ######
{ time \
ShortStack \
--genomefile "${genome_fasta}" \
--readfile ${trimmed_fastqs_list} \
--known_miRNAs ${deep_dive_data_dir}/${mirbase_mature_fasta_version} \
--dn_mirna \
--threads ${threads} \
--outdir ${output_dir_top}/ShortStack_out \
&> ${output_dir_top}/shortstack.log ; } \
2>> ${output_dir_top}/shortstack.log
```

## 5.3 Check runtime

``` bash
# Load bash variables into memory
source .bashvars

tail -n 3 ${output_dir_top}/shortstack.log \
| grep "real" \
| awk '{print "ShortStack runtime:" "\t" $2}'
```

    ShortStack runtime: 101m1.563s

# 6 Results

## 6.1 ShortStack synopsis

``` bash
# Load bash variables into memory
source .bashvars

tail -n 20 ${output_dir_top}/shortstack.log
```

    Screening of possible de novo microRNAs

    No microRNA loci were found!

    Writing final files

    Non-MIRNA loci by DicerCall:
    N 7191
    23 45
    22 39
    21 18
    24 13

    Thu 30 Nov 2023 17:31:28 -0800 PST
    Run Completed!

    real    101m1.563s
    user    1946m43.958s
    sys 621m8.203s

ShortStack didn’t identify *any* miRNAs.

## 6.2 Inspect `Results.txt`

``` bash
# Load bash variables into memory
source .bashvars

head ${output_dir_top}/ShortStack_out/Results.txt

echo ""
echo "----------------------------------------------------------"
echo ""

echo "Nummber of potential loci:"
awk '(NR>1)' ${output_dir_top}/ShortStack_out/Results.txt | wc -l
```

    Locus   Name    Chrom   Start   End Length  Reads   UniqueReads FracTop Strand  MajorRNA    MajorRNAReads   Short   Long    21  22  23  24  DicerCall   MIRNA   known_miRNAs
    Pocillopora_meandrina_HIv1___Sc0000000:9092-9520    Cluster_1   Pocillopora_meandrina_HIv1___Sc0000000  9092    9520    429 22848   607 0.5324317226890757  .   GGGGGUAUAGCUCAGUGGUAGAGCA   4414    2819    8702    1285    8803    362 877 N   N   NA
    Pocillopora_meandrina_HIv1___Sc0000000:53578-53997  Cluster_2   Pocillopora_meandrina_HIv1___Sc0000000  53578   53997   420 570 27  0.5035087719298246  .   GCCUAAGUUGCUUGGAACA 139 566 4   0   0   0   0   N   N   NA
    Pocillopora_meandrina_HIv1___Sc0000000:150243-150718    Cluster_3   Pocillopora_meandrina_HIv1___Sc0000000  150243  150718  476 7840    538 0.49923469387755104 .   UUAUGUGAUGAGUAUGUUAAUGUAC   875 1692    3509    1272    749 280 338 N   N   NA
    Pocillopora_meandrina_HIv1___Sc0000000:173728-174150    Cluster_4   Pocillopora_meandrina_HIv1___Sc0000000  173728  174150  423 2521    135 0.50059500198334    .   CAACCAGAUCACAGCAAUCAAA  438 212 10  83  1269    887 60  22  N   NA
    Pocillopora_meandrina_HIv1___Sc0000000:187652-188072    Cluster_5   Pocillopora_meandrina_HIv1___Sc0000000  187652  188072  421 439 54  0.46924829157175396 .   AUAAAUGUCACUACAAGAAACCUGA   103 0   423 2   2   5   7   N   N   NA
    Pocillopora_meandrina_HIv1___Sc0000000:485727-486254    Cluster_6   Pocillopora_meandrina_HIv1___Sc0000000  485727  486254  528 1431    279 0.5122292103424179  .   UUGCACUAGAACAGACUGUGCUUCC   257 132 1224    14  22  6   33  N   N   NA
    Pocillopora_meandrina_HIv1___Sc0000000:503576-504000    Cluster_7   Pocillopora_meandrina_HIv1___Sc0000000  503576  504000  425 403 122 0.9280397022332506  +   GGGGGGGGGGGGGGGGGGGCCCCCG   54  16  371 8   0   4   4   N   N   NA
    Pocillopora_meandrina_HIv1___Sc0000000:525310-527341    Cluster_8   Pocillopora_meandrina_HIv1___Sc0000000  525310  527341  2032    31666   4090    0.4850628434282827  .   UUUUCGUCACUUUCUUCAGCCUCAG   1838    275 29434   101 203 617 1036    N   N   NA
    Pocillopora_meandrina_HIv1___Sc0000000:541267-541723    Cluster_9   Pocillopora_meandrina_HIv1___Sc0000000  541267  541723  457 1123    107 0.4701691896705254  .   UCGGUGUGCUAUCGUGUUAGUGAGU   185 0   1112    0   5   4   2   N   N   NA

    ----------------------------------------------------------

    Nummber of potential loci:
    7306

Column 20 of the `Results.txt` file identifies if a cluster is a miRNA
or not (`Y` or `N`).

``` bash
# Load bash variables into memory
source .bashvars

echo "Number of loci characterized as miRNA:"
awk '$20=="Y" {print $0}' ${output_dir_top}/ShortStack_out/Results.txt \
| wc -l
echo ""

echo "----------------------------------------------------------"

echo ""
echo "Number of loci _not_ characterized as miRNA:"
awk '$20=="N" {print $0}' ${output_dir_top}/ShortStack_out/Results.txt \
| wc -l
```

    Number of loci characterized as miRNA:
    0

    ----------------------------------------------------------

    Number of loci _not_ characterized as miRNA:
    7306

Column 21 of the `Results.txt` file identifies if a cluster aligned to a
known miRNA (miRBase) or not (`Y` or `NA`).

Since there are no miRNAs, the following code will *not* print any
output.

The `echo` command after the `awk` command is simply there to prove that
the chunk executed.

``` bash
# Load bash variables into memory
source .bashvars

echo "Number of loci matching miRBase miRNAs:"
awk '$21!="NA" {print $0}' ${output_dir_top}/ShortStack_out/Results.txt \
| wc -l
echo ""

echo "----------------------------------------------------------"

echo ""
echo "Number of loci _not_ matching miRBase miRNAs:"
awk '$21=="NA" {print $0}' ${output_dir_top}/ShortStack_out/Results.txt \
| wc -l
```

    Number of loci matching miRBase miRNAs:
    86

    ----------------------------------------------------------

    Number of loci _not_ matching miRBase miRNAs:
    7221

Although there are loci with matches to miRBase miRNAs, ShortStack did
*not* annotated these clusters as miRNAs likely [because they do not
*also* match secondary structure
criteria](https://github.com/MikeAxtell/ShortStack#mirna-annotation).

### 6.2.1 Directory tree of all ShortStack outputs

Many of these are large (by GitHub standards) BAM files, so will not be
added to the repo.

Additionally, it’s unlikely we’ll utilize most of the other files
(bigwig) generated by ShortStack.

``` bash
# Load bash variables into memory
source .bashvars

tree -h ${output_dir_top}/
```

    /home/shared/8TB_HDD_02/shedurkin/deep-dive/F-Pmea/output/13-Pmea-sRNAseq-ShortStack/
    ├── [ 28K]  shortstack.log
    └── [ 20K]  ShortStack_out
        ├── [ 45K]  alignment_details.tsv
        ├── [730K]  Counts.txt
        ├── [160K]  known_miRNAs.gff3
        ├── [1.8M]  known_miRNAs_unaligned.fasta
        ├── [ 13M]  merged_alignments_21_m.bw
        ├── [ 13M]  merged_alignments_21_p.bw
        ├── [ 13M]  merged_alignments_22_m.bw
        ├── [ 13M]  merged_alignments_22_p.bw
        ├── [ 24M]  merged_alignments_23-24_m.bw
        ├── [ 24M]  merged_alignments_23-24_p.bw
        ├── [2.1G]  merged_alignments.bam
        ├── [151K]  merged_alignments.bam.csi
        ├── [ 95M]  merged_alignments_other_m.bw
        ├── [ 96M]  merged_alignments_other_p.bw
        ├── [ 24M]  merged_alignments_sRNA-POC-47-S1-TP2.flexbar_trim.25bp_1.bw
        ├── [ 24M]  merged_alignments_sRNA-POC-47-S1-TP2.flexbar_trim.25bp_2.bw
        ├── [ 34M]  merged_alignments_sRNA-POC-48-S1-TP2.flexbar_trim.25bp_1.bw
        ├── [ 33M]  merged_alignments_sRNA-POC-48-S1-TP2.flexbar_trim.25bp_2.bw
        ├── [ 26M]  merged_alignments_sRNA-POC-50-S1-TP2.flexbar_trim.25bp_1.bw
        ├── [ 26M]  merged_alignments_sRNA-POC-50-S1-TP2.flexbar_trim.25bp_2.bw
        ├── [ 43M]  merged_alignments_sRNA-POC-53-S1-TP2.flexbar_trim.25bp_1.bw
        ├── [ 43M]  merged_alignments_sRNA-POC-53-S1-TP2.flexbar_trim.25bp_2.bw
        ├── [ 55M]  merged_alignments_sRNA-POC-57-S1-TP2.flexbar_trim.25bp_1.bw
        ├── [ 54M]  merged_alignments_sRNA-POC-57-S1-TP2.flexbar_trim.25bp_2.bw
        ├── [922K]  Results.gff3
        ├── [1.5M]  Results.txt
        ├── [194M]  sRNA-POC-47-S1-TP2.flexbar_trim.25bp_1.bam
        ├── [170K]  sRNA-POC-47-S1-TP2.flexbar_trim.25bp_1.bam.csi
        ├── [204M]  sRNA-POC-47-S1-TP2.flexbar_trim.25bp_2.bam
        ├── [172K]  sRNA-POC-47-S1-TP2.flexbar_trim.25bp_2.bam.csi
        ├── [206M]  sRNA-POC-48-S1-TP2.flexbar_trim.25bp_1.bam
        ├── [170K]  sRNA-POC-48-S1-TP2.flexbar_trim.25bp_1.bam.csi
        ├── [214M]  sRNA-POC-48-S1-TP2.flexbar_trim.25bp_2.bam
        ├── [170K]  sRNA-POC-48-S1-TP2.flexbar_trim.25bp_2.bam.csi
        ├── [188M]  sRNA-POC-50-S1-TP2.flexbar_trim.25bp_1.bam
        ├── [172K]  sRNA-POC-50-S1-TP2.flexbar_trim.25bp_1.bam.csi
        ├── [196M]  sRNA-POC-50-S1-TP2.flexbar_trim.25bp_2.bam
        ├── [173K]  sRNA-POC-50-S1-TP2.flexbar_trim.25bp_2.bam.csi
        ├── [242M]  sRNA-POC-53-S1-TP2.flexbar_trim.25bp_1.bam
        ├── [169K]  sRNA-POC-53-S1-TP2.flexbar_trim.25bp_1.bam.csi
        ├── [248M]  sRNA-POC-53-S1-TP2.flexbar_trim.25bp_2.bam
        ├── [169K]  sRNA-POC-53-S1-TP2.flexbar_trim.25bp_2.bam.csi
        ├── [225M]  sRNA-POC-57-S1-TP2.flexbar_trim.25bp_1.bam
        ├── [159K]  sRNA-POC-57-S1-TP2.flexbar_trim.25bp_1.bam.csi
        ├── [233M]  sRNA-POC-57-S1-TP2.flexbar_trim.25bp_2.bam
        └── [160K]  sRNA-POC-57-S1-TP2.flexbar_trim.25bp_2.bam.csi

    1 directory, 47 files

------------------------------------------------------------------------

# 7 Citations

<div id="refs" class="references csl-bib-body hanging-indent">

<div id="ref-axtell2013a" class="csl-entry">

Axtell, Michael J. 2013. “ShortStack: Comprehensive Annotation and
Quantification of Small RNA Genes.” *RNA* 19 (6): 740–51.
<https://doi.org/10.1261/rna.035279.112>.

</div>

<div id="ref-johnson2016a" class="csl-entry">

Johnson, Nathan R, Jonathan M Yeoh, Ceyda Coruh, and Michael J Axtell.
2016. “Improved Placement of Multi-Mapping Small RNAs.” *G3
Genes\|Genomes\|Genetics* 6 (7): 2103–11.
<https://doi.org/10.1534/g3.116.030452>.

</div>

<div id="ref-shahid2014" class="csl-entry">

Shahid, Saima, and Michael J. Axtell. 2014. “Identification and
Annotation of Small RNA Genes Using ShortStack.” *Methods* 67 (1):
20–27. <https://doi.org/10.1016/j.ymeth.2013.10.004>.

</div>

</div>
