% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{article}
\usepackage{amsmath,amssymb}
\usepackage{lmodern}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\usepackage[margin=1in]{geometry}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}
\newlength{\cslhangindent}
\setlength{\cslhangindent}{1.5em}
\newlength{\csllabelwidth}
\setlength{\csllabelwidth}{3em}
\newlength{\cslentryspacingunit} % times entry-spacing
\setlength{\cslentryspacingunit}{\parskip}
\newenvironment{CSLReferences}[2] % #1 hanging-ident, #2 entry spacing
 {% don't indent paragraphs
  \setlength{\parindent}{0pt}
  % turn on hanging indent if param 1 is 1
  \ifodd #1
  \let\oldpar\par
  \def\par{\hangindent=\cslhangindent\oldpar}
  \fi
  % set entry spacing
  \setlength{\parskip}{#2\cslentryspacingunit}
 }%
 {}
\usepackage{calc}
\newcommand{\CSLBlock}[1]{#1\hfill\break}
\newcommand{\CSLLeftMargin}[1]{\parbox[t]{\csllabelwidth}{#1}}
\newcommand{\CSLRightInline}[1]{\parbox[t]{\linewidth - \csllabelwidth}{#1}\break}
\newcommand{\CSLIndent}[1]{\hspace{\cslhangindent}#1}
\usepackage{booktabs}
\usepackage{longtable}
\usepackage{array}
\usepackage{multirow}
\usepackage{wrapfig}
\usepackage{float}
\usepackage{colortbl}
\usepackage{pdflscape}
\usepackage{tabu}
\usepackage{threeparttable}
\usepackage{threeparttablex}
\usepackage[normalem]{ulem}
\usepackage{makecell}
\usepackage{xcolor}
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same} % disable monospaced font for URLs
\hypersetup{
  pdftitle={13-Pmea-sRNAseq-ShortStack},
  pdfauthor={Sam White (modified by K Durkin for P. meandrina analysis)},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}

\title{13-Pmea-sRNAseq-ShortStack}
\author{Sam White (modified by K Durkin for P. meandrina analysis)}
\date{2023-11-30}

\begin{document}
\maketitle

{
\setcounter{tocdepth}{2}
\tableofcontents
}
Use \href{https://github.com/MikeAxtell/ShortStack}{ShortStack} (\protect\hyperlink{ref-axtell2013a}{Axtell 2013}; \protect\hyperlink{ref-shahid2014}{Shahid and Axtell 2014}; \protect\hyperlink{ref-johnson2016a}{Johnson et al. 2016})to perform alignment of sRNAseq data and annotation of sRNA-producing genes.

The \emph{P.meandrina} genome will be used as the reference genome.

\begin{center}\rule{0.5\linewidth}{0.5pt}\end{center}

Inputs:

\begin{itemize}
\item
  Requires trimmed sRNAseq files generated by \href{https://github.com/urol-e5/deep-dive/blob/main/F-Pmea/code/08-Pmea-sRNAseq-trimming.Rmd}{08-Pmea-sRNAseq-trimming.Rmd}

  \begin{itemize}
  \tightlist
  \item
    Filenames formatted: \texttt{*flexbar\_trim.25bp*.gz}
  \end{itemize}
\item
  \emph{P.meandrina} genome FastA. See \href{https://github.com/urol-e5/deep-dive/blob/main/F-Pmea/code/12-Pmea-sRNAseq-MirMachine.Rmd}{12-Pmea-sRNAseq-MirMachine.Rmd} for download info if needed.
\end{itemize}

Outputs:

\begin{itemize}
\tightlist
\item
  See \href{https://github.com/MikeAxtell/ShortStack\#outputs}{ShortStack outputs documentation} for full list and detailed descriptions.
\end{itemize}

Software requirements:

\begin{itemize}
\tightlist
\item
  Utilizes a \href{https://github.com/MikeAxtell/ShortStack\#installation}{ShortStack} Conda/Mamba environment, per the installation instructions.
\end{itemize}

Replace with name of your ShortStack environment and the path to the corresponding conda installation (find this \emph{after} you've activated the environment).

E.g.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Activate environment}
\ExtensionTok{conda}\NormalTok{ activate ShortStack4\_env}

\CommentTok{\# Find conda path}
\FunctionTok{which}\NormalTok{ conda}
\end{Highlighting}
\end{Shaded}

\begin{center}\rule{0.5\linewidth}{0.5pt}\end{center}

\hypertarget{set-r-variables}{%
\section{Set R variables}\label{set-r-variables}}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{shortstack\_conda\_env\_name }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\StringTok{"ShortStack4\_env"}\NormalTok{)}
\NormalTok{shortstack\_cond\_path }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\StringTok{"/home/sam/programs/mambaforge/condabin/conda"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{create-a-bash-variables-file}{%
\section{Create a Bash variables file}\label{create-a-bash-variables-file}}

This allows usage of Bash variables across R Markdown chunks.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{\{}
\BuiltInTok{echo} \StringTok{"\#\#\#\# Assign Variables \#\#\#\#"}
\BuiltInTok{echo} \StringTok{""}

\BuiltInTok{echo} \StringTok{"\# Trimmed FastQ naming pattern"}
\BuiltInTok{echo} \StringTok{"export trimmed\_fastqs\_pattern=\textquotesingle{}*flexbar\_trim.25bp*.fastq.gz\textquotesingle{}"}

\BuiltInTok{echo} \StringTok{"\# Data directories"}
\BuiltInTok{echo} \StringTok{\textquotesingle{}export deep\_dive\_dir=/home/shared/8TB\_HDD\_02/shedurkin/deep{-}dive\textquotesingle{}}
\BuiltInTok{echo} \StringTok{\textquotesingle{}export deep\_dive\_data\_dir="$\{deep\_dive\_dir\}/data"\textquotesingle{}}
\BuiltInTok{echo} \StringTok{\textquotesingle{}export output\_dir\_top=$\{deep\_dive\_dir\}/F{-}Pmea/output/13{-}Pmea{-}sRNAseq{-}ShortStack\textquotesingle{}}
\BuiltInTok{echo} \StringTok{\textquotesingle{}export trimmed\_fastqs\_dir="$\{deep\_dive\_dir\}/F{-}Pmea/output/08{-}Pmea{-}sRNAseq{-}trimming/trimmed{-}reads"\textquotesingle{}}
\BuiltInTok{echo} \StringTok{""}

\BuiltInTok{echo} \StringTok{"\# Input/Output files"}
\BuiltInTok{echo} \StringTok{\textquotesingle{}export genome\_fasta\_dir=$\{deep\_dive\_dir\}/F{-}Pmea/data/Pmea\textquotesingle{}}
\BuiltInTok{echo} \StringTok{\textquotesingle{}export genome\_fasta\_name="Pocillopora\_meandrina\_HIv1.assembly.fasta"\textquotesingle{}}
\BuiltInTok{echo} \StringTok{\textquotesingle{}export shortstack\_genome\_fasta\_name="Pocillopora\_meandrina\_HIv1.assembly.fa"\textquotesingle{}}
\BuiltInTok{echo} \StringTok{\textquotesingle{}export mirbase\_mature\_fasta=mature.fa\textquotesingle{}}
\BuiltInTok{echo} \StringTok{\textquotesingle{}export mirbase\_mature\_fasta\_version=mirbase{-}mature{-}v22.1.fa\textquotesingle{}}
\BuiltInTok{echo} \StringTok{\textquotesingle{}export genome\_fasta="$\{genome\_fasta\_dir\}/$\{shortstack\_genome\_fasta\_name\}"\textquotesingle{}}
\BuiltInTok{echo} \StringTok{""}

\BuiltInTok{echo} \StringTok{"\# External data URLs"}
\BuiltInTok{echo} \StringTok{\textquotesingle{}export mirbase\_fasta\_url="https://mirbase.org/download\_version\_files/22.1/"\textquotesingle{}}
\BuiltInTok{echo} \StringTok{""}

\BuiltInTok{echo} \StringTok{"\# Set number of CPUs to use"}
\BuiltInTok{echo} \StringTok{\textquotesingle{}export threads=46\textquotesingle{}}
\BuiltInTok{echo} \StringTok{""}

\BuiltInTok{echo} \StringTok{"\# Initialize arrays"}
\BuiltInTok{echo} \StringTok{\textquotesingle{}export trimmed\_fastqs\_array=()\textquotesingle{}}


\KeywordTok{\}} \OperatorTok{\textgreater{}}\NormalTok{ .bashvars}

\FunctionTok{cat}\NormalTok{ .bashvars}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
#### Assign Variables ####

# Trimmed FastQ naming pattern
export trimmed_fastqs_pattern='*flexbar_trim.25bp*.fastq.gz'
# Data directories
export deep_dive_dir=/home/shared/8TB_HDD_02/shedurkin/deep-dive
export deep_dive_data_dir="${deep_dive_dir}/data"
export output_dir_top=${deep_dive_dir}/F-Pmea/output/13-Pmea-sRNAseq-ShortStack
export trimmed_fastqs_dir="${deep_dive_dir}/F-Pmea/output/08-Pmea-sRNAseq-trimming/trimmed-reads"

# Input/Output files
export genome_fasta_dir=${deep_dive_dir}/F-Pmea/data/Pmea
export genome_fasta_name="Pocillopora_meandrina_HIv1.assembly.fasta"
export shortstack_genome_fasta_name="Pocillopora_meandrina_HIv1.assembly.fa"
export mirbase_mature_fasta=mature.fa
export mirbase_mature_fasta_version=mirbase-mature-v22.1.fa
export genome_fasta="${genome_fasta_dir}/${shortstack_genome_fasta_name}"

# External data URLs
export mirbase_fasta_url="https://mirbase.org/download_version_files/22.1/"

# Set number of CPUs to use
export threads=46

# Initialize arrays
export trimmed_fastqs_array=()
\end{verbatim}

\hypertarget{load-shortstack-conda-environment}{%
\section{\texorpdfstring{Load \href{https://github.com/MikeAxtell/ShortStack}{ShortStack} conda environment}{Load ShortStack conda environment}}\label{load-shortstack-conda-environment}}

If this is successful, the first line of output should show that the Python being used is the one in your {[}ShortStack{]}(\url{https://github.com/MikeAxtell/ShortStack} conda environment path.

E.g.

\texttt{python:\ \ \ \ \ \ \ \ \ /home/sam/programs/mambaforge/envs/mirmachine\_env/bin/python}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{use\_condaenv}\NormalTok{(}\AttributeTok{condaenv =}\NormalTok{ shortstack\_conda\_env\_name, }\AttributeTok{conda =}\NormalTok{ shortstack\_cond\_path)}

\CommentTok{\# Check successful env loading}
\FunctionTok{py\_config}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
python:         /home/sam/programs/mambaforge/envs/ShortStack4_env/bin/python
libpython:      /home/sam/programs/mambaforge/envs/ShortStack4_env/lib/libpython3.10.so
pythonhome:     /home/sam/programs/mambaforge/envs/ShortStack4_env:/home/sam/programs/mambaforge/envs/ShortStack4_env
version:        3.10.13 | packaged by conda-forge | (main, Oct 26 2023, 18:07:37) [GCC 12.3.0]
numpy:          /home/sam/programs/mambaforge/envs/ShortStack4_env/lib/python3.10/site-packages/numpy
numpy_version:  1.26.0

NOTE: Python version was forced by use_python() function
\end{verbatim}

\hypertarget{download-mirbase-mature-mirna-fasta}{%
\section{\texorpdfstring{Download \href{https://mirbase.org/}{miRBase} mature miRNA FastA}{Download miRBase mature miRNA FastA}}\label{download-mirbase-mature-mirna-fasta}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load bash variables into memory}
\BuiltInTok{source}\NormalTok{ .bashvars}

\FunctionTok{wget} \DataTypeTok{\textbackslash{}}
\NormalTok{{-}{-}directory{-}prefix }\VariableTok{$\{deep\_dive\_data\_dir\}} \DataTypeTok{\textbackslash{}}
\NormalTok{{-}{-}recursive }\DataTypeTok{\textbackslash{}}
\NormalTok{{-}{-}no{-}check{-}certificate }\DataTypeTok{\textbackslash{}}
\NormalTok{{-}{-}continue }\DataTypeTok{\textbackslash{}}
\NormalTok{{-}{-}no{-}host{-}directories }\DataTypeTok{\textbackslash{}}
\NormalTok{{-}{-}no{-}directories }\DataTypeTok{\textbackslash{}}
\NormalTok{{-}{-}no{-}parent }\DataTypeTok{\textbackslash{}}
\NormalTok{{-}{-}quiet }\DataTypeTok{\textbackslash{}}
\NormalTok{{-}{-}execute robots=off }\DataTypeTok{\textbackslash{}}
 \VariableTok{$\{mirbase\_fasta\_url\}}\NormalTok{/}\VariableTok{$\{mirbase\_mature\_fasta\}}

\CommentTok{\# Rename to indicate miRBase FastA version}
\FunctionTok{mv} \VariableTok{$\{deep\_dive\_data\_dir\}}\NormalTok{/}\VariableTok{$\{mirbase\_mature\_fasta\}} \VariableTok{$\{deep\_dive\_data\_dir\}}\NormalTok{/}\VariableTok{$\{mirbase\_mature\_fasta\_version\}}

\FunctionTok{ls} \AttributeTok{{-}lh} \StringTok{"}\VariableTok{$\{deep\_dive\_data\_dir\}}\StringTok{"}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
total 13M
drwxr-xr-x 2 shedurkin labmembers 4.0K Nov 14 09:39 blast_dbs
-rw-r--r-- 1 shedurkin labmembers 3.7M Dec  1 08:48 mirbase-mature-v22.1.fa
-rw-r--r-- 1 shedurkin labmembers 3.7M Nov 30 08:45 mirbase-mature-v22.1-no_spaces.fa
-rw-r--r-- 1 shedurkin labmembers 3.7M Nov 14 15:35 mirbase-mature-v22.1-no_U.fa
-rw-r--r-- 1 shedurkin labmembers 726K Nov 14 09:39 mirgene-mature-all-v2.1.fa
-rw-r--r-- 1 shedurkin labmembers 726K Nov 14 15:35 mirgene-mature-all-v2.1-no_U.fa
\end{verbatim}

\hypertarget{run-shortstack}{%
\section{Run ShortStack}\label{run-shortstack}}

\hypertarget{modify-genome-filename-for-shortstack-compatability}{%
\subsection{Modify genome filename for ShortStack compatability}\label{modify-genome-filename-for-shortstack-compatability}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load bash variables into memory}
\BuiltInTok{source}\NormalTok{ .bashvars}

\CommentTok{\# Copy genome FastA to ShortStack{-}compatible filename (ending with .fa)}
\FunctionTok{cp} \VariableTok{$\{genome\_fasta\_dir\}}\NormalTok{/}\VariableTok{$\{genome\_fasta\_name\}} \VariableTok{$\{genome\_fasta\_dir\}}\NormalTok{/}\VariableTok{$\{shortstack\_genome\_fasta\_name\}}

\CommentTok{\# Confirm}
\FunctionTok{ls} \AttributeTok{{-}lh} \VariableTok{$\{genome\_fasta\_dir\}}\NormalTok{/}\VariableTok{$\{shortstack\_genome\_fasta\_name\}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
-rw-r--r-- 1 shedurkin labmembers 360M Dec  1 08:46 /home/shared/8TB_HDD_02/shedurkin/deep-dive/F-Pmea/data/Pmea/Pocillopora_meandrina_HIv1.assembly.fa
\end{verbatim}

\hypertarget{excecute-shortstack-command}{%
\subsection{Excecute ShortStack command}\label{excecute-shortstack-command}}

Uses the \texttt{-\/-dn\_mirna} option to identify miRNAs in the genome, without relying on the \texttt{-\/-known\_miRNAs}.

This part of the code redirects the output of \texttt{time} to the end of \texttt{shortstack.log} file.

\begin{itemize}
\tightlist
\item
  \texttt{;\ \}\ \textbackslash{}\ 2\textgreater{}\textgreater{}\ \$\{output\_dir\_top\}/shortstack.log}
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load bash variables into memory}
\BuiltInTok{source}\NormalTok{ .bashvars}

\CommentTok{\# Create array of trimmed FastQs}
\VariableTok{trimmed\_fastqs\_array}\OperatorTok{=}\VariableTok{($\{trimmed\_fastqs\_dir\}}\NormalTok{/}\VariableTok{$\{trimmed\_fastqs\_pattern\})}


\CommentTok{\# Pass array contents to new variable as space{-}delimited list}
\VariableTok{trimmed\_fastqs\_list}\OperatorTok{=}\VariableTok{$(}\BuiltInTok{echo} \StringTok{"}\VariableTok{$\{trimmed\_fastqs\_array}\OperatorTok{[*]}\VariableTok{\}}\StringTok{"}\VariableTok{)}


\CommentTok{\#\#\#\#\#\# Run ShortStack \#\#\#\#\#\#}
\KeywordTok{\{} \BuiltInTok{time} \DataTypeTok{\textbackslash{}}
\NormalTok{ShortStack }\DataTypeTok{\textbackslash{}}
\NormalTok{{-}{-}genomefile }\StringTok{"}\VariableTok{$\{genome\_fasta\}}\StringTok{"} \DataTypeTok{\textbackslash{}}
\NormalTok{{-}{-}readfile }\VariableTok{$\{trimmed\_fastqs\_list\}} \DataTypeTok{\textbackslash{}}
\NormalTok{{-}{-}known\_miRNAs }\VariableTok{$\{deep\_dive\_data\_dir\}}\NormalTok{/}\VariableTok{$\{mirbase\_mature\_fasta\_version\}} \DataTypeTok{\textbackslash{}}
\NormalTok{{-}{-}dn\_mirna }\DataTypeTok{\textbackslash{}}
\NormalTok{{-}{-}threads }\VariableTok{$\{threads\}} \DataTypeTok{\textbackslash{}}
\NormalTok{{-}{-}outdir }\VariableTok{$\{output\_dir\_top\}}\NormalTok{/ShortStack\_out }\DataTypeTok{\textbackslash{}}
\OperatorTok{\&\textgreater{}} \VariableTok{$\{output\_dir\_top\}}\NormalTok{/shortstack.log }\KeywordTok{;} \KeywordTok{\}} \DataTypeTok{\textbackslash{}}
\DecValTok{2}\OperatorTok{\textgreater{}\textgreater{}} \VariableTok{$\{output\_dir\_top\}}\NormalTok{/shortstack.log}
\end{Highlighting}
\end{Shaded}

\hypertarget{check-runtime}{%
\subsection{Check runtime}\label{check-runtime}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load bash variables into memory}
\BuiltInTok{source}\NormalTok{ .bashvars}

\FunctionTok{tail} \AttributeTok{{-}n}\NormalTok{ 3 }\VariableTok{$\{output\_dir\_top\}}\NormalTok{/shortstack.log }\DataTypeTok{\textbackslash{}}
\KeywordTok{|} \FunctionTok{grep} \StringTok{"real"} \DataTypeTok{\textbackslash{}}
\KeywordTok{|} \FunctionTok{awk} \StringTok{\textquotesingle{}\{print "ShortStack runtime:" "\textbackslash{}t" $2\}\textquotesingle{}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
ShortStack runtime: 101m1.563s
\end{verbatim}

\hypertarget{results}{%
\section{Results}\label{results}}

\hypertarget{shortstack-synopsis}{%
\subsection{ShortStack synopsis}\label{shortstack-synopsis}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load bash variables into memory}
\BuiltInTok{source}\NormalTok{ .bashvars}

\FunctionTok{tail} \AttributeTok{{-}n}\NormalTok{ 20 }\VariableTok{$\{output\_dir\_top\}}\NormalTok{/shortstack.log}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}

Screening of possible de novo microRNAs

No microRNA loci were found!

Writing final files

Non-MIRNA loci by DicerCall:
N 7191
23 45
22 39
21 18
24 13

Thu 30 Nov 2023 17:31:28 -0800 PST
Run Completed!

real    101m1.563s
user    1946m43.958s
sys 621m8.203s
\end{verbatim}

ShortStack didn't identify \emph{any} miRNAs.

\hypertarget{inspect-results.txt}{%
\subsection{\texorpdfstring{Inspect \texttt{Results.txt}}{Inspect Results.txt}}\label{inspect-results.txt}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load bash variables into memory}
\BuiltInTok{source}\NormalTok{ .bashvars}

\FunctionTok{head} \VariableTok{$\{output\_dir\_top\}}\NormalTok{/ShortStack\_out/Results.txt}

\BuiltInTok{echo} \StringTok{""}
\BuiltInTok{echo} \StringTok{"{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}"}
\BuiltInTok{echo} \StringTok{""}

\BuiltInTok{echo} \StringTok{"Nummber of potential loci:"}
\FunctionTok{awk} \StringTok{\textquotesingle{}(NR\textgreater{}1)\textquotesingle{}} \VariableTok{$\{output\_dir\_top\}}\NormalTok{/ShortStack\_out/Results.txt }\KeywordTok{|} \FunctionTok{wc} \AttributeTok{{-}l}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
Locus   Name    Chrom   Start   End Length  Reads   UniqueReads FracTop Strand  MajorRNA    MajorRNAReads   Short   Long    21  22  23  24  DicerCall   MIRNA   known_miRNAs
Pocillopora_meandrina_HIv1___Sc0000000:9092-9520    Cluster_1   Pocillopora_meandrina_HIv1___Sc0000000  9092    9520    429 22848   607 0.5324317226890757  .   GGGGGUAUAGCUCAGUGGUAGAGCA   4414    2819    8702    1285    8803    362 877 N   N   NA
Pocillopora_meandrina_HIv1___Sc0000000:53578-53997  Cluster_2   Pocillopora_meandrina_HIv1___Sc0000000  53578   53997   420 570 27  0.5035087719298246  .   GCCUAAGUUGCUUGGAACA 139 566 4   0   0   0   0   N   N   NA
Pocillopora_meandrina_HIv1___Sc0000000:150243-150718    Cluster_3   Pocillopora_meandrina_HIv1___Sc0000000  150243  150718  476 7840    538 0.49923469387755104 .   UUAUGUGAUGAGUAUGUUAAUGUAC   875 1692    3509    1272    749 280 338 N   N   NA
Pocillopora_meandrina_HIv1___Sc0000000:173728-174150    Cluster_4   Pocillopora_meandrina_HIv1___Sc0000000  173728  174150  423 2521    135 0.50059500198334    .   CAACCAGAUCACAGCAAUCAAA  438 212 10  83  1269    887 60  22  N   NA
Pocillopora_meandrina_HIv1___Sc0000000:187652-188072    Cluster_5   Pocillopora_meandrina_HIv1___Sc0000000  187652  188072  421 439 54  0.46924829157175396 .   AUAAAUGUCACUACAAGAAACCUGA   103 0   423 2   2   5   7   N   N   NA
Pocillopora_meandrina_HIv1___Sc0000000:485727-486254    Cluster_6   Pocillopora_meandrina_HIv1___Sc0000000  485727  486254  528 1431    279 0.5122292103424179  .   UUGCACUAGAACAGACUGUGCUUCC   257 132 1224    14  22  6   33  N   N   NA
Pocillopora_meandrina_HIv1___Sc0000000:503576-504000    Cluster_7   Pocillopora_meandrina_HIv1___Sc0000000  503576  504000  425 403 122 0.9280397022332506  +   GGGGGGGGGGGGGGGGGGGCCCCCG   54  16  371 8   0   4   4   N   N   NA
Pocillopora_meandrina_HIv1___Sc0000000:525310-527341    Cluster_8   Pocillopora_meandrina_HIv1___Sc0000000  525310  527341  2032    31666   4090    0.4850628434282827  .   UUUUCGUCACUUUCUUCAGCCUCAG   1838    275 29434   101 203 617 1036    N   N   NA
Pocillopora_meandrina_HIv1___Sc0000000:541267-541723    Cluster_9   Pocillopora_meandrina_HIv1___Sc0000000  541267  541723  457 1123    107 0.4701691896705254  .   UCGGUGUGCUAUCGUGUUAGUGAGU   185 0   1112    0   5   4   2   N   N   NA

----------------------------------------------------------

Nummber of potential loci:
7306
\end{verbatim}

Column 20 of the \texttt{Results.txt} file identifies if a cluster is a miRNA or not (\texttt{Y} or \texttt{N}).

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load bash variables into memory}
\BuiltInTok{source}\NormalTok{ .bashvars}

\BuiltInTok{echo} \StringTok{"Number of loci characterized as miRNA:"}
\FunctionTok{awk} \StringTok{\textquotesingle{}$20=="Y" \{print $0\}\textquotesingle{}} \VariableTok{$\{output\_dir\_top\}}\NormalTok{/ShortStack\_out/Results.txt }\DataTypeTok{\textbackslash{}}
\KeywordTok{|} \FunctionTok{wc} \AttributeTok{{-}l}
\BuiltInTok{echo} \StringTok{""}

\BuiltInTok{echo} \StringTok{"{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}"}

\BuiltInTok{echo} \StringTok{""}
\BuiltInTok{echo} \StringTok{"Number of loci \_not\_ characterized as miRNA:"}
\FunctionTok{awk} \StringTok{\textquotesingle{}$20=="N" \{print $0\}\textquotesingle{}} \VariableTok{$\{output\_dir\_top\}}\NormalTok{/ShortStack\_out/Results.txt }\DataTypeTok{\textbackslash{}}
\KeywordTok{|} \FunctionTok{wc} \AttributeTok{{-}l}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
Number of loci characterized as miRNA:
0

----------------------------------------------------------

Number of loci _not_ characterized as miRNA:
7306
\end{verbatim}

Column 21 of the \texttt{Results.txt} file identifies if a cluster aligned to a known miRNA (miRBase) or not (\texttt{Y} or \texttt{NA}).

Since there are no miRNAs, the following code will \emph{not} print any output.

The \texttt{echo} command after the \texttt{awk} command is simply there to prove that the chunk executed.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load bash variables into memory}
\BuiltInTok{source}\NormalTok{ .bashvars}

\BuiltInTok{echo} \StringTok{"Number of loci matching miRBase miRNAs:"}
\FunctionTok{awk} \StringTok{\textquotesingle{}$21!="NA" \{print $0\}\textquotesingle{}} \VariableTok{$\{output\_dir\_top\}}\NormalTok{/ShortStack\_out/Results.txt }\DataTypeTok{\textbackslash{}}
\KeywordTok{|} \FunctionTok{wc} \AttributeTok{{-}l}
\BuiltInTok{echo} \StringTok{""}

\BuiltInTok{echo} \StringTok{"{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}{-}"}

\BuiltInTok{echo} \StringTok{""}
\BuiltInTok{echo} \StringTok{"Number of loci \_not\_ matching miRBase miRNAs:"}
\FunctionTok{awk} \StringTok{\textquotesingle{}$21=="NA" \{print $0\}\textquotesingle{}} \VariableTok{$\{output\_dir\_top\}}\NormalTok{/ShortStack\_out/Results.txt }\DataTypeTok{\textbackslash{}}
\KeywordTok{|} \FunctionTok{wc} \AttributeTok{{-}l}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
Number of loci matching miRBase miRNAs:
86

----------------------------------------------------------

Number of loci _not_ matching miRBase miRNAs:
7221
\end{verbatim}

Although there are loci with matches to miRBase miRNAs, ShortStack did \emph{not} annotated these clusters as miRNAs likely \href{https://github.com/MikeAxtell/ShortStack\#mirna-annotation}{because they do not \emph{also} match secondary structure criteria}.

\hypertarget{directory-tree-of-all-shortstack-outputs}{%
\subsubsection{Directory tree of all ShortStack outputs}\label{directory-tree-of-all-shortstack-outputs}}

Many of these are large (by GitHub standards) BAM files, so will not be added to the repo.

Additionally, it's unlikely we'll utilize most of the other files (bigwig) generated by ShortStack.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load bash variables into memory}
\BuiltInTok{source}\NormalTok{ .bashvars}

\ExtensionTok{tree} \AttributeTok{{-}h} \VariableTok{$\{output\_dir\_top\}}\NormalTok{/}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
/home/shared/8TB_HDD_02/shedurkin/deep-dive/F-Pmea/output/13-Pmea-sRNAseq-ShortStack/
├── [ 28K]  shortstack.log
└── [ 20K]  ShortStack_out
    ├── [ 45K]  alignment_details.tsv
    ├── [730K]  Counts.txt
    ├── [160K]  known_miRNAs.gff3
    ├── [1.8M]  known_miRNAs_unaligned.fasta
    ├── [ 13M]  merged_alignments_21_m.bw
    ├── [ 13M]  merged_alignments_21_p.bw
    ├── [ 13M]  merged_alignments_22_m.bw
    ├── [ 13M]  merged_alignments_22_p.bw
    ├── [ 24M]  merged_alignments_23-24_m.bw
    ├── [ 24M]  merged_alignments_23-24_p.bw
    ├── [2.1G]  merged_alignments.bam
    ├── [151K]  merged_alignments.bam.csi
    ├── [ 95M]  merged_alignments_other_m.bw
    ├── [ 96M]  merged_alignments_other_p.bw
    ├── [ 24M]  merged_alignments_sRNA-POC-47-S1-TP2.flexbar_trim.25bp_1.bw
    ├── [ 24M]  merged_alignments_sRNA-POC-47-S1-TP2.flexbar_trim.25bp_2.bw
    ├── [ 34M]  merged_alignments_sRNA-POC-48-S1-TP2.flexbar_trim.25bp_1.bw
    ├── [ 33M]  merged_alignments_sRNA-POC-48-S1-TP2.flexbar_trim.25bp_2.bw
    ├── [ 26M]  merged_alignments_sRNA-POC-50-S1-TP2.flexbar_trim.25bp_1.bw
    ├── [ 26M]  merged_alignments_sRNA-POC-50-S1-TP2.flexbar_trim.25bp_2.bw
    ├── [ 43M]  merged_alignments_sRNA-POC-53-S1-TP2.flexbar_trim.25bp_1.bw
    ├── [ 43M]  merged_alignments_sRNA-POC-53-S1-TP2.flexbar_trim.25bp_2.bw
    ├── [ 55M]  merged_alignments_sRNA-POC-57-S1-TP2.flexbar_trim.25bp_1.bw
    ├── [ 54M]  merged_alignments_sRNA-POC-57-S1-TP2.flexbar_trim.25bp_2.bw
    ├── [922K]  Results.gff3
    ├── [1.5M]  Results.txt
    ├── [194M]  sRNA-POC-47-S1-TP2.flexbar_trim.25bp_1.bam
    ├── [170K]  sRNA-POC-47-S1-TP2.flexbar_trim.25bp_1.bam.csi
    ├── [204M]  sRNA-POC-47-S1-TP2.flexbar_trim.25bp_2.bam
    ├── [172K]  sRNA-POC-47-S1-TP2.flexbar_trim.25bp_2.bam.csi
    ├── [206M]  sRNA-POC-48-S1-TP2.flexbar_trim.25bp_1.bam
    ├── [170K]  sRNA-POC-48-S1-TP2.flexbar_trim.25bp_1.bam.csi
    ├── [214M]  sRNA-POC-48-S1-TP2.flexbar_trim.25bp_2.bam
    ├── [170K]  sRNA-POC-48-S1-TP2.flexbar_trim.25bp_2.bam.csi
    ├── [188M]  sRNA-POC-50-S1-TP2.flexbar_trim.25bp_1.bam
    ├── [172K]  sRNA-POC-50-S1-TP2.flexbar_trim.25bp_1.bam.csi
    ├── [196M]  sRNA-POC-50-S1-TP2.flexbar_trim.25bp_2.bam
    ├── [173K]  sRNA-POC-50-S1-TP2.flexbar_trim.25bp_2.bam.csi
    ├── [242M]  sRNA-POC-53-S1-TP2.flexbar_trim.25bp_1.bam
    ├── [169K]  sRNA-POC-53-S1-TP2.flexbar_trim.25bp_1.bam.csi
    ├── [248M]  sRNA-POC-53-S1-TP2.flexbar_trim.25bp_2.bam
    ├── [169K]  sRNA-POC-53-S1-TP2.flexbar_trim.25bp_2.bam.csi
    ├── [225M]  sRNA-POC-57-S1-TP2.flexbar_trim.25bp_1.bam
    ├── [159K]  sRNA-POC-57-S1-TP2.flexbar_trim.25bp_1.bam.csi
    ├── [233M]  sRNA-POC-57-S1-TP2.flexbar_trim.25bp_2.bam
    └── [160K]  sRNA-POC-57-S1-TP2.flexbar_trim.25bp_2.bam.csi

1 directory, 47 files
\end{verbatim}

\begin{center}\rule{0.5\linewidth}{0.5pt}\end{center}

\hypertarget{citations}{%
\section*{Citations}\label{citations}}
\addcontentsline{toc}{section}{Citations}

\hypertarget{refs}{}
\begin{CSLReferences}{1}{0}
\leavevmode\vadjust pre{\hypertarget{ref-axtell2013a}{}}%
Axtell, Michael J. 2013. {``ShortStack: Comprehensive Annotation and Quantification of Small RNA Genes.''} \emph{RNA} 19 (6): 740--51. \url{https://doi.org/10.1261/rna.035279.112}.

\leavevmode\vadjust pre{\hypertarget{ref-johnson2016a}{}}%
Johnson, Nathan R, Jonathan M Yeoh, Ceyda Coruh, and Michael J Axtell. 2016. {``Improved Placement of Multi-Mapping Small RNAs.''} \emph{G3 Genes\textbar Genomes\textbar Genetics} 6 (7): 2103--11. \url{https://doi.org/10.1534/g3.116.030452}.

\leavevmode\vadjust pre{\hypertarget{ref-shahid2014}{}}%
Shahid, Saima, and Michael J. Axtell. 2014. {``Identification and Annotation of Small RNA Genes Using ShortStack.''} \emph{Methods} 67 (1): 20--27. \url{https://doi.org/10.1016/j.ymeth.2013.10.004}.

\end{CSLReferences}

\end{document}
