---
title: "Closest genomic features to ncRNAs - Ptuh"
author: "Jill Ashey"
date: "2024-08-20"
output: html_document
---

## Closest genomic features to ncRNAs in Pocillopora tuahiniensis

This code will investigate the proximity of ncRNAs to other genomic features in Pocillopora tuahiniensis. The `bash` code snippets were done on the URI HPC server, Andromeda. I copied the Ptuh miRNA [gff](https://github.com/urol-e5/deep-dive/blob/main/F-Pmea/output/13.2.1-Pmea-sRNAseq-ShortStack-31bp-fastp-merged-cnidarian_miRBase/ShortStack_out/Results.gff3) and [fasta](https://github.com/urol-e5/deep-dive/blob/main/F-Pmea/output/13.2.1-Pmea-sRNAseq-ShortStack-31bp-fastp-merged-cnidarian_miRBase/ShortStack_out/mir.fasta) files onto Andromeda. We used the Pmeandrina genomic information for Ptuh. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Load packages
library(tidyverse)
```

## miRNA

Sort gene gff file. The gene gff file only has information about transcripts, CDSs, and exons. 
```{bash}
cd /data/putnamlab/jillashey/genome/Pmea
sort -k1,1 -k4,4n Pocillopora_meandrina_HIv1.genes.gff3 > Pocillopora_meandrina_HIv1.genes_sorted.gff3
```

Sort miRNA gff file and run `bedtools closest`.
```{bash}
cd /data/putnamlab/jillashey/e5/ncRNA_gff

sort -k1,1 -k4,4n Ptuh_Results.gff3 > Ptuh_Results_sorted.gff3

interactive 
module load BEDTools/2.30.0-GCC-11.3.0
bedtools closest -a Ptuh_Results_sorted.gff3 -b /data/putnamlab/jillashey/genome/Pmea/Pocillopora_meandrina_HIv1.genes_sorted.gff3 > Ptuh_output.bed

wc -l Ptuh_output.bed 
21604 Ptuh_output.bed

head Ptuh_output.bed 
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	Unknown_sRNA_locus	9092	9521	10813	+	.	ID=Cluster_1;DicerCall=N;MIRNA=N	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	CDS	10771	11117	.	+	0	Parent=Pocillopora_meandrina_HIv1___RNAseq.g20902.t1
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	Unknown_sRNA_locus	9092	9521	10813	+	.	ID=Cluster_1;DicerCall=N;MIRNA=N	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	exon	10771	11117	.	+	0	Parent=Pocillopora_meandrina_HIv1___RNAseq.g20902.t1
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	Unknown_sRNA_locus	9092	9521	10813	+	.	ID=Cluster_1;DicerCall=N;MIRNA=N	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	transcript	10771	23652	.	+	.	ID=Pocillopora_meandrina_HIv1___RNAseq.g20902.t1
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	Unknown_sRNA_locus	53578	53997	287	+	.	ID=Cluster_2;DicerCall=N;MIRNA=N	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	transcript	52050	53624	.	-	.	ID=Pocillopora_meandrina_HIv1___RNAseq.g20906.t1
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	Unknown_sRNA_locus	53578	53997	287	+	.	ID=Cluster_2;DicerCall=N;MIRNA=N	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	CDS	53573	53624	.	-	0	Parent=Pocillopora_meandrina_HIv1___RNAseq.g20906.t1
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	Unknown_sRNA_locus	53578	53997	287	+	.	ID=Cluster_2;DicerCall=N;MIRNA=N	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	exon	53573	53624	.	-	0	Parent=Pocillopora_meandrina_HIv1___RNAseq.g20906.t1
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	Unknown_sRNA_locus	150243	150718	2549	-	.	ID=Cluster_3;DicerCall=N;MIRNA=N	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	transcript	143552	155669	.	-	.	ID=Pocillopora_meandrina_HIv1___RNAseq.g20914.t1
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	Unknown_sRNA_locus	150243	150718	2549	-	.	ID=Cluster_3;DicerCall=N;MIRNA=N	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	CDS	150290	150371	.	-	1	Parent=Pocillopora_meandrina_HIv1___RNAseq.g20914.t1
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	Unknown_sRNA_locus	150243	150718	2549	-	.	ID=Cluster_3;DicerCall=N;MIRNA=N	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	exon	150290	150371	.	-	1	Parent=Pocillopora_meandrina_HIv1___RNAseq.g20914.t1
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	Unknown_sRNA_locus	150243	150718	2549	-	.	ID=Cluster_3;DicerCall=N;MIRNA=N	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	CDS	150573	150661	.	-	0	Parent=Pocillopora_meandrina_HIv1___RNAseq.g20914.t1
```

Success! Remove the unknown sRNA loci. 
```{bash}
awk 'BEGIN {OFS="\t"} $3 != "Unknown_sRNA_locus"' Ptuh_output.bed > filtered_Ptuh_output.bed
wc -l filtered_Ptuh_output.bed 
405 filtered_Ptuh_output.bed

head filtered_Ptuh_output.bed
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	siRNA22_locus	173728	174150	1257	+	.	ID=Cluster_4;DicerCall=22;MIRNA=N	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	CDS	174509	175333	.	-	0	Parent=Pocillopora_meandrina_HIv1___RNAseq.g20918.t1
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	siRNA22_locus	173728	174150	1257	+	.	ID=Cluster_4;DicerCall=22;MIRNA=N	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	exon	174509	175333	.	-	0	Parent=Pocillopora_meandrina_HIv1___RNAseq.g20918.t1
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	siRNA22_locus	173728	174150	1257	+	.	ID=Cluster_4;DicerCall=22;MIRNA=N	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	transcript	174509	176444	.	-	.	ID=Pocillopora_meandrina_HIv1___RNAseq.g20918.t1
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	MIRNA_hairpin	818027	818120	12096	+	.	ID=Cluster_19;DicerCall=23;MIRNA=Y	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	transcript	816355	820160	.	+	.	ID=Pocillopora_meandrina_HIv1___RNAseq.g21001.t1
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	mature_miRNA	818049	818070	3240	+	.	ID=Cluster_19.mature;Parent=Cluster_19	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	transcript	816355	820160	.	+	.	ID=Pocillopora_meandrina_HIv1___RNAseq.g21001.t1
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	miRNA-star	818079	818100	9	+	.	ID=Cluster_19.star;Parent=Cluster_19	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	transcript	816355	820160	.	+	.	ID=Pocillopora_meandrina_HIv1___RNAseq.g21001.t1
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	MIRNA_hairpin	2872019	2872110	177	+	.	ID=Cluster_34;DicerCall=21;MIRNA=Y	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	transcript	2868586	2871318	.	+	.	ID=Pocillopora_meandrina_HIv1___TS.g25957.t1
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	MIRNA_hairpin	2872019	2872110	177	+	.	ID=Cluster_34;DicerCall=21;MIRNA=Y	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	CDS	2870522	2871318	.	+	2	Parent=Pocillopora_meandrina_HIv1___TS.g25957.t1
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	MIRNA_hairpin	2872019	2872110	177	+	.	ID=Cluster_34;DicerCall=21;MIRNA=Y	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	exon	2870522	2871318	.	+	2	Parent=Pocillopora_meandrina_HIv1___TS.g25957.t1
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	mature_miRNA	2872041	2872061	110	+	.	ID=Cluster_34.mature;Parent=Cluster_34	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	transcript	2868586	2871318	.	+	.	ID=Pocillopora_meandrina_HIv1___TS.g25957.t1
```

If I subtract the 5th column (representing the end of the miRNA feature) by the 13th column (representing the beginning of the genomic feature), this will tell me if there is overlap of the two features. If the value is positive, there is overlap. If the value is negative, there is no overlap between the features. 
```{bash}
awk '{ $(NF+1) = $13 - $5 }1' OFS='\t' filtered_Ptuh_output.bed > filtered_Ptuh_output_overlap.bed

head filtered_Ptuh_output_overlap.bed
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	siRNA22_locus	173728	174150	1257	+	.	ID=Cluster_4;DicerCall=22;MIRNA=Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	CDS	174509	175333	.	-	0	Parent=Pocillopora_meandrina_HIv1___RNAseq.g20918.t1	359
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	siRNA22_locus	173728	174150	1257	+	.	ID=Cluster_4;DicerCall=22;MIRNA=Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	exon	174509	175333	.	-	0	Parent=Pocillopora_meandrina_HIv1___RNAseq.g20918.t1	359
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	siRNA22_locus	173728	174150	1257	+	.	ID=Cluster_4;DicerCall=22;MIRNA=Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	transcript	174509	176444	.	-	.	ID=Pocillopora_meandrina_HIv1___RNAseq.g20918.t1	359
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	MIRNA_hairpin	818027	818120	12096	+	.	ID=Cluster_19;DicerCall=23;MIRNA=Y	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	transcript	816355	820160	.	+	.	ID=Pocillopora_meandrina_HIv1___RNAseq.g21001.t1	-1765
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	mature_miRNA	818049	818070	3240	+	.	ID=Cluster_19.mature;Parent=Cluster_19	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	transcript	816355	820160	.	+	.	ID=Pocillopora_meandrina_HIv1___RNAseq.g21001.t1	-1715
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	miRNA-star	818079	818100	9	+	.	ID=Cluster_19.star;Parent=Cluster_19	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	transcript	816355	820160	.	+	.	ID=Pocillopora_meandrina_HIv1___RNAseq.g21001.t1	-1745
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	MIRNA_hairpin	2872019	2872110	177	+	.	ID=Cluster_34;DicerCall=21;MIRNA=Y	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	transcript	2868586	2871318	.	+	.	ID=Pocillopora_meandrina_HIv1___TS.g25957.t1	-3524
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	MIRNA_hairpin	2872019	2872110	177	+	.	ID=Cluster_34;DicerCall=21;MIRNA=Y	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	CDS	2870522	2871318	.	+	2	Parent=Pocillopora_meandrina_HIv1___TS.g25957.t1	-1588
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	MIRNA_hairpin	2872019	2872110	177	+	.	ID=Cluster_34;DicerCall=21;MIRNA=Y	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	exon	2870522	2871318	.	+	2	Parent=Pocillopora_meandrina_HIv1___TS.g25957.t1	-1588
Pocillopora_meandrina_HIv1___Sc0000000	ShortStack	mature_miRNA	2872041	2872061	110	+	.	ID=Cluster_34.mature;Parent=Cluster_34	Pocillopora_meandrina_HIv1___Sc0000000	AUGUSTUS	transcript	2868586	2871318	.	+	.	ID=Pocillopora_meandrina_HIv1___TS.g25957.t1	-3475
```

Count number of positive and negative values.
```{bash}
awk '{if ($NF < 0) neg++; else if ($NF > 0) pos++} END {print "Positive count:", pos; print "Negative count:", neg}' filtered_Ptuh_output_overlap.bed

Positive count: 99
Negative count: 306
```

Moving into R to visualize and calculate more stats.

Load Ptuh bed file into R. 
```{r}
bed <- read.delim("../output/19-bedtools-closest/miRNA_filtered_Ptuh_output_overlap.bed", header = F)
head(bed)
```

Keep specific columns and rename them. I'm going to label columns that relate to miRNAs with an "m" and columns that relate to genomic features with a "gf".
```{r}
bed_filt <- bed %>%
  dplyr::select(V1, V2, V3, V4, V5, V7, V9, V10, V11, V12, V13, V14, V16, V18, V19) %>%
  dplyr::rename("m_chrom" = "V1", "m_source" = "V2", "m_type" = "V3", "m_start" = "V4", "m_end" = "V5", "m_strand" = "V7", "m_attr" = "V9", "gf_chrom" = "V10", "gf_source" = "V11", "gf_type" = "V12", "gf_start" = "V13", "gf_end" = "V14", "gf_strand" = "V16", "gf_attr" = "V18", "overlap" = "V19")

unique(bed_filt$m_type)
unique(bed_filt$gf_type)
```

Remove the siRNA m_types and make column that has cluster IDs for miRNAs. 
```{r}
bed_filt <- bed_filt %>%
  dplyr::filter(!grepl("siRNA", m_type)) %>%
  mutate(m_ID = sub(";.*", "", m_attr))

# Remove .mature and .star from the ID column 
bed_filt$m_ID <- gsub(".mature", "", bed_filt$m_ID)
bed_filt$m_ID <- gsub(".star", "", bed_filt$m_ID)

length(unique(bed_filt$m_ID)) # check how many miRNAs are there - should be 37 for Ptuh
```

Discuss with e5 group how to present/visualize this data. 







